{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cluster/project/beltrao/gankin/software/miniconda3/envs/vnn/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[EnsemblRelease(release=75, species='homo_sapiens'),\n",
       " EnsemblRelease(release=111, species='homo_sapiens')]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "import swifter\n",
    "\n",
    "from pyensembl import EnsemblRelease\n",
    "from pyensembl.genome import Genome\n",
    "from pyensembl.shell import collect_all_installed_ensembl_releases\n",
    "collect_all_installed_ensembl_releases()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load UKB BIM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/tmp.20761925.dgankin/ipykernel_3600543/1028760957.py:5: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  bim_df = pd.read_csv(bim_file, sep='\\t', header=None)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs28659788</td>\n",
       "      <td>0</td>\n",
       "      <td>723307</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>rs116587930</td>\n",
       "      <td>0</td>\n",
       "      <td>727841</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>rs116720794</td>\n",
       "      <td>0</td>\n",
       "      <td>729632</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>rs3131972</td>\n",
       "      <td>0</td>\n",
       "      <td>752721</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>rs12184325</td>\n",
       "      <td>0</td>\n",
       "      <td>754105</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805421</th>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-92047842</td>\n",
       "      <td>0</td>\n",
       "      <td>16337</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805422</th>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-79443531</td>\n",
       "      <td>0</td>\n",
       "      <td>16356</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805423</th>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-79443532</td>\n",
       "      <td>0</td>\n",
       "      <td>16362</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805424</th>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-89025709</td>\n",
       "      <td>0</td>\n",
       "      <td>16390</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805425</th>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-79381726</td>\n",
       "      <td>0</td>\n",
       "      <td>16391</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>805426 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom            snp  cm     pos a1 a2\n",
       "0          1     rs28659788   0  723307  C  G\n",
       "1          1    rs116587930   0  727841  G  A\n",
       "2          1    rs116720794   0  729632  C  T\n",
       "3          1      rs3131972   0  752721  A  G\n",
       "4          1     rs12184325   0  754105  C  T\n",
       "...      ...            ...  ..     ... .. ..\n",
       "805421    MT  Affx-92047842   0   16337  C  T\n",
       "805422    MT  Affx-79443531   0   16356  T  C\n",
       "805423    MT  Affx-79443532   0   16362  T  C\n",
       "805424    MT  Affx-89025709   0   16390  G  A\n",
       "805425    MT  Affx-79381726   0   16391  G  A\n",
       "\n",
       "[805426 rows x 6 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_dir = \"/cluster/project/beltrao/gankin/vnn/data\"\n",
    "\n",
    "# Read the .bim file to get SNP information\n",
    "bim_file = data_dir + \"/ukb.bim\"\n",
    "bim_df = pd.read_csv(bim_file, sep='\\t', header=None)\n",
    "bim_df.columns = ['chrom', 'snp', 'cm', 'pos', 'a1', 'a2']\n",
    "bim_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Open Targets data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>gene_id_any_distance</th>\n",
       "      <th>gene_id_any</th>\n",
       "      <th>gene_id_prot_coding_distance</th>\n",
       "      <th>gene_id_prot_coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>30794</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>30794</td>\n",
       "      <td>rs1311411458</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 0.177, 'raw': -0.402967}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>34625.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>34625.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>55330</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>55330</td>\n",
       "      <td>rs185215913</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>{'phred': 0.431, 'raw': -0.275244}</td>\n",
       "      <td>{'gnomad_afr': 0.016711590296495958, 'gnomad_a...</td>\n",
       "      <td>10089.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>10089.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>57248</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>57248</td>\n",
       "      <td>rs1249676487</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>{'phred': 7.268, 'raw': 0.304146}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>8171.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>8171.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>66381</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>66381</td>\n",
       "      <td>rs538833530</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 1.389, 'raw': -0.101295}</td>\n",
       "      <td>{'gnomad_afr': 0.15321011673151752, 'gnomad_am...</td>\n",
       "      <td>962.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>962.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>67631</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>67631</td>\n",
       "      <td>rs533896527</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.033, 'raw': -0.632235}</td>\n",
       "      <td>{'gnomad_afr': 0.041812315138478086, 'gnomad_a...</td>\n",
       "      <td>2212.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>2212.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72878704</th>\n",
       "      <td>X</td>\n",
       "      <td>156006466</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>155236131</td>\n",
       "      <td>rs1218662062</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 1.118, 'raw': -0.133403}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>8770.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "      <td>8770.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72878705</th>\n",
       "      <td>X</td>\n",
       "      <td>156006493</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>155236158</td>\n",
       "      <td>rs1417497789</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 1.776, 'raw': -0.063666}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0011848341...</td>\n",
       "      <td>8797.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "      <td>8797.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72878706</th>\n",
       "      <td>X</td>\n",
       "      <td>156008488</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>155238153</td>\n",
       "      <td>rs186430584</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 3.876, 'raw': 0.076619}</td>\n",
       "      <td>{'gnomad_afr': 0.030184331797235023, 'gnomad_a...</td>\n",
       "      <td>10792.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "      <td>10792.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72878707</th>\n",
       "      <td>X</td>\n",
       "      <td>156009035</td>\n",
       "      <td>CTG</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>155238700</td>\n",
       "      <td>rs1381721936</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': None, 'raw': None}</td>\n",
       "      <td>{'gnomad_afr': 0.0008291873963515755, 'gnomad_...</td>\n",
       "      <td>11339.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "      <td>11339.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72878708</th>\n",
       "      <td>X</td>\n",
       "      <td>156027432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>155257097</td>\n",
       "      <td>rs1041001787</td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>{'phred': 9.819, 'raw': 0.548259}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>6606.0</td>\n",
       "      <td>ENSG00000182484</td>\n",
       "      <td>6606.0</td>\n",
       "      <td>ENSG00000182484</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72878709 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0             1      30794          G          T          1         30794   \n",
       "1             1      55330          G          A          1         55330   \n",
       "2             1      57248          C          T          1         57248   \n",
       "3             1      66381          T          A          1         66381   \n",
       "4             1      67631          G          C          1         67631   \n",
       "...         ...        ...        ...        ...        ...           ...   \n",
       "72878704      X  156006466          G          C          X     155236131   \n",
       "72878705      X  156006493          C          T          X     155236158   \n",
       "72878706      X  156008488          C          T          X     155238153   \n",
       "72878707      X  156009035        CTG          C          X     155238700   \n",
       "72878708      X  156027432          C          T          X     155257097   \n",
       "\n",
       "                 rs_id             most_severe_consequence  \\\n",
       "0         rs1311411458                      intron_variant   \n",
       "1          rs185215913             downstream_gene_variant   \n",
       "2         rs1249676487             downstream_gene_variant   \n",
       "3          rs538833530               upstream_gene_variant   \n",
       "4          rs533896527               upstream_gene_variant   \n",
       "...                ...                                 ...   \n",
       "72878704  rs1218662062                      intron_variant   \n",
       "72878705  rs1417497789                      intron_variant   \n",
       "72878706   rs186430584                      intron_variant   \n",
       "72878707  rs1381721936                      intron_variant   \n",
       "72878708  rs1041001787  non_coding_transcript_exon_variant   \n",
       "\n",
       "                                        cadd  \\\n",
       "0         {'phred': 0.177, 'raw': -0.402967}   \n",
       "1         {'phred': 0.431, 'raw': -0.275244}   \n",
       "2          {'phred': 7.268, 'raw': 0.304146}   \n",
       "3         {'phred': 1.389, 'raw': -0.101295}   \n",
       "4         {'phred': 0.033, 'raw': -0.632235}   \n",
       "...                                      ...   \n",
       "72878704  {'phred': 1.118, 'raw': -0.133403}   \n",
       "72878705  {'phred': 1.776, 'raw': -0.063666}   \n",
       "72878706   {'phred': 3.876, 'raw': 0.076619}   \n",
       "72878707        {'phred': None, 'raw': None}   \n",
       "72878708   {'phred': 9.819, 'raw': 0.548259}   \n",
       "\n",
       "                                                         af  \\\n",
       "0         {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...   \n",
       "1         {'gnomad_afr': 0.016711590296495958, 'gnomad_a...   \n",
       "2         {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...   \n",
       "3         {'gnomad_afr': 0.15321011673151752, 'gnomad_am...   \n",
       "4         {'gnomad_afr': 0.041812315138478086, 'gnomad_a...   \n",
       "...                                                     ...   \n",
       "72878704  {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...   \n",
       "72878705  {'gnomad_afr': 0.0, 'gnomad_amr': 0.0011848341...   \n",
       "72878706  {'gnomad_afr': 0.030184331797235023, 'gnomad_a...   \n",
       "72878707  {'gnomad_afr': 0.0008291873963515755, 'gnomad_...   \n",
       "72878708  {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...   \n",
       "\n",
       "          gene_id_any_distance      gene_id_any  gene_id_prot_coding_distance  \\\n",
       "0                      34625.0  ENSG00000186092                       34625.0   \n",
       "1                      10089.0  ENSG00000186092                       10089.0   \n",
       "2                       8171.0  ENSG00000186092                        8171.0   \n",
       "3                        962.0  ENSG00000186092                         962.0   \n",
       "4                       2212.0  ENSG00000186092                        2212.0   \n",
       "...                        ...              ...                           ...   \n",
       "72878704                8770.0  ENSG00000124334                        8770.0   \n",
       "72878705                8797.0  ENSG00000124334                        8797.0   \n",
       "72878706               10792.0  ENSG00000124334                       10792.0   \n",
       "72878707               11339.0  ENSG00000124334                       11339.0   \n",
       "72878708                6606.0  ENSG00000182484                        6606.0   \n",
       "\n",
       "         gene_id_prot_coding  \n",
       "0            ENSG00000186092  \n",
       "1            ENSG00000186092  \n",
       "2            ENSG00000186092  \n",
       "3            ENSG00000186092  \n",
       "4            ENSG00000186092  \n",
       "...                      ...  \n",
       "72878704     ENSG00000124334  \n",
       "72878705     ENSG00000124334  \n",
       "72878706     ENSG00000124334  \n",
       "72878707     ENSG00000124334  \n",
       "72878708     ENSG00000182484  \n",
       "\n",
       "[72878709 rows x 14 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_index_parquet = data_dir + \"/complete_variant_index.parquet\"\n",
    "variant_index = pd.read_parquet(variant_index_parquet) # takes a few minutes\n",
    "variant_index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge with UKB SNP data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>gene_id_any_distance</th>\n",
       "      <th>gene_id_any</th>\n",
       "      <th>gene_id_prot_coding_distance</th>\n",
       "      <th>gene_id_prot_coding</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1169204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1104584</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.61, 'raw': -0.223786}</td>\n",
       "      <td>{'gnomad_afr': 0.1370131366674349, 'gnomad_amr...</td>\n",
       "      <td>4676.0</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>4676.0</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>1</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>0</td>\n",
       "      <td>1104584</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>12098.0</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>12098.0</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>1</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>0</td>\n",
       "      <td>1969752</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2194104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>2125543</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 4.241, 'raw': 0.098415}</td>\n",
       "      <td>{'gnomad_afr': 0.00034498620055197794, 'gnomad...</td>\n",
       "      <td>18616.0</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>18616.0</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>0</td>\n",
       "      <td>2125543</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2592668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>2524107</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 23.9, 'raw': 3.118097}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>6177.0</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>6177.0</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>0</td>\n",
       "      <td>2524107</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2828773</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2745338</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.058, 'raw': -0.555543}</td>\n",
       "      <td>{'gnomad_afr': 0.0060821666284140465, 'gnomad_...</td>\n",
       "      <td>27080.0</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>27080.0</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>0</td>\n",
       "      <td>2745338</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747754</th>\n",
       "      <td>X</td>\n",
       "      <td>152014563</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>151183035</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 2.456, 'raw': -0.011208}</td>\n",
       "      <td>{'gnomad_afr': 0.009298393913778529, 'gnomad_a...</td>\n",
       "      <td>39883.0</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>39883.0</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>X</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>0</td>\n",
       "      <td>151183035</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747755</th>\n",
       "      <td>X</td>\n",
       "      <td>153459591</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>152725049</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 2.968, 'raw': 0.022059}</td>\n",
       "      <td>{'gnomad_afr': 0.0032312925170068026, 'gnomad_...</td>\n",
       "      <td>10996.0</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>10996.0</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>0</td>\n",
       "      <td>152725049</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747756</th>\n",
       "      <td>X</td>\n",
       "      <td>153781540</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>153046995</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 22.5, 'raw': 2.480113}</td>\n",
       "      <td>{'gnomad_afr': 0.00017018379850238256, 'gnomad...</td>\n",
       "      <td>5128.0</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>5128.0</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>X</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>0</td>\n",
       "      <td>153046995</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747757</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>6780.0</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>6780.0</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>0</td>\n",
       "      <td>153762634</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747758</th>\n",
       "      <td>X</td>\n",
       "      <td>155780194</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>X</td>\n",
       "      <td>155009856</td>\n",
       "      <td>rs700449</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>{'phred': 5.474, 'raw': 0.175592}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0011876484...</td>\n",
       "      <td>61096.0</td>\n",
       "      <td>ENSG00000185973</td>\n",
       "      <td>61096.0</td>\n",
       "      <td>ENSG00000185973</td>\n",
       "      <td>XY</td>\n",
       "      <td>rs700449</td>\n",
       "      <td>0</td>\n",
       "      <td>155009856</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>747759 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0           1    1169204          T          C          1       1104584   \n",
       "1           1    2038313          C          T          1       1969752   \n",
       "2           1    2194104          A          G          1       2125543   \n",
       "3           1    2592668          C          T          1       2524107   \n",
       "4           1    2828773          C          A          1       2745338   \n",
       "...       ...        ...        ...        ...        ...           ...   \n",
       "747754      X  152014563          G          A          X     151183035   \n",
       "747755      X  153459591          C          T          X     152725049   \n",
       "747756      X  153781540          A          C          X     153046995   \n",
       "747757      X  154534419          G          A          X     153762634   \n",
       "747758      X  155780194          T          G          X     155009856   \n",
       "\n",
       "              rs_id most_severe_consequence  \\\n",
       "0       rs112695918   upstream_gene_variant   \n",
       "1        rs74372554      intergenic_variant   \n",
       "2       rs146607591        missense_variant   \n",
       "3       rs146100312        missense_variant   \n",
       "4       rs116076619      intergenic_variant   \n",
       "...             ...                     ...   \n",
       "747754  rs145754054      intergenic_variant   \n",
       "747755    rs5945389          intron_variant   \n",
       "747756  rs199915647        missense_variant   \n",
       "747757    rs5030868        missense_variant   \n",
       "747758     rs700449     3_prime_UTR_variant   \n",
       "\n",
       "                                      cadd  \\\n",
       "0        {'phred': 0.61, 'raw': -0.223786}   \n",
       "1       {'phred': 0.309, 'raw': -0.324059}   \n",
       "2        {'phred': 4.241, 'raw': 0.098415}   \n",
       "3         {'phred': 23.9, 'raw': 3.118097}   \n",
       "4       {'phred': 0.058, 'raw': -0.555543}   \n",
       "...                                    ...   \n",
       "747754  {'phred': 2.456, 'raw': -0.011208}   \n",
       "747755   {'phred': 2.968, 'raw': 0.022059}   \n",
       "747756    {'phred': 22.5, 'raw': 2.480113}   \n",
       "747757    {'phred': 24.7, 'raw': 3.374046}   \n",
       "747758   {'phred': 5.474, 'raw': 0.175592}   \n",
       "\n",
       "                                                       af  \\\n",
       "0       {'gnomad_afr': 0.1370131366674349, 'gnomad_amr...   \n",
       "1       {'gnomad_afr': 0.01078971533516988, 'gnomad_am...   \n",
       "2       {'gnomad_afr': 0.00034498620055197794, 'gnomad...   \n",
       "3       {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...   \n",
       "4       {'gnomad_afr': 0.0060821666284140465, 'gnomad_...   \n",
       "...                                                   ...   \n",
       "747754  {'gnomad_afr': 0.009298393913778529, 'gnomad_a...   \n",
       "747755  {'gnomad_afr': 0.0032312925170068026, 'gnomad_...   \n",
       "747756  {'gnomad_afr': 0.00017018379850238256, 'gnomad...   \n",
       "747757  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...   \n",
       "747758  {'gnomad_afr': 0.0, 'gnomad_amr': 0.0011876484...   \n",
       "\n",
       "        gene_id_any_distance      gene_id_any  gene_id_prot_coding_distance  \\\n",
       "0                     4676.0  ENSG00000162571                        4676.0   \n",
       "1                    12098.0  ENSG00000067606                       12098.0   \n",
       "2                    18616.0  ENSG00000162585                       18616.0   \n",
       "3                     6177.0  ENSG00000157870                        6177.0   \n",
       "4                    27080.0  ENSG00000215912                       27080.0   \n",
       "...                      ...              ...                           ...   \n",
       "747754               39883.0  ENSG00000102287                       39883.0   \n",
       "747755               10996.0  ENSG00000183479                       10996.0   \n",
       "747756                5128.0  ENSG00000184343                        5128.0   \n",
       "747757                6780.0  ENSG00000269335                        6780.0   \n",
       "747758               61096.0  ENSG00000185973                       61096.0   \n",
       "\n",
       "       gene_id_prot_coding chrom          snp  cm        pos a1 a2  \n",
       "0          ENSG00000162571     1  rs112695918   0    1104584  T  C  \n",
       "1          ENSG00000067606     1   rs74372554   0    1969752  C  T  \n",
       "2          ENSG00000162585     1  rs146607591   0    2125543  A  G  \n",
       "3          ENSG00000157870     1  rs146100312   0    2524107  C  T  \n",
       "4          ENSG00000215912     1  rs116076619   0    2745338  C  A  \n",
       "...                    ...   ...          ...  ..        ... .. ..  \n",
       "747754     ENSG00000102287     X  rs145754054   0  151183035  G  A  \n",
       "747755     ENSG00000183479     X    rs5945389   0  152725049  C  T  \n",
       "747756     ENSG00000184343     X  rs199915647   0  153046995  A  C  \n",
       "747757     ENSG00000269335     X    rs5030868   0  153762634  G  A  \n",
       "747758     ENSG00000185973    XY     rs700449   0  155009856  T  C  \n",
       "\n",
       "[747759 rows x 20 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge bim with variant index\n",
    "ot_bim = variant_index.merge(bim_df, left_on='rs_id', right_on='snp', how='inner')\n",
    "ot_bim"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene ids to gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_data = EnsemblRelease(species='human', release=75)\n",
    "def get_gene_name(ensembl_id):\n",
    "    try:\n",
    "        return human_data.gene_name_of_gene_id(ensembl_id)\n",
    "    except:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Pandas Apply: 100%|██████████| 18817/18817 [10:42<00:00, 29.27it/s]\n"
     ]
    }
   ],
   "source": [
    "name_map = pd.DataFrame()\n",
    "name_map[\"gene_id\"] = ot_bim.gene_id_prot_coding.unique()\n",
    "name_map[\"gene_name\"]  = name_map[\"gene_id\"].swifter.apply(lambda x: get_gene_name(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene Id not mappable:  298\n",
      "Unique gene names:  18510\n"
     ]
    }
   ],
   "source": [
    "# count None in gene names\n",
    "print(\"Gene Id not mappable: \", name_map.gene_name.isnull().sum())\n",
    "print(\"Unique gene names: \", name_map.gene_name.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "ot_bim = pd.merge(ot_bim, name_map, left_on='gene_id_prot_coding', right_on='gene_id', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_id_prot_coding_distance</th>\n",
       "      <th>gene_id_prot_coding</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1169204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1104584</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.61, 'raw': -0.223786}</td>\n",
       "      <td>{'gnomad_afr': 0.1370131366674349, 'gnomad_amr...</td>\n",
       "      <td>...</td>\n",
       "      <td>4676.0</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>1</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>0</td>\n",
       "      <td>1104584</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>TTLL10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>12098.0</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>1</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>0</td>\n",
       "      <td>1969752</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2194104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>2125543</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 4.241, 'raw': 0.098415}</td>\n",
       "      <td>{'gnomad_afr': 0.00034498620055197794, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>18616.0</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>0</td>\n",
       "      <td>2125543</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>C1orf86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2592668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>2524107</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 23.9, 'raw': 3.118097}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>6177.0</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>0</td>\n",
       "      <td>2524107</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>FAM213B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2828773</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2745338</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.058, 'raw': -0.555543}</td>\n",
       "      <td>{'gnomad_afr': 0.0060821666284140465, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>27080.0</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>0</td>\n",
       "      <td>2745338</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>TTC34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747754</th>\n",
       "      <td>X</td>\n",
       "      <td>152014563</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>151183035</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 2.456, 'raw': -0.011208}</td>\n",
       "      <td>{'gnomad_afr': 0.009298393913778529, 'gnomad_a...</td>\n",
       "      <td>...</td>\n",
       "      <td>39883.0</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>X</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>0</td>\n",
       "      <td>151183035</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>GABRE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747755</th>\n",
       "      <td>X</td>\n",
       "      <td>153459591</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>152725049</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 2.968, 'raw': 0.022059}</td>\n",
       "      <td>{'gnomad_afr': 0.0032312925170068026, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>10996.0</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>0</td>\n",
       "      <td>152725049</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>TREX2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747756</th>\n",
       "      <td>X</td>\n",
       "      <td>153781540</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>153046995</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 22.5, 'raw': 2.480113}</td>\n",
       "      <td>{'gnomad_afr': 0.00017018379850238256, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>5128.0</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>X</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>0</td>\n",
       "      <td>153046995</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>SRPK3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747757</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>6780.0</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>0</td>\n",
       "      <td>153762634</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747758</th>\n",
       "      <td>X</td>\n",
       "      <td>155780194</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>X</td>\n",
       "      <td>155009856</td>\n",
       "      <td>rs700449</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>{'phred': 5.474, 'raw': 0.175592}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0011876484...</td>\n",
       "      <td>...</td>\n",
       "      <td>61096.0</td>\n",
       "      <td>ENSG00000185973</td>\n",
       "      <td>XY</td>\n",
       "      <td>rs700449</td>\n",
       "      <td>0</td>\n",
       "      <td>155009856</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000185973</td>\n",
       "      <td>TMLHE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>741340 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0           1    1169204          T          C          1       1104584   \n",
       "1           1    2038313          C          T          1       1969752   \n",
       "2           1    2194104          A          G          1       2125543   \n",
       "3           1    2592668          C          T          1       2524107   \n",
       "4           1    2828773          C          A          1       2745338   \n",
       "...       ...        ...        ...        ...        ...           ...   \n",
       "747754      X  152014563          G          A          X     151183035   \n",
       "747755      X  153459591          C          T          X     152725049   \n",
       "747756      X  153781540          A          C          X     153046995   \n",
       "747757      X  154534419          G          A          X     153762634   \n",
       "747758      X  155780194          T          G          X     155009856   \n",
       "\n",
       "              rs_id most_severe_consequence  \\\n",
       "0       rs112695918   upstream_gene_variant   \n",
       "1        rs74372554      intergenic_variant   \n",
       "2       rs146607591        missense_variant   \n",
       "3       rs146100312        missense_variant   \n",
       "4       rs116076619      intergenic_variant   \n",
       "...             ...                     ...   \n",
       "747754  rs145754054      intergenic_variant   \n",
       "747755    rs5945389          intron_variant   \n",
       "747756  rs199915647        missense_variant   \n",
       "747757    rs5030868        missense_variant   \n",
       "747758     rs700449     3_prime_UTR_variant   \n",
       "\n",
       "                                      cadd  \\\n",
       "0        {'phred': 0.61, 'raw': -0.223786}   \n",
       "1       {'phred': 0.309, 'raw': -0.324059}   \n",
       "2        {'phred': 4.241, 'raw': 0.098415}   \n",
       "3         {'phred': 23.9, 'raw': 3.118097}   \n",
       "4       {'phred': 0.058, 'raw': -0.555543}   \n",
       "...                                    ...   \n",
       "747754  {'phred': 2.456, 'raw': -0.011208}   \n",
       "747755   {'phred': 2.968, 'raw': 0.022059}   \n",
       "747756    {'phred': 22.5, 'raw': 2.480113}   \n",
       "747757    {'phred': 24.7, 'raw': 3.374046}   \n",
       "747758   {'phred': 5.474, 'raw': 0.175592}   \n",
       "\n",
       "                                                       af  ...  \\\n",
       "0       {'gnomad_afr': 0.1370131366674349, 'gnomad_amr...  ...   \n",
       "1       {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   \n",
       "2       {'gnomad_afr': 0.00034498620055197794, 'gnomad...  ...   \n",
       "3       {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...  ...   \n",
       "4       {'gnomad_afr': 0.0060821666284140465, 'gnomad_...  ...   \n",
       "...                                                   ...  ...   \n",
       "747754  {'gnomad_afr': 0.009298393913778529, 'gnomad_a...  ...   \n",
       "747755  {'gnomad_afr': 0.0032312925170068026, 'gnomad_...  ...   \n",
       "747756  {'gnomad_afr': 0.00017018379850238256, 'gnomad...  ...   \n",
       "747757  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   \n",
       "747758  {'gnomad_afr': 0.0, 'gnomad_amr': 0.0011876484...  ...   \n",
       "\n",
       "        gene_id_prot_coding_distance gene_id_prot_coding  chrom          snp  \\\n",
       "0                             4676.0     ENSG00000162571      1  rs112695918   \n",
       "1                            12098.0     ENSG00000067606      1   rs74372554   \n",
       "2                            18616.0     ENSG00000162585      1  rs146607591   \n",
       "3                             6177.0     ENSG00000157870      1  rs146100312   \n",
       "4                            27080.0     ENSG00000215912      1  rs116076619   \n",
       "...                              ...                 ...    ...          ...   \n",
       "747754                       39883.0     ENSG00000102287      X  rs145754054   \n",
       "747755                       10996.0     ENSG00000183479      X    rs5945389   \n",
       "747756                        5128.0     ENSG00000184343      X  rs199915647   \n",
       "747757                        6780.0     ENSG00000269335      X    rs5030868   \n",
       "747758                       61096.0     ENSG00000185973     XY     rs700449   \n",
       "\n",
       "       cm        pos  a1  a2          gene_id gene_name  \n",
       "0       0    1104584   T   C  ENSG00000162571    TTLL10  \n",
       "1       0    1969752   C   T  ENSG00000067606     PRKCZ  \n",
       "2       0    2125543   A   G  ENSG00000162585   C1orf86  \n",
       "3       0    2524107   C   T  ENSG00000157870   FAM213B  \n",
       "4       0    2745338   C   A  ENSG00000215912     TTC34  \n",
       "...    ..        ...  ..  ..              ...       ...  \n",
       "747754  0  151183035   G   A  ENSG00000102287     GABRE  \n",
       "747755  0  152725049   C   T  ENSG00000183479     TREX2  \n",
       "747756  0  153046995   A   C  ENSG00000184343     SRPK3  \n",
       "747757  0  153762634   G   A  ENSG00000269335     IKBKG  \n",
       "747758  0  155009856   T   C  ENSG00000185973     TMLHE  \n",
       "\n",
       "[741340 rows x 22 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove all None gene names\n",
    "ot_bim = ot_bim[ot_bim.gene_name.notnull()]\n",
    "ot_bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "731084"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ot_bim.rs_id.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_id_prot_coding_distance</th>\n",
       "      <th>gene_id_prot_coding</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1169204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1104584</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.61, 'raw': -0.223786}</td>\n",
       "      <td>{'gnomad_afr': 0.1370131366674349, 'gnomad_amr...</td>\n",
       "      <td>...</td>\n",
       "      <td>4676.0</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>1</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>0</td>\n",
       "      <td>1104584</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>TTLL10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>12098.0</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>1</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>0</td>\n",
       "      <td>1969752</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2194104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>2125543</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 4.241, 'raw': 0.098415}</td>\n",
       "      <td>{'gnomad_afr': 0.00034498620055197794, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>18616.0</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>0</td>\n",
       "      <td>2125543</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>C1orf86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2592668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>2524107</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 23.9, 'raw': 3.118097}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>6177.0</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>0</td>\n",
       "      <td>2524107</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>FAM213B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2828773</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2745338</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.058, 'raw': -0.555543}</td>\n",
       "      <td>{'gnomad_afr': 0.0060821666284140465, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>27080.0</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>0</td>\n",
       "      <td>2745338</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>TTC34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747753</th>\n",
       "      <td>X</td>\n",
       "      <td>151609900</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>X</td>\n",
       "      <td>150778372</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 3.523, 'raw': 0.055668}</td>\n",
       "      <td>{'gnomad_afr': 0.11482740118046861, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>46225.0</td>\n",
       "      <td>ENSG00000166049</td>\n",
       "      <td>X</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>0</td>\n",
       "      <td>150778372</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000166049</td>\n",
       "      <td>PASD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747754</th>\n",
       "      <td>X</td>\n",
       "      <td>152014563</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>151183035</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 2.456, 'raw': -0.011208}</td>\n",
       "      <td>{'gnomad_afr': 0.009298393913778529, 'gnomad_a...</td>\n",
       "      <td>...</td>\n",
       "      <td>39883.0</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>X</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>0</td>\n",
       "      <td>151183035</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>GABRE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747755</th>\n",
       "      <td>X</td>\n",
       "      <td>153459591</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>152725049</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 2.968, 'raw': 0.022059}</td>\n",
       "      <td>{'gnomad_afr': 0.0032312925170068026, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>10996.0</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>0</td>\n",
       "      <td>152725049</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>TREX2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747756</th>\n",
       "      <td>X</td>\n",
       "      <td>153781540</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>153046995</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 22.5, 'raw': 2.480113}</td>\n",
       "      <td>{'gnomad_afr': 0.00017018379850238256, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>5128.0</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>X</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>0</td>\n",
       "      <td>153046995</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>SRPK3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747757</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>6780.0</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>0</td>\n",
       "      <td>153762634</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>731084 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0           1    1169204          T          C          1       1104584   \n",
       "1           1    2038313          C          T          1       1969752   \n",
       "2           1    2194104          A          G          1       2125543   \n",
       "3           1    2592668          C          T          1       2524107   \n",
       "4           1    2828773          C          A          1       2745338   \n",
       "...       ...        ...        ...        ...        ...           ...   \n",
       "747753      X  151609900          A          G          X     150778372   \n",
       "747754      X  152014563          G          A          X     151183035   \n",
       "747755      X  153459591          C          T          X     152725049   \n",
       "747756      X  153781540          A          C          X     153046995   \n",
       "747757      X  154534419          G          A          X     153762634   \n",
       "\n",
       "              rs_id most_severe_consequence  \\\n",
       "0       rs112695918   upstream_gene_variant   \n",
       "1        rs74372554      intergenic_variant   \n",
       "2       rs146607591        missense_variant   \n",
       "3       rs146100312        missense_variant   \n",
       "4       rs116076619      intergenic_variant   \n",
       "...             ...                     ...   \n",
       "747753   rs58448897          intron_variant   \n",
       "747754  rs145754054      intergenic_variant   \n",
       "747755    rs5945389          intron_variant   \n",
       "747756  rs199915647        missense_variant   \n",
       "747757    rs5030868        missense_variant   \n",
       "\n",
       "                                      cadd  \\\n",
       "0        {'phred': 0.61, 'raw': -0.223786}   \n",
       "1       {'phred': 0.309, 'raw': -0.324059}   \n",
       "2        {'phred': 4.241, 'raw': 0.098415}   \n",
       "3         {'phred': 23.9, 'raw': 3.118097}   \n",
       "4       {'phred': 0.058, 'raw': -0.555543}   \n",
       "...                                    ...   \n",
       "747753   {'phred': 3.523, 'raw': 0.055668}   \n",
       "747754  {'phred': 2.456, 'raw': -0.011208}   \n",
       "747755   {'phred': 2.968, 'raw': 0.022059}   \n",
       "747756    {'phred': 22.5, 'raw': 2.480113}   \n",
       "747757    {'phred': 24.7, 'raw': 3.374046}   \n",
       "\n",
       "                                                       af  ...  \\\n",
       "0       {'gnomad_afr': 0.1370131366674349, 'gnomad_amr...  ...   \n",
       "1       {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   \n",
       "2       {'gnomad_afr': 0.00034498620055197794, 'gnomad...  ...   \n",
       "3       {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...  ...   \n",
       "4       {'gnomad_afr': 0.0060821666284140465, 'gnomad_...  ...   \n",
       "...                                                   ...  ...   \n",
       "747753  {'gnomad_afr': 0.11482740118046861, 'gnomad_am...  ...   \n",
       "747754  {'gnomad_afr': 0.009298393913778529, 'gnomad_a...  ...   \n",
       "747755  {'gnomad_afr': 0.0032312925170068026, 'gnomad_...  ...   \n",
       "747756  {'gnomad_afr': 0.00017018379850238256, 'gnomad...  ...   \n",
       "747757  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   \n",
       "\n",
       "        gene_id_prot_coding_distance gene_id_prot_coding  chrom          snp  \\\n",
       "0                             4676.0     ENSG00000162571      1  rs112695918   \n",
       "1                            12098.0     ENSG00000067606      1   rs74372554   \n",
       "2                            18616.0     ENSG00000162585      1  rs146607591   \n",
       "3                             6177.0     ENSG00000157870      1  rs146100312   \n",
       "4                            27080.0     ENSG00000215912      1  rs116076619   \n",
       "...                              ...                 ...    ...          ...   \n",
       "747753                       46225.0     ENSG00000166049      X   rs58448897   \n",
       "747754                       39883.0     ENSG00000102287      X  rs145754054   \n",
       "747755                       10996.0     ENSG00000183479      X    rs5945389   \n",
       "747756                        5128.0     ENSG00000184343      X  rs199915647   \n",
       "747757                        6780.0     ENSG00000269335      X    rs5030868   \n",
       "\n",
       "       cm        pos  a1  a2          gene_id gene_name  \n",
       "0       0    1104584   T   C  ENSG00000162571    TTLL10  \n",
       "1       0    1969752   C   T  ENSG00000067606     PRKCZ  \n",
       "2       0    2125543   A   G  ENSG00000162585   C1orf86  \n",
       "3       0    2524107   C   T  ENSG00000157870   FAM213B  \n",
       "4       0    2745338   C   A  ENSG00000215912     TTC34  \n",
       "...    ..        ...  ..  ..              ...       ...  \n",
       "747753  0  150778372   A   G  ENSG00000166049     PASD1  \n",
       "747754  0  151183035   G   A  ENSG00000102287     GABRE  \n",
       "747755  0  152725049   C   T  ENSG00000183479     TREX2  \n",
       "747756  0  153046995   A   C  ENSG00000184343     SRPK3  \n",
       "747757  0  153762634   G   A  ENSG00000269335     IKBKG  \n",
       "\n",
       "[731084 rows x 22 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ot_bim = ot_bim.drop_duplicates(subset=[\"rs_id\", \"gene_name\"])\n",
    "# rename rs_id col to snp\n",
    "ot_bim = ot_bim.rename({\"rs_id\":\"snp\"})\n",
    "ot_bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_bed_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>maf</th>\n",
       "      <th>nchrobs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>rs28659788</td>\n",
       "      <td>0</td>\n",
       "      <td>723307</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116587930</td>\n",
       "      <td>0</td>\n",
       "      <td>727841</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116720794</td>\n",
       "      <td>0</td>\n",
       "      <td>729632</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>rs3131972</td>\n",
       "      <td>0</td>\n",
       "      <td>752721</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>rs12184325</td>\n",
       "      <td>0</td>\n",
       "      <td>754105</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805421</th>\n",
       "      <td>805421</td>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-92047842</td>\n",
       "      <td>0</td>\n",
       "      <td>16337</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805422</th>\n",
       "      <td>805422</td>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-79443531</td>\n",
       "      <td>0</td>\n",
       "      <td>16356</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805423</th>\n",
       "      <td>805423</td>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-79443532</td>\n",
       "      <td>0</td>\n",
       "      <td>16362</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805424</th>\n",
       "      <td>805424</td>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-89025709</td>\n",
       "      <td>0</td>\n",
       "      <td>16390</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805425</th>\n",
       "      <td>805425</td>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-79381726</td>\n",
       "      <td>0</td>\n",
       "      <td>16391</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>805426 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        snp_bed_id chrom            snp  a1      a2 maf nchrobs\n",
       "0                0     1     rs28659788   0  723307   C       G\n",
       "1                1     1    rs116587930   0  727841   G       A\n",
       "2                2     1    rs116720794   0  729632   C       T\n",
       "3                3     1      rs3131972   0  752721   A       G\n",
       "4                4     1     rs12184325   0  754105   C       T\n",
       "...            ...   ...            ...  ..     ...  ..     ...\n",
       "805421      805421    MT  Affx-92047842   0   16337   C       T\n",
       "805422      805422    MT  Affx-79443531   0   16356   T       C\n",
       "805423      805423    MT  Affx-79443532   0   16362   T       C\n",
       "805424      805424    MT  Affx-89025709   0   16390   G       A\n",
       "805425      805425    MT  Affx-79381726   0   16391   G       A\n",
       "\n",
       "[805426 rows x 7 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get snp_bed_id from bim_df \n",
    "bim_df = bim_df.reset_index()\n",
    "#bim_df.columns = ['snp_bed_id', 'chrom', 'snp', 'cm', 'pos', 'a1', 'a2']\n",
    "bim_df.columns = ['snp_bed_id', 'chrom', 'snp', 'cm', 'pos', 'a1', 'a2']\n",
    "bim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "ot_bim = ot_bim.merge(bim_df[[\"snp\", \"snp_bed_id\"]], left_on=\"snp\", right_on=\"snp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_id_prot_coding</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>snp_bed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1169204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1104584</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.61, 'raw': -0.223786}</td>\n",
       "      <td>{'gnomad_afr': 0.1370131366674349, 'gnomad_amr...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>1</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>0</td>\n",
       "      <td>1104584</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>TTLL10</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>1</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>0</td>\n",
       "      <td>1969752</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2194104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>2125543</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 4.241, 'raw': 0.098415}</td>\n",
       "      <td>{'gnomad_afr': 0.00034498620055197794, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>0</td>\n",
       "      <td>2125543</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>C1orf86</td>\n",
       "      <td>648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2592668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>2524107</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 23.9, 'raw': 3.118097}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>0</td>\n",
       "      <td>2524107</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>FAM213B</td>\n",
       "      <td>838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2828773</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2745338</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.058, 'raw': -0.555543}</td>\n",
       "      <td>{'gnomad_afr': 0.0060821666284140465, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>0</td>\n",
       "      <td>2745338</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>TTC34</td>\n",
       "      <td>889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731079</th>\n",
       "      <td>X</td>\n",
       "      <td>151609900</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>X</td>\n",
       "      <td>150778372</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 3.523, 'raw': 0.055668}</td>\n",
       "      <td>{'gnomad_afr': 0.11482740118046861, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000166049</td>\n",
       "      <td>X</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>0</td>\n",
       "      <td>150778372</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000166049</td>\n",
       "      <td>PASD1</td>\n",
       "      <td>802181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731080</th>\n",
       "      <td>X</td>\n",
       "      <td>152014563</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>151183035</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 2.456, 'raw': -0.011208}</td>\n",
       "      <td>{'gnomad_afr': 0.009298393913778529, 'gnomad_a...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>X</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>0</td>\n",
       "      <td>151183035</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>GABRE</td>\n",
       "      <td>802336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731081</th>\n",
       "      <td>X</td>\n",
       "      <td>153459591</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>152725049</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 2.968, 'raw': 0.022059}</td>\n",
       "      <td>{'gnomad_afr': 0.0032312925170068026, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>0</td>\n",
       "      <td>152725049</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>TREX2</td>\n",
       "      <td>802547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731082</th>\n",
       "      <td>X</td>\n",
       "      <td>153781540</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>153046995</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 22.5, 'raw': 2.480113}</td>\n",
       "      <td>{'gnomad_afr': 0.00017018379850238256, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>X</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>0</td>\n",
       "      <td>153046995</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>SRPK3</td>\n",
       "      <td>802648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731083</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>0</td>\n",
       "      <td>153762634</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>731084 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0           1    1169204          T          C          1       1104584   \n",
       "1           1    2038313          C          T          1       1969752   \n",
       "2           1    2194104          A          G          1       2125543   \n",
       "3           1    2592668          C          T          1       2524107   \n",
       "4           1    2828773          C          A          1       2745338   \n",
       "...       ...        ...        ...        ...        ...           ...   \n",
       "731079      X  151609900          A          G          X     150778372   \n",
       "731080      X  152014563          G          A          X     151183035   \n",
       "731081      X  153459591          C          T          X     152725049   \n",
       "731082      X  153781540          A          C          X     153046995   \n",
       "731083      X  154534419          G          A          X     153762634   \n",
       "\n",
       "              rs_id most_severe_consequence  \\\n",
       "0       rs112695918   upstream_gene_variant   \n",
       "1        rs74372554      intergenic_variant   \n",
       "2       rs146607591        missense_variant   \n",
       "3       rs146100312        missense_variant   \n",
       "4       rs116076619      intergenic_variant   \n",
       "...             ...                     ...   \n",
       "731079   rs58448897          intron_variant   \n",
       "731080  rs145754054      intergenic_variant   \n",
       "731081    rs5945389          intron_variant   \n",
       "731082  rs199915647        missense_variant   \n",
       "731083    rs5030868        missense_variant   \n",
       "\n",
       "                                      cadd  \\\n",
       "0        {'phred': 0.61, 'raw': -0.223786}   \n",
       "1       {'phred': 0.309, 'raw': -0.324059}   \n",
       "2        {'phred': 4.241, 'raw': 0.098415}   \n",
       "3         {'phred': 23.9, 'raw': 3.118097}   \n",
       "4       {'phred': 0.058, 'raw': -0.555543}   \n",
       "...                                    ...   \n",
       "731079   {'phred': 3.523, 'raw': 0.055668}   \n",
       "731080  {'phred': 2.456, 'raw': -0.011208}   \n",
       "731081   {'phred': 2.968, 'raw': 0.022059}   \n",
       "731082    {'phred': 22.5, 'raw': 2.480113}   \n",
       "731083    {'phred': 24.7, 'raw': 3.374046}   \n",
       "\n",
       "                                                       af  ...  \\\n",
       "0       {'gnomad_afr': 0.1370131366674349, 'gnomad_amr...  ...   \n",
       "1       {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   \n",
       "2       {'gnomad_afr': 0.00034498620055197794, 'gnomad...  ...   \n",
       "3       {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...  ...   \n",
       "4       {'gnomad_afr': 0.0060821666284140465, 'gnomad_...  ...   \n",
       "...                                                   ...  ...   \n",
       "731079  {'gnomad_afr': 0.11482740118046861, 'gnomad_am...  ...   \n",
       "731080  {'gnomad_afr': 0.009298393913778529, 'gnomad_a...  ...   \n",
       "731081  {'gnomad_afr': 0.0032312925170068026, 'gnomad_...  ...   \n",
       "731082  {'gnomad_afr': 0.00017018379850238256, 'gnomad...  ...   \n",
       "731083  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   \n",
       "\n",
       "        gene_id_prot_coding chrom          snp cm        pos a1  a2  \\\n",
       "0           ENSG00000162571     1  rs112695918  0    1104584  T   C   \n",
       "1           ENSG00000067606     1   rs74372554  0    1969752  C   T   \n",
       "2           ENSG00000162585     1  rs146607591  0    2125543  A   G   \n",
       "3           ENSG00000157870     1  rs146100312  0    2524107  C   T   \n",
       "4           ENSG00000215912     1  rs116076619  0    2745338  C   A   \n",
       "...                     ...   ...          ... ..        ... ..  ..   \n",
       "731079      ENSG00000166049     X   rs58448897  0  150778372  A   G   \n",
       "731080      ENSG00000102287     X  rs145754054  0  151183035  G   A   \n",
       "731081      ENSG00000183479     X    rs5945389  0  152725049  C   T   \n",
       "731082      ENSG00000184343     X  rs199915647  0  153046995  A   C   \n",
       "731083      ENSG00000269335     X    rs5030868  0  153762634  G   A   \n",
       "\n",
       "                gene_id gene_name snp_bed_id  \n",
       "0       ENSG00000162571    TTLL10        220  \n",
       "1       ENSG00000067606     PRKCZ        599  \n",
       "2       ENSG00000162585   C1orf86        648  \n",
       "3       ENSG00000157870   FAM213B        838  \n",
       "4       ENSG00000215912     TTC34        889  \n",
       "...                 ...       ...        ...  \n",
       "731079  ENSG00000166049     PASD1     802181  \n",
       "731080  ENSG00000102287     GABRE     802336  \n",
       "731081  ENSG00000183479     TREX2     802547  \n",
       "731082  ENSG00000184343     SRPK3     802648  \n",
       "731083  ENSG00000269335     IKBKG     802968  \n",
       "\n",
       "[731084 rows x 23 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ot_bim"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load GO term to gene mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "import pickle "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_path = \"../../data/GO_biological_process.pickle\"\n",
    "G = pickle.load(open(go_path, 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load go terms and genes\n",
    "goterms_df = pd.read_csv(\"../../ontology_operations/data/GO_terms_biological_processes_merged_EBI_uniprot.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>merged</th>\n",
       "      <th>len_merged</th>\n",
       "      <th>name</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GO:0000002</th>\n",
       "      <td>SLC25A33;PRSS15;MEF2;MGME1;PIF1;AAC1;C15orf20;...</td>\n",
       "      <td>21</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>SLC25A33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0000002</th>\n",
       "      <td>SLC25A33;PRSS15;MEF2;MGME1;PIF1;AAC1;C15orf20;...</td>\n",
       "      <td>21</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>PRSS15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0000002</th>\n",
       "      <td>SLC25A33;PRSS15;MEF2;MGME1;PIF1;AAC1;C15orf20;...</td>\n",
       "      <td>21</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>MEF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0000002</th>\n",
       "      <td>SLC25A33;PRSS15;MEF2;MGME1;PIF1;AAC1;C15orf20;...</td>\n",
       "      <td>21</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>MGME1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0000002</th>\n",
       "      <td>SLC25A33;PRSS15;MEF2;MGME1;PIF1;AAC1;C15orf20;...</td>\n",
       "      <td>21</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>PIF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:2001272</th>\n",
       "      <td>OK/SW-cl.26;RNF78;RFPL1L;RPS3;RFPL1</td>\n",
       "      <td>5</td>\n",
       "      <td>positive regulation of cysteine-type endopepti...</td>\n",
       "      <td>RFPL1L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:2001272</th>\n",
       "      <td>OK/SW-cl.26;RNF78;RFPL1L;RPS3;RFPL1</td>\n",
       "      <td>5</td>\n",
       "      <td>positive regulation of cysteine-type endopepti...</td>\n",
       "      <td>RPS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:2001272</th>\n",
       "      <td>OK/SW-cl.26;RNF78;RFPL1L;RPS3;RFPL1</td>\n",
       "      <td>5</td>\n",
       "      <td>positive regulation of cysteine-type endopepti...</td>\n",
       "      <td>RFPL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:2001304</th>\n",
       "      <td>CYP4F3;LTB4H</td>\n",
       "      <td>2</td>\n",
       "      <td>lipoxin B4 metabolic process</td>\n",
       "      <td>CYP4F3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:2001304</th>\n",
       "      <td>CYP4F3;LTB4H</td>\n",
       "      <td>2</td>\n",
       "      <td>lipoxin B4 metabolic process</td>\n",
       "      <td>LTB4H</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>286507 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       merged  len_merged  \\\n",
       "GO:0000002  SLC25A33;PRSS15;MEF2;MGME1;PIF1;AAC1;C15orf20;...          21   \n",
       "GO:0000002  SLC25A33;PRSS15;MEF2;MGME1;PIF1;AAC1;C15orf20;...          21   \n",
       "GO:0000002  SLC25A33;PRSS15;MEF2;MGME1;PIF1;AAC1;C15orf20;...          21   \n",
       "GO:0000002  SLC25A33;PRSS15;MEF2;MGME1;PIF1;AAC1;C15orf20;...          21   \n",
       "GO:0000002  SLC25A33;PRSS15;MEF2;MGME1;PIF1;AAC1;C15orf20;...          21   \n",
       "...                                                       ...         ...   \n",
       "GO:2001272                OK/SW-cl.26;RNF78;RFPL1L;RPS3;RFPL1           5   \n",
       "GO:2001272                OK/SW-cl.26;RNF78;RFPL1L;RPS3;RFPL1           5   \n",
       "GO:2001272                OK/SW-cl.26;RNF78;RFPL1L;RPS3;RFPL1           5   \n",
       "GO:2001304                                       CYP4F3;LTB4H           2   \n",
       "GO:2001304                                       CYP4F3;LTB4H           2   \n",
       "\n",
       "                                                         name gene_name  \n",
       "GO:0000002                   mitochondrial genome maintenance  SLC25A33  \n",
       "GO:0000002                   mitochondrial genome maintenance    PRSS15  \n",
       "GO:0000002                   mitochondrial genome maintenance      MEF2  \n",
       "GO:0000002                   mitochondrial genome maintenance     MGME1  \n",
       "GO:0000002                   mitochondrial genome maintenance      PIF1  \n",
       "...                                                       ...       ...  \n",
       "GO:2001272  positive regulation of cysteine-type endopepti...    RFPL1L  \n",
       "GO:2001272  positive regulation of cysteine-type endopepti...      RPS3  \n",
       "GO:2001272  positive regulation of cysteine-type endopepti...     RFPL1  \n",
       "GO:2001304                       lipoxin B4 metabolic process    CYP4F3  \n",
       "GO:2001304                       lipoxin B4 metabolic process     LTB4H  \n",
       "\n",
       "[286507 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "goterms_df[\"gene_name\"] = goterms_df[\"merged\"].apply(lambda x: x.split(\";\"))\n",
    "goterms_df = goterms_df.explode(\"gene_name\")\n",
    "goterms_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df = pd.DataFrame()\n",
    "gene_df[\"gene_name\"] = goterms_df.gene_name.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_bim = ot_bim.merge(gene_df, on=\"gene_name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16042"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_bim.gene_name.nunique() # some genes not mapped to go terms .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_id_prot_coding</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>snp_bed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1169204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1104584</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.61, 'raw': -0.223786}</td>\n",
       "      <td>{'gnomad_afr': 0.1370131366674349, 'gnomad_amr...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>1</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>0</td>\n",
       "      <td>1104584</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>TTLL10</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>1</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>0</td>\n",
       "      <td>1969752</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2194104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>2125543</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 4.241, 'raw': 0.098415}</td>\n",
       "      <td>{'gnomad_afr': 0.00034498620055197794, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>0</td>\n",
       "      <td>2125543</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>C1orf86</td>\n",
       "      <td>648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2592668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>2524107</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 23.9, 'raw': 3.118097}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>0</td>\n",
       "      <td>2524107</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>FAM213B</td>\n",
       "      <td>838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>3041036</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>2957600</td>\n",
       "      <td>rs12409277</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 1.823, 'raw': -0.059547}</td>\n",
       "      <td>{'gnomad_afr': 0.13893357848770396, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000169717</td>\n",
       "      <td>1</td>\n",
       "      <td>rs12409277</td>\n",
       "      <td>0</td>\n",
       "      <td>2957600</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000169717</td>\n",
       "      <td>ACTRT2</td>\n",
       "      <td>958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644889</th>\n",
       "      <td>X</td>\n",
       "      <td>151609900</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>X</td>\n",
       "      <td>150778372</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 3.523, 'raw': 0.055668}</td>\n",
       "      <td>{'gnomad_afr': 0.11482740118046861, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000166049</td>\n",
       "      <td>X</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>0</td>\n",
       "      <td>150778372</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000166049</td>\n",
       "      <td>PASD1</td>\n",
       "      <td>802181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644890</th>\n",
       "      <td>X</td>\n",
       "      <td>152014563</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>151183035</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 2.456, 'raw': -0.011208}</td>\n",
       "      <td>{'gnomad_afr': 0.009298393913778529, 'gnomad_a...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>X</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>0</td>\n",
       "      <td>151183035</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>GABRE</td>\n",
       "      <td>802336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644891</th>\n",
       "      <td>X</td>\n",
       "      <td>153459591</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>152725049</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 2.968, 'raw': 0.022059}</td>\n",
       "      <td>{'gnomad_afr': 0.0032312925170068026, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>0</td>\n",
       "      <td>152725049</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>TREX2</td>\n",
       "      <td>802547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644892</th>\n",
       "      <td>X</td>\n",
       "      <td>153781540</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>153046995</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 22.5, 'raw': 2.480113}</td>\n",
       "      <td>{'gnomad_afr': 0.00017018379850238256, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>X</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>0</td>\n",
       "      <td>153046995</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>SRPK3</td>\n",
       "      <td>802648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644893</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>0</td>\n",
       "      <td>153762634</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>644894 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0           1    1169204          T          C          1       1104584   \n",
       "1           1    2038313          C          T          1       1969752   \n",
       "2           1    2194104          A          G          1       2125543   \n",
       "3           1    2592668          C          T          1       2524107   \n",
       "4           1    3041036          T          C          1       2957600   \n",
       "...       ...        ...        ...        ...        ...           ...   \n",
       "644889      X  151609900          A          G          X     150778372   \n",
       "644890      X  152014563          G          A          X     151183035   \n",
       "644891      X  153459591          C          T          X     152725049   \n",
       "644892      X  153781540          A          C          X     153046995   \n",
       "644893      X  154534419          G          A          X     153762634   \n",
       "\n",
       "              rs_id most_severe_consequence  \\\n",
       "0       rs112695918   upstream_gene_variant   \n",
       "1        rs74372554      intergenic_variant   \n",
       "2       rs146607591        missense_variant   \n",
       "3       rs146100312        missense_variant   \n",
       "4        rs12409277      intergenic_variant   \n",
       "...             ...                     ...   \n",
       "644889   rs58448897          intron_variant   \n",
       "644890  rs145754054      intergenic_variant   \n",
       "644891    rs5945389          intron_variant   \n",
       "644892  rs199915647        missense_variant   \n",
       "644893    rs5030868        missense_variant   \n",
       "\n",
       "                                      cadd  \\\n",
       "0        {'phred': 0.61, 'raw': -0.223786}   \n",
       "1       {'phred': 0.309, 'raw': -0.324059}   \n",
       "2        {'phred': 4.241, 'raw': 0.098415}   \n",
       "3         {'phred': 23.9, 'raw': 3.118097}   \n",
       "4       {'phred': 1.823, 'raw': -0.059547}   \n",
       "...                                    ...   \n",
       "644889   {'phred': 3.523, 'raw': 0.055668}   \n",
       "644890  {'phred': 2.456, 'raw': -0.011208}   \n",
       "644891   {'phred': 2.968, 'raw': 0.022059}   \n",
       "644892    {'phred': 22.5, 'raw': 2.480113}   \n",
       "644893    {'phred': 24.7, 'raw': 3.374046}   \n",
       "\n",
       "                                                       af  ...  \\\n",
       "0       {'gnomad_afr': 0.1370131366674349, 'gnomad_amr...  ...   \n",
       "1       {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   \n",
       "2       {'gnomad_afr': 0.00034498620055197794, 'gnomad...  ...   \n",
       "3       {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...  ...   \n",
       "4       {'gnomad_afr': 0.13893357848770396, 'gnomad_am...  ...   \n",
       "...                                                   ...  ...   \n",
       "644889  {'gnomad_afr': 0.11482740118046861, 'gnomad_am...  ...   \n",
       "644890  {'gnomad_afr': 0.009298393913778529, 'gnomad_a...  ...   \n",
       "644891  {'gnomad_afr': 0.0032312925170068026, 'gnomad_...  ...   \n",
       "644892  {'gnomad_afr': 0.00017018379850238256, 'gnomad...  ...   \n",
       "644893  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   \n",
       "\n",
       "        gene_id_prot_coding chrom          snp cm        pos a1  a2  \\\n",
       "0           ENSG00000162571     1  rs112695918  0    1104584  T   C   \n",
       "1           ENSG00000067606     1   rs74372554  0    1969752  C   T   \n",
       "2           ENSG00000162585     1  rs146607591  0    2125543  A   G   \n",
       "3           ENSG00000157870     1  rs146100312  0    2524107  C   T   \n",
       "4           ENSG00000169717     1   rs12409277  0    2957600  T   C   \n",
       "...                     ...   ...          ... ..        ... ..  ..   \n",
       "644889      ENSG00000166049     X   rs58448897  0  150778372  A   G   \n",
       "644890      ENSG00000102287     X  rs145754054  0  151183035  G   A   \n",
       "644891      ENSG00000183479     X    rs5945389  0  152725049  C   T   \n",
       "644892      ENSG00000184343     X  rs199915647  0  153046995  A   C   \n",
       "644893      ENSG00000269335     X    rs5030868  0  153762634  G   A   \n",
       "\n",
       "                gene_id gene_name snp_bed_id  \n",
       "0       ENSG00000162571    TTLL10        220  \n",
       "1       ENSG00000067606     PRKCZ        599  \n",
       "2       ENSG00000162585   C1orf86        648  \n",
       "3       ENSG00000157870   FAM213B        838  \n",
       "4       ENSG00000169717    ACTRT2        958  \n",
       "...                 ...       ...        ...  \n",
       "644889  ENSG00000166049     PASD1     802181  \n",
       "644890  ENSG00000102287     GABRE     802336  \n",
       "644891  ENSG00000183479     TREX2     802547  \n",
       "644892  ENSG00000184343     SRPK3     802648  \n",
       "644893  ENSG00000269335     IKBKG     802968  \n",
       "\n",
       "[644894 rows x 23 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_bed_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TTLL10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>PRKCZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>C1orf86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>FAM213B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ACTRT2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16037</th>\n",
       "      <td>16037</td>\n",
       "      <td>RPL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16038</th>\n",
       "      <td>16038</td>\n",
       "      <td>HOXA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16039</th>\n",
       "      <td>16039</td>\n",
       "      <td>TAF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16040</th>\n",
       "      <td>16040</td>\n",
       "      <td>TPRN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16041</th>\n",
       "      <td>16041</td>\n",
       "      <td>NOMO1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16042 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_bed_id gene_name\n",
       "0                0    TTLL10\n",
       "1                1     PRKCZ\n",
       "2                2   C1orf86\n",
       "3                3   FAM213B\n",
       "4                4    ACTRT2\n",
       "...            ...       ...\n",
       "16037        16037     RPL18\n",
       "16038        16038     HOXA3\n",
       "16039        16039      TAF6\n",
       "16040        16040      TPRN\n",
       "16041        16041     NOMO1\n",
       "\n",
       "[16042 rows x 2 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df = pd.DataFrame()\n",
    "gene_df[\"gene_name\"] = go_bim.gene_name.unique()\n",
    "gene_df.reset_index(inplace=True)\n",
    "gene_df.columns = [\"gene_bed_id\", \"gene_name\"]\n",
    "gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>...</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>snp_bed_id</th>\n",
       "      <th>gene_bed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1169204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1104584</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.61, 'raw': -0.223786}</td>\n",
       "      <td>{'gnomad_afr': 0.1370131366674349, 'gnomad_amr...</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>0</td>\n",
       "      <td>1104584</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>TTLL10</td>\n",
       "      <td>220</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>0</td>\n",
       "      <td>1969752</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2194104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>2125543</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 4.241, 'raw': 0.098415}</td>\n",
       "      <td>{'gnomad_afr': 0.00034498620055197794, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>0</td>\n",
       "      <td>2125543</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>C1orf86</td>\n",
       "      <td>648</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2592668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>2524107</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 23.9, 'raw': 3.118097}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>0</td>\n",
       "      <td>2524107</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>FAM213B</td>\n",
       "      <td>838</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>3041036</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>2957600</td>\n",
       "      <td>rs12409277</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 1.823, 'raw': -0.059547}</td>\n",
       "      <td>{'gnomad_afr': 0.13893357848770396, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>rs12409277</td>\n",
       "      <td>0</td>\n",
       "      <td>2957600</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000169717</td>\n",
       "      <td>ACTRT2</td>\n",
       "      <td>958</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644889</th>\n",
       "      <td>X</td>\n",
       "      <td>151609900</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>X</td>\n",
       "      <td>150778372</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 3.523, 'raw': 0.055668}</td>\n",
       "      <td>{'gnomad_afr': 0.11482740118046861, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>0</td>\n",
       "      <td>150778372</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000166049</td>\n",
       "      <td>PASD1</td>\n",
       "      <td>802181</td>\n",
       "      <td>9628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644890</th>\n",
       "      <td>X</td>\n",
       "      <td>152014563</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>151183035</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 2.456, 'raw': -0.011208}</td>\n",
       "      <td>{'gnomad_afr': 0.009298393913778529, 'gnomad_a...</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>0</td>\n",
       "      <td>151183035</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>GABRE</td>\n",
       "      <td>802336</td>\n",
       "      <td>7660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644891</th>\n",
       "      <td>X</td>\n",
       "      <td>153459591</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>152725049</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 2.968, 'raw': 0.022059}</td>\n",
       "      <td>{'gnomad_afr': 0.0032312925170068026, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>0</td>\n",
       "      <td>152725049</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>TREX2</td>\n",
       "      <td>802547</td>\n",
       "      <td>2413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644892</th>\n",
       "      <td>X</td>\n",
       "      <td>153781540</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>153046995</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 22.5, 'raw': 2.480113}</td>\n",
       "      <td>{'gnomad_afr': 0.00017018379850238256, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>0</td>\n",
       "      <td>153046995</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>SRPK3</td>\n",
       "      <td>802648</td>\n",
       "      <td>12949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644893</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>0</td>\n",
       "      <td>153762634</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "      <td>13478</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>644894 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0           1    1169204          T          C          1       1104584   \n",
       "1           1    2038313          C          T          1       1969752   \n",
       "2           1    2194104          A          G          1       2125543   \n",
       "3           1    2592668          C          T          1       2524107   \n",
       "4           1    3041036          T          C          1       2957600   \n",
       "...       ...        ...        ...        ...        ...           ...   \n",
       "644889      X  151609900          A          G          X     150778372   \n",
       "644890      X  152014563          G          A          X     151183035   \n",
       "644891      X  153459591          C          T          X     152725049   \n",
       "644892      X  153781540          A          C          X     153046995   \n",
       "644893      X  154534419          G          A          X     153762634   \n",
       "\n",
       "              rs_id most_severe_consequence  \\\n",
       "0       rs112695918   upstream_gene_variant   \n",
       "1        rs74372554      intergenic_variant   \n",
       "2       rs146607591        missense_variant   \n",
       "3       rs146100312        missense_variant   \n",
       "4        rs12409277      intergenic_variant   \n",
       "...             ...                     ...   \n",
       "644889   rs58448897          intron_variant   \n",
       "644890  rs145754054      intergenic_variant   \n",
       "644891    rs5945389          intron_variant   \n",
       "644892  rs199915647        missense_variant   \n",
       "644893    rs5030868        missense_variant   \n",
       "\n",
       "                                      cadd  \\\n",
       "0        {'phred': 0.61, 'raw': -0.223786}   \n",
       "1       {'phred': 0.309, 'raw': -0.324059}   \n",
       "2        {'phred': 4.241, 'raw': 0.098415}   \n",
       "3         {'phred': 23.9, 'raw': 3.118097}   \n",
       "4       {'phred': 1.823, 'raw': -0.059547}   \n",
       "...                                    ...   \n",
       "644889   {'phred': 3.523, 'raw': 0.055668}   \n",
       "644890  {'phred': 2.456, 'raw': -0.011208}   \n",
       "644891   {'phred': 2.968, 'raw': 0.022059}   \n",
       "644892    {'phred': 22.5, 'raw': 2.480113}   \n",
       "644893    {'phred': 24.7, 'raw': 3.374046}   \n",
       "\n",
       "                                                       af  ...  chrom  \\\n",
       "0       {'gnomad_afr': 0.1370131366674349, 'gnomad_amr...  ...      1   \n",
       "1       {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...      1   \n",
       "2       {'gnomad_afr': 0.00034498620055197794, 'gnomad...  ...      1   \n",
       "3       {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...  ...      1   \n",
       "4       {'gnomad_afr': 0.13893357848770396, 'gnomad_am...  ...      1   \n",
       "...                                                   ...  ...    ...   \n",
       "644889  {'gnomad_afr': 0.11482740118046861, 'gnomad_am...  ...      X   \n",
       "644890  {'gnomad_afr': 0.009298393913778529, 'gnomad_a...  ...      X   \n",
       "644891  {'gnomad_afr': 0.0032312925170068026, 'gnomad_...  ...      X   \n",
       "644892  {'gnomad_afr': 0.00017018379850238256, 'gnomad...  ...      X   \n",
       "644893  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...      X   \n",
       "\n",
       "                snp  cm        pos a1 a2          gene_id  gene_name  \\\n",
       "0       rs112695918   0    1104584  T  C  ENSG00000162571     TTLL10   \n",
       "1        rs74372554   0    1969752  C  T  ENSG00000067606      PRKCZ   \n",
       "2       rs146607591   0    2125543  A  G  ENSG00000162585    C1orf86   \n",
       "3       rs146100312   0    2524107  C  T  ENSG00000157870    FAM213B   \n",
       "4        rs12409277   0    2957600  T  C  ENSG00000169717     ACTRT2   \n",
       "...             ...  ..        ... .. ..              ...        ...   \n",
       "644889   rs58448897   0  150778372  A  G  ENSG00000166049      PASD1   \n",
       "644890  rs145754054   0  151183035  G  A  ENSG00000102287      GABRE   \n",
       "644891    rs5945389   0  152725049  C  T  ENSG00000183479      TREX2   \n",
       "644892  rs199915647   0  153046995  A  C  ENSG00000184343      SRPK3   \n",
       "644893    rs5030868   0  153762634  G  A  ENSG00000269335      IKBKG   \n",
       "\n",
       "       snp_bed_id gene_bed_id  \n",
       "0             220           0  \n",
       "1             599           1  \n",
       "2             648           2  \n",
       "3             838           3  \n",
       "4             958           4  \n",
       "...           ...         ...  \n",
       "644889     802181        9628  \n",
       "644890     802336        7660  \n",
       "644891     802547        2413  \n",
       "644892     802648       12949  \n",
       "644893     802968       13478  \n",
       "\n",
       "[644894 rows x 24 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_bim = go_bim.merge(gene_df, on=\"gene_name\")\n",
    "go_bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_snp_ids = go_bim.snp_bed_id.unique() # list of all snp ids, enumerate and map to ids\n",
    "snp_id_dict = dict(zip(all_snp_ids, range(len(all_snp_ids))))\n",
    "# snp position in gene_level bed file\n",
    "go_bim['snp_gbed_id'] = go_bim['snp_bed_id'].map(snp_id_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>snp_bed_id</th>\n",
       "      <th>snp_gbed_id</th>\n",
       "      <th>gene_bed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs112695918</td>\n",
       "      <td>TTLL10</td>\n",
       "      <td>220</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs74372554</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs146607591</td>\n",
       "      <td>C1orf86</td>\n",
       "      <td>648</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs146100312</td>\n",
       "      <td>FAM213B</td>\n",
       "      <td>838</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs12409277</td>\n",
       "      <td>ACTRT2</td>\n",
       "      <td>958</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644889</th>\n",
       "      <td>rs58448897</td>\n",
       "      <td>PASD1</td>\n",
       "      <td>802181</td>\n",
       "      <td>644889</td>\n",
       "      <td>9628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644890</th>\n",
       "      <td>rs145754054</td>\n",
       "      <td>GABRE</td>\n",
       "      <td>802336</td>\n",
       "      <td>644890</td>\n",
       "      <td>7660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644891</th>\n",
       "      <td>rs5945389</td>\n",
       "      <td>TREX2</td>\n",
       "      <td>802547</td>\n",
       "      <td>644891</td>\n",
       "      <td>2413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644892</th>\n",
       "      <td>rs199915647</td>\n",
       "      <td>SRPK3</td>\n",
       "      <td>802648</td>\n",
       "      <td>644892</td>\n",
       "      <td>12949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644893</th>\n",
       "      <td>rs5030868</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "      <td>644893</td>\n",
       "      <td>13478</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>644894 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                snp gene_name  snp_bed_id  snp_gbed_id  gene_bed_id\n",
       "0       rs112695918    TTLL10         220            0            0\n",
       "1        rs74372554     PRKCZ         599            1            1\n",
       "2       rs146607591   C1orf86         648            2            2\n",
       "3       rs146100312   FAM213B         838            3            3\n",
       "4        rs12409277    ACTRT2         958            4            4\n",
       "...             ...       ...         ...          ...          ...\n",
       "644889   rs58448897     PASD1      802181       644889         9628\n",
       "644890  rs145754054     GABRE      802336       644890         7660\n",
       "644891    rs5945389     TREX2      802547       644891         2413\n",
       "644892  rs199915647     SRPK3      802648       644892        12949\n",
       "644893    rs5030868     IKBKG      802968       644893        13478\n",
       "\n",
       "[644894 rows x 5 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_bim[[\"snp\", \"gene_name\", \"snp_bed_id\", \"snp_gbed_id\", \"gene_bed_id\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_bim[[\"snp\", \"gene_name\", \"snp_bed_id\", \"snp_gbed_id\", \"gene_bed_id\"]].to_csv(\"../data/go_biological_processes_bim.csv\", index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_bim.to_csv(\"../data/go_biological_processes_bim_full.csv\", index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prune network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "import pickle\n",
    "\n",
    "go_path = \"../../ontology_operations/data/GO_biological_process.pickle\"\n",
    "G = pickle.load(open(go_path, 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_bim = pd.read_csv(\"../data/go_biological_processes_bim_full.csv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_go = go_bim.merge(goterms_df.reset_index(), on=\"gene_name\") # .... lot's of multiple connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>...</th>\n",
       "      <th>a2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>snp_bed_id</th>\n",
       "      <th>gene_bed_id</th>\n",
       "      <th>snp_gbed_id</th>\n",
       "      <th>index</th>\n",
       "      <th>merged</th>\n",
       "      <th>len_merged</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1169204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1104584</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.61, 'raw': -0.223786}</td>\n",
       "      <td>{'gnomad_afr': 0.1370131366674349, 'gnomad_amr...</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>TTLL10</td>\n",
       "      <td>220</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GO:0018094</td>\n",
       "      <td>TTLL8;TTLL3;TTLL10</td>\n",
       "      <td>3</td>\n",
       "      <td>protein polyglycylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>GO:0000226</td>\n",
       "      <td>C6orf97;DYNC1LI1;NF68;HAUS6;B7Z3Y3;LL5B;PARCC1...</td>\n",
       "      <td>346</td>\n",
       "      <td>microtubule cytoskeleton organization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>GO:0001954</td>\n",
       "      <td>DAG1;PLEKHA2;CCL25;CCR7;SCAP1;KIAA0457;CSF1;SC...</td>\n",
       "      <td>75</td>\n",
       "      <td>positive regulation of cell-matrix adhesion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>GO:0006468</td>\n",
       "      <td>STK22A;CDK13;SNRK;PRKACG;DAPK;ZAK;PSKH2;FLT2;D...</td>\n",
       "      <td>1149</td>\n",
       "      <td>protein phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>GO:0006954</td>\n",
       "      <td>LY96;SCN11A;OLR1;C4A;ZAK;TNIP2;F2RL1;LYN;KCNJ8...</td>\n",
       "      <td>1106</td>\n",
       "      <td>inflammatory response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861945</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "      <td>13478</td>\n",
       "      <td>644893</td>\n",
       "      <td>GO:0050852</td>\n",
       "      <td>PTPN6;UNQ744/PRO1472;MNAB;BT2.1;PS1;HLA-DP1B;M...</td>\n",
       "      <td>271</td>\n",
       "      <td>T cell receptor signaling pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861946</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "      <td>13478</td>\n",
       "      <td>644893</td>\n",
       "      <td>GO:0050862</td>\n",
       "      <td>TAPA1;CD81;CARD11;KCNN4;CCR7;RPS3;IKBKG;NFKB3;...</td>\n",
       "      <td>52</td>\n",
       "      <td>positive regulation of T cell receptor signali...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861947</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "      <td>13478</td>\n",
       "      <td>644893</td>\n",
       "      <td>GO:0051092</td>\n",
       "      <td>CARMA2;PIDD1;NFKB3;COP;CLAN1;TRIM62;HSPC187;IK...</td>\n",
       "      <td>322</td>\n",
       "      <td>positive regulation of NF-kappaB transcription...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861948</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "      <td>13478</td>\n",
       "      <td>644893</td>\n",
       "      <td>GO:0051650</td>\n",
       "      <td>RAB7;RAB11;IKBKG;FIP3;SNAPIN;TESK1;TCIRG1;ATP6...</td>\n",
       "      <td>15</td>\n",
       "      <td>establishment of vesicle localization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861949</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "      <td>13478</td>\n",
       "      <td>644893</td>\n",
       "      <td>GO:2000060</td>\n",
       "      <td>NRF2;CCP1;KIAA0457;MJD1;L3MBTL3;FIP3;SEL10;APO...</td>\n",
       "      <td>76</td>\n",
       "      <td>positive regulation of ubiquitin-dependent pro...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5861950 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0            1    1169204          T          C          1       1104584   \n",
       "1            1    2038313          C          T          1       1969752   \n",
       "2            1    2038313          C          T          1       1969752   \n",
       "3            1    2038313          C          T          1       1969752   \n",
       "4            1    2038313          C          T          1       1969752   \n",
       "...        ...        ...        ...        ...        ...           ...   \n",
       "5861945      X  154534419          G          A          X     153762634   \n",
       "5861946      X  154534419          G          A          X     153762634   \n",
       "5861947      X  154534419          G          A          X     153762634   \n",
       "5861948      X  154534419          G          A          X     153762634   \n",
       "5861949      X  154534419          G          A          X     153762634   \n",
       "\n",
       "               rs_id most_severe_consequence  \\\n",
       "0        rs112695918   upstream_gene_variant   \n",
       "1         rs74372554      intergenic_variant   \n",
       "2         rs74372554      intergenic_variant   \n",
       "3         rs74372554      intergenic_variant   \n",
       "4         rs74372554      intergenic_variant   \n",
       "...              ...                     ...   \n",
       "5861945    rs5030868        missense_variant   \n",
       "5861946    rs5030868        missense_variant   \n",
       "5861947    rs5030868        missense_variant   \n",
       "5861948    rs5030868        missense_variant   \n",
       "5861949    rs5030868        missense_variant   \n",
       "\n",
       "                                       cadd  \\\n",
       "0         {'phred': 0.61, 'raw': -0.223786}   \n",
       "1        {'phred': 0.309, 'raw': -0.324059}   \n",
       "2        {'phred': 0.309, 'raw': -0.324059}   \n",
       "3        {'phred': 0.309, 'raw': -0.324059}   \n",
       "4        {'phred': 0.309, 'raw': -0.324059}   \n",
       "...                                     ...   \n",
       "5861945    {'phred': 24.7, 'raw': 3.374046}   \n",
       "5861946    {'phred': 24.7, 'raw': 3.374046}   \n",
       "5861947    {'phred': 24.7, 'raw': 3.374046}   \n",
       "5861948    {'phred': 24.7, 'raw': 3.374046}   \n",
       "5861949    {'phred': 24.7, 'raw': 3.374046}   \n",
       "\n",
       "                                                        af  ...  a2  \\\n",
       "0        {'gnomad_afr': 0.1370131366674349, 'gnomad_amr...  ...   C   \n",
       "1        {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   T   \n",
       "2        {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   T   \n",
       "3        {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   T   \n",
       "4        {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   T   \n",
       "...                                                    ...  ...  ..   \n",
       "5861945  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   A   \n",
       "5861946  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   A   \n",
       "5861947  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   A   \n",
       "5861948  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   A   \n",
       "5861949  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   A   \n",
       "\n",
       "                 gene_id  gene_name snp_bed_id gene_bed_id snp_gbed_id  \\\n",
       "0        ENSG00000162571     TTLL10        220           0           0   \n",
       "1        ENSG00000067606      PRKCZ        599           1           1   \n",
       "2        ENSG00000067606      PRKCZ        599           1           1   \n",
       "3        ENSG00000067606      PRKCZ        599           1           1   \n",
       "4        ENSG00000067606      PRKCZ        599           1           1   \n",
       "...                  ...        ...        ...         ...         ...   \n",
       "5861945  ENSG00000269335      IKBKG     802968       13478      644893   \n",
       "5861946  ENSG00000269335      IKBKG     802968       13478      644893   \n",
       "5861947  ENSG00000269335      IKBKG     802968       13478      644893   \n",
       "5861948  ENSG00000269335      IKBKG     802968       13478      644893   \n",
       "5861949  ENSG00000269335      IKBKG     802968       13478      644893   \n",
       "\n",
       "              index                                             merged  \\\n",
       "0        GO:0018094                                 TTLL8;TTLL3;TTLL10   \n",
       "1        GO:0000226  C6orf97;DYNC1LI1;NF68;HAUS6;B7Z3Y3;LL5B;PARCC1...   \n",
       "2        GO:0001954  DAG1;PLEKHA2;CCL25;CCR7;SCAP1;KIAA0457;CSF1;SC...   \n",
       "3        GO:0006468  STK22A;CDK13;SNRK;PRKACG;DAPK;ZAK;PSKH2;FLT2;D...   \n",
       "4        GO:0006954  LY96;SCN11A;OLR1;C4A;ZAK;TNIP2;F2RL1;LYN;KCNJ8...   \n",
       "...             ...                                                ...   \n",
       "5861945  GO:0050852  PTPN6;UNQ744/PRO1472;MNAB;BT2.1;PS1;HLA-DP1B;M...   \n",
       "5861946  GO:0050862  TAPA1;CD81;CARD11;KCNN4;CCR7;RPS3;IKBKG;NFKB3;...   \n",
       "5861947  GO:0051092  CARMA2;PIDD1;NFKB3;COP;CLAN1;TRIM62;HSPC187;IK...   \n",
       "5861948  GO:0051650  RAB7;RAB11;IKBKG;FIP3;SNAPIN;TESK1;TCIRG1;ATP6...   \n",
       "5861949  GO:2000060  NRF2;CCP1;KIAA0457;MJD1;L3MBTL3;FIP3;SEL10;APO...   \n",
       "\n",
       "        len_merged                                               name  \n",
       "0                3                            protein polyglycylation  \n",
       "1              346              microtubule cytoskeleton organization  \n",
       "2               75        positive regulation of cell-matrix adhesion  \n",
       "3             1149                            protein phosphorylation  \n",
       "4             1106                              inflammatory response  \n",
       "...            ...                                                ...  \n",
       "5861945        271                  T cell receptor signaling pathway  \n",
       "5861946         52  positive regulation of T cell receptor signali...  \n",
       "5861947        322  positive regulation of NF-kappaB transcription...  \n",
       "5861948         15              establishment of vesicle localization  \n",
       "5861949         76  positive regulation of ubiquitin-dependent pro...  \n",
       "\n",
       "[5861950 rows x 29 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11364"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_go[\"index\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_terms = merged_go[\"index\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GO_0060922',\n",
       " 'GO_0006658',\n",
       " 'GO_0097264',\n",
       " 'GO_0061311',\n",
       " 'GO_0035268',\n",
       " 'GO_0030213',\n",
       " 'GO_0021935',\n",
       " 'GO_0021722',\n",
       " 'GO_1902416',\n",
       " 'GO_1903043',\n",
       " 'GO_0032088',\n",
       " 'GO_0090494',\n",
       " 'GO_0046513',\n",
       " 'GO_0097279',\n",
       " 'GO_0031173',\n",
       " 'GO_0098664',\n",
       " 'GO_0050806',\n",
       " 'GO_1902949',\n",
       " 'GO_0072716',\n",
       " 'GO_0097528',\n",
       " 'GO_0042462',\n",
       " 'GO_0034201',\n",
       " 'GO_0006265',\n",
       " 'GO_0070262',\n",
       " 'GO_0061951',\n",
       " 'GO_0035694',\n",
       " 'GO_0008610',\n",
       " 'GO_0046833',\n",
       " 'GO_0016185',\n",
       " 'GO_2000347',\n",
       " 'GO_0060441',\n",
       " 'GO_0043011',\n",
       " 'GO_0048822',\n",
       " 'GO_2000331',\n",
       " 'GO_0060447',\n",
       " 'GO_0034392',\n",
       " 'GO_0032376',\n",
       " 'GO_0045333',\n",
       " 'GO_0001111',\n",
       " 'GO_0140896',\n",
       " 'GO_0002759',\n",
       " 'GO_0002517',\n",
       " 'GO_0055064',\n",
       " 'GO_0002581',\n",
       " 'GO_0106089',\n",
       " 'GO_0051450',\n",
       " 'GO_1901189',\n",
       " 'GO_0035910',\n",
       " 'GO_0042635',\n",
       " 'GO_0034148',\n",
       " 'GO_0010569',\n",
       " 'GO_1904045',\n",
       " 'GO_0034088',\n",
       " 'GO_0070889',\n",
       " 'GO_0009583',\n",
       " 'GO_0065002',\n",
       " 'GO_1904108',\n",
       " 'GO_2000768',\n",
       " 'GO_0019083',\n",
       " 'GO_1990115',\n",
       " 'GO_1990466',\n",
       " 'GO_0009396',\n",
       " 'GO_2000172',\n",
       " 'GO_0044381',\n",
       " 'GO_0046052',\n",
       " 'GO_0061092',\n",
       " 'GO_0043584',\n",
       " 'GO_0010037',\n",
       " 'GO_0062112',\n",
       " 'GO_0003018',\n",
       " 'GO_0060770',\n",
       " 'GO_0018910',\n",
       " 'GO_1904613',\n",
       " 'GO_0000461',\n",
       " 'GO_0042246',\n",
       " 'GO_1902530',\n",
       " 'GO_0097403',\n",
       " 'GO_0106091',\n",
       " 'GO_0035822',\n",
       " 'GO_2000655',\n",
       " 'GO_0072592',\n",
       " 'GO_0036305',\n",
       " 'GO_0043393',\n",
       " 'GO_0048266',\n",
       " 'GO_0120168',\n",
       " 'GO_0099504',\n",
       " 'GO_1904640',\n",
       " 'GO_0002364',\n",
       " 'GO_0030970',\n",
       " 'GO_1903367',\n",
       " 'GO_0048144',\n",
       " 'GO_0051987',\n",
       " 'GO_0038146',\n",
       " 'GO_0050927',\n",
       " 'GO_0071559',\n",
       " 'GO_0015876',\n",
       " 'GO_1901382',\n",
       " 'GO_0072721',\n",
       " 'GO_0007160',\n",
       " 'GO_0006196',\n",
       " 'GO_0051459',\n",
       " 'GO_0033326',\n",
       " 'GO_0021879',\n",
       " 'GO_0015964',\n",
       " 'GO_0042692',\n",
       " 'GO_0060268',\n",
       " 'GO_0002540',\n",
       " 'GO_1903817',\n",
       " 'GO_0031635',\n",
       " 'GO_0031923',\n",
       " 'GO_0002812',\n",
       " 'GO_0070845',\n",
       " 'GO_0045167',\n",
       " 'GO_0071072',\n",
       " 'GO_0000578',\n",
       " 'GO_0009154',\n",
       " 'GO_1902991',\n",
       " 'GO_2000291',\n",
       " 'GO_2000564',\n",
       " 'GO_0072107',\n",
       " 'GO_0086069',\n",
       " 'GO_0060137',\n",
       " 'GO_1902558',\n",
       " 'GO_0006904',\n",
       " 'GO_0006264',\n",
       " 'GO_1902856',\n",
       " 'GO_1904685',\n",
       " 'GO_0003032',\n",
       " 'GO_0060586',\n",
       " 'GO_0043508',\n",
       " 'GO_0160063',\n",
       " 'GO_0051561',\n",
       " 'GO_0032201',\n",
       " 'GO_0051977',\n",
       " 'GO_0001836',\n",
       " 'GO_0016199',\n",
       " 'GO_0061817',\n",
       " 'GO_1904356',\n",
       " 'GO_1905144',\n",
       " 'GO_1905048',\n",
       " 'GO_0032897',\n",
       " 'GO_1904828',\n",
       " 'GO_0006041',\n",
       " 'GO_0071409',\n",
       " 'GO_0048073',\n",
       " 'GO_2000521',\n",
       " 'GO_2000984',\n",
       " 'GO_0045324',\n",
       " 'GO_1904960',\n",
       " 'GO_0099562',\n",
       " 'GO_0048137',\n",
       " 'GO_0001826',\n",
       " 'GO_0050651',\n",
       " 'GO_0032057',\n",
       " 'GO_0045453',\n",
       " 'GO_0071875',\n",
       " 'GO_0061355',\n",
       " 'GO_1905375',\n",
       " 'GO_0048808',\n",
       " 'GO_0072278',\n",
       " 'GO_0042102',\n",
       " 'GO_0046668',\n",
       " 'GO_1901159',\n",
       " 'GO_0071205',\n",
       " 'GO_0035094',\n",
       " 'GO_0042117',\n",
       " 'GO_0006116',\n",
       " 'GO_0010351',\n",
       " 'GO_1905706',\n",
       " 'GO_0070307',\n",
       " 'GO_0030865',\n",
       " 'GO_0060915',\n",
       " 'GO_0060129',\n",
       " 'GO_0021512',\n",
       " 'GO_0010976',\n",
       " 'GO_0031643',\n",
       " 'GO_0034231',\n",
       " 'GO_0038095',\n",
       " 'GO_0008210',\n",
       " 'GO_0046037',\n",
       " 'GO_0071380',\n",
       " 'GO_0070417',\n",
       " 'GO_0097377',\n",
       " 'GO_0048378',\n",
       " 'GO_0070184',\n",
       " 'GO_0014012',\n",
       " 'GO_0036153',\n",
       " 'GO_0160203',\n",
       " 'GO_0097070',\n",
       " 'GO_1900142',\n",
       " 'GO_0040018',\n",
       " 'GO_1904803',\n",
       " 'GO_0002315',\n",
       " 'GO_0001678',\n",
       " 'GO_0035239',\n",
       " 'GO_0016031',\n",
       " 'GO_0003104',\n",
       " 'GO_0050975',\n",
       " 'GO_0051794',\n",
       " 'GO_0050869',\n",
       " 'GO_0032929',\n",
       " 'GO_0032534',\n",
       " 'GO_0061188',\n",
       " 'GO_0006668',\n",
       " 'GO_0030997',\n",
       " 'GO_1903010',\n",
       " 'GO_2000158',\n",
       " 'GO_0042113',\n",
       " 'GO_0002224',\n",
       " 'GO_0030432',\n",
       " 'GO_1990180',\n",
       " 'GO_0046651',\n",
       " 'GO_0006576',\n",
       " 'GO_0002230',\n",
       " 'GO_0046226',\n",
       " 'GO_0045606',\n",
       " 'GO_2000368',\n",
       " 'GO_0007253',\n",
       " 'GO_0036115',\n",
       " 'GO_0070948',\n",
       " 'GO_0034255',\n",
       " 'GO_1903494',\n",
       " 'GO_0007617',\n",
       " 'GO_0097326',\n",
       " 'GO_0051153',\n",
       " 'GO_0043609',\n",
       " 'GO_0060765',\n",
       " 'GO_0036306',\n",
       " 'GO_1902966',\n",
       " 'GO_0036293',\n",
       " 'GO_0002605',\n",
       " 'GO_0032571',\n",
       " 'GO_0036520',\n",
       " 'GO_1903421',\n",
       " 'GO_1904442',\n",
       " 'GO_1902960',\n",
       " 'GO_1904867',\n",
       " 'GO_0140460',\n",
       " 'GO_0035513',\n",
       " 'GO_0050863',\n",
       " 'GO_0006091',\n",
       " 'GO_0021899',\n",
       " 'GO_0036446',\n",
       " 'GO_0009624',\n",
       " 'GO_0006782',\n",
       " 'GO_0061900',\n",
       " 'GO_0099173',\n",
       " 'GO_0034199',\n",
       " 'GO_0001680',\n",
       " 'GO_0071383',\n",
       " 'GO_1902490',\n",
       " 'GO_0051152',\n",
       " 'GO_0097039',\n",
       " 'GO_2000105',\n",
       " 'GO_0030050',\n",
       " 'GO_0002227',\n",
       " 'GO_0060648',\n",
       " 'GO_1903914',\n",
       " 'GO_0106071',\n",
       " 'GO_0061072',\n",
       " 'GO_0007019',\n",
       " 'GO_0051654',\n",
       " 'GO_1902943',\n",
       " 'GO_0021678',\n",
       " 'GO_0030335',\n",
       " 'GO_0097106',\n",
       " 'GO_0006089',\n",
       " 'GO_1903052',\n",
       " 'GO_0045842',\n",
       " 'GO_0006891',\n",
       " 'GO_1905772',\n",
       " 'GO_0006903',\n",
       " 'GO_0031339',\n",
       " 'GO_1990806',\n",
       " 'GO_0043586',\n",
       " 'GO_0003009',\n",
       " 'GO_0045948',\n",
       " 'GO_0006268',\n",
       " 'GO_1904816',\n",
       " 'GO_0035701',\n",
       " 'GO_0015787',\n",
       " 'GO_0120034',\n",
       " 'GO_2000860',\n",
       " 'GO_0030878',\n",
       " 'GO_0071225',\n",
       " 'GO_0006635',\n",
       " 'GO_0007341',\n",
       " 'GO_0072168',\n",
       " 'GO_0002387',\n",
       " 'GO_0032786',\n",
       " 'GO_1903373',\n",
       " 'GO_0048627',\n",
       " 'GO_0098698',\n",
       " 'GO_1903996',\n",
       " 'GO_0000266',\n",
       " 'GO_0055017',\n",
       " 'GO_0036486',\n",
       " 'GO_0034030',\n",
       " 'GO_0021782',\n",
       " 'GO_0070383',\n",
       " 'GO_0061033',\n",
       " 'GO_0010482',\n",
       " 'GO_0032211',\n",
       " 'GO_0055059',\n",
       " 'GO_0001946',\n",
       " 'GO_0090158',\n",
       " 'GO_0032796',\n",
       " 'GO_0021548',\n",
       " 'GO_0072180',\n",
       " 'GO_0001832',\n",
       " 'GO_0014003',\n",
       " 'GO_0017157',\n",
       " 'GO_2000052',\n",
       " 'GO_0015791',\n",
       " 'GO_0003062',\n",
       " 'GO_0051851',\n",
       " 'GO_1990547',\n",
       " 'GO_0098813',\n",
       " 'GO_0001976',\n",
       " 'GO_1900273',\n",
       " 'GO_0072161',\n",
       " 'GO_1903017',\n",
       " 'GO_0021522',\n",
       " 'GO_0006203',\n",
       " 'GO_0046449',\n",
       " 'GO_0019086',\n",
       " 'GO_0034340',\n",
       " 'GO_0019449',\n",
       " 'GO_0060112',\n",
       " 'GO_0032107',\n",
       " 'GO_0022400',\n",
       " 'GO_0010890',\n",
       " 'GO_0002761',\n",
       " 'GO_0014859',\n",
       " 'GO_0035644',\n",
       " 'GO_0090258',\n",
       " 'GO_0048333',\n",
       " 'GO_0031547',\n",
       " 'GO_0003081',\n",
       " 'GO_2000870',\n",
       " 'GO_1905457',\n",
       " 'GO_0030910',\n",
       " 'GO_1903141',\n",
       " 'GO_0090189',\n",
       " 'GO_0071642',\n",
       " 'GO_0097529',\n",
       " 'GO_0002062',\n",
       " 'GO_0007171',\n",
       " 'GO_0030857',\n",
       " 'GO_0021775',\n",
       " 'GO_0045684',\n",
       " 'GO_0060481',\n",
       " 'GO_1990641',\n",
       " 'GO_0021658',\n",
       " 'GO_0097048',\n",
       " 'GO_0090169',\n",
       " 'GO_0021934',\n",
       " 'GO_0042320',\n",
       " 'GO_0046368',\n",
       " 'GO_0032224',\n",
       " 'GO_1990138',\n",
       " 'GO_0017158',\n",
       " 'GO_1905454',\n",
       " 'GO_0150078',\n",
       " 'GO_0050857',\n",
       " 'GO_1902041',\n",
       " 'GO_0001880',\n",
       " 'GO_0060545',\n",
       " 'GO_2001240',\n",
       " 'GO_1903580',\n",
       " 'GO_2001257',\n",
       " 'GO_0016479',\n",
       " 'GO_0055111',\n",
       " 'GO_0036261',\n",
       " 'GO_0141112',\n",
       " 'GO_0060117',\n",
       " 'GO_0001994',\n",
       " 'GO_0060440',\n",
       " 'GO_0060876',\n",
       " 'GO_0071976',\n",
       " 'GO_1904300',\n",
       " 'GO_0046111',\n",
       " 'GO_0021779',\n",
       " 'GO_0051232',\n",
       " 'GO_1901723',\n",
       " 'GO_0090272',\n",
       " 'GO_0003285',\n",
       " 'GO_0007197',\n",
       " 'GO_1903249',\n",
       " 'GO_0060068',\n",
       " 'GO_0045637',\n",
       " 'GO_1990832',\n",
       " 'GO_0010606',\n",
       " 'GO_0000454',\n",
       " 'GO_0050766',\n",
       " 'GO_0022029',\n",
       " 'GO_0002253',\n",
       " 'GO_0097494',\n",
       " 'GO_0072579',\n",
       " 'GO_1904772',\n",
       " 'GO_0098925',\n",
       " 'GO_1990086',\n",
       " 'GO_0014736',\n",
       " 'GO_1905737',\n",
       " 'GO_0034356',\n",
       " 'GO_0048634',\n",
       " 'GO_1904375',\n",
       " 'GO_0060510',\n",
       " 'GO_0031591',\n",
       " 'GO_1903142',\n",
       " 'GO_2000541',\n",
       " 'GO_0006809',\n",
       " 'GO_0007339',\n",
       " 'GO_0032525',\n",
       " 'GO_0030238',\n",
       " 'GO_2000758',\n",
       " 'GO_0071726',\n",
       " 'GO_0043323',\n",
       " 'GO_1903441',\n",
       " 'GO_0070447',\n",
       " 'GO_0032790',\n",
       " 'GO_0043161',\n",
       " 'GO_0030336',\n",
       " 'GO_0035675',\n",
       " 'GO_0006678',\n",
       " 'GO_0140273',\n",
       " 'GO_2000696',\n",
       " 'GO_2001243',\n",
       " 'GO_1904036',\n",
       " 'GO_0097091',\n",
       " 'GO_0070305',\n",
       " 'GO_0060369',\n",
       " 'GO_0072136',\n",
       " 'GO_0031990',\n",
       " 'GO_0061764',\n",
       " 'GO_0150107',\n",
       " 'GO_0055062',\n",
       " 'GO_0097647',\n",
       " 'GO_0070585',\n",
       " 'GO_0014896',\n",
       " 'GO_1905716',\n",
       " 'GO_1905281',\n",
       " 'GO_1902603',\n",
       " 'GO_2000900',\n",
       " 'GO_0072530',\n",
       " 'GO_0007612',\n",
       " 'GO_0033059',\n",
       " 'GO_0019402',\n",
       " 'GO_0097355',\n",
       " 'GO_0038157',\n",
       " 'GO_0009231',\n",
       " 'GO_1902255',\n",
       " 'GO_1904401',\n",
       " 'GO_0015808',\n",
       " 'GO_0006956',\n",
       " 'GO_0001576',\n",
       " 'GO_0060875',\n",
       " 'GO_0034370',\n",
       " 'GO_0006488',\n",
       " 'GO_1901186',\n",
       " 'GO_1902362',\n",
       " 'GO_0051294',\n",
       " 'GO_0010766',\n",
       " 'GO_0071107',\n",
       " 'GO_0048167',\n",
       " 'GO_0051047',\n",
       " 'GO_0033688',\n",
       " 'GO_0090119',\n",
       " 'GO_0050849',\n",
       " 'GO_0021675',\n",
       " 'GO_0002666',\n",
       " 'GO_0062029',\n",
       " 'GO_2000548',\n",
       " 'GO_0097376',\n",
       " 'GO_0090050',\n",
       " 'GO_0048703',\n",
       " 'GO_1903434',\n",
       " 'GO_0010730',\n",
       " 'GO_0002528',\n",
       " 'GO_1905898',\n",
       " 'GO_0060122',\n",
       " 'GO_0022898',\n",
       " 'GO_0035352',\n",
       " 'GO_0035279',\n",
       " 'GO_1903788',\n",
       " 'GO_0061445',\n",
       " 'GO_0002575',\n",
       " 'GO_0150036',\n",
       " 'GO_0106134',\n",
       " 'GO_0062177',\n",
       " 'GO_2001108',\n",
       " 'GO_0006166',\n",
       " 'GO_0021822',\n",
       " 'GO_0050996',\n",
       " 'GO_0031054',\n",
       " 'GO_0071259',\n",
       " 'GO_0021562',\n",
       " 'GO_0061078',\n",
       " 'GO_1903457',\n",
       " 'GO_0090238',\n",
       " 'GO_0033626',\n",
       " 'GO_2000439',\n",
       " 'GO_0050771',\n",
       " 'GO_0036515',\n",
       " 'GO_2000276',\n",
       " 'GO_0006542',\n",
       " 'GO_0051093',\n",
       " 'GO_1902109',\n",
       " 'GO_0021915',\n",
       " 'GO_0002548',\n",
       " 'GO_0006468',\n",
       " 'GO_0045059',\n",
       " 'GO_2000741',\n",
       " 'GO_0098877',\n",
       " 'GO_0021937',\n",
       " 'GO_0048021',\n",
       " 'GO_0021874',\n",
       " 'GO_0072227',\n",
       " 'GO_0002520',\n",
       " 'GO_0061017',\n",
       " 'GO_0120192',\n",
       " 'GO_0060374',\n",
       " 'GO_0010556',\n",
       " 'GO_0031103',\n",
       " 'GO_0032401',\n",
       " 'GO_0019226',\n",
       " 'GO_1903830',\n",
       " 'GO_0010256',\n",
       " 'GO_0043415',\n",
       " 'GO_0033215',\n",
       " 'GO_0002432',\n",
       " 'GO_0002854',\n",
       " 'GO_1904380',\n",
       " 'GO_1901652',\n",
       " 'GO_0071073',\n",
       " 'GO_1903094',\n",
       " 'GO_0008615',\n",
       " 'GO_0051622',\n",
       " 'GO_0141065',\n",
       " 'GO_0110061',\n",
       " 'GO_0021570',\n",
       " 'GO_0000425',\n",
       " 'GO_0072538',\n",
       " 'GO_1904808',\n",
       " 'GO_0010559',\n",
       " 'GO_0010957',\n",
       " 'GO_1902115',\n",
       " 'GO_0060330',\n",
       " 'GO_0070507',\n",
       " 'GO_0007076',\n",
       " 'GO_0034112',\n",
       " 'GO_0060471',\n",
       " 'GO_0051494',\n",
       " 'GO_0050731',\n",
       " 'GO_0010517',\n",
       " 'GO_0048211',\n",
       " 'GO_0045910',\n",
       " 'GO_0097284',\n",
       " 'GO_0003097',\n",
       " 'GO_0002069',\n",
       " 'GO_0051665',\n",
       " 'GO_2000195',\n",
       " 'GO_0046580',\n",
       " 'GO_0033153',\n",
       " 'GO_0086003',\n",
       " 'GO_0006649',\n",
       " 'GO_0048352',\n",
       " 'GO_0043331',\n",
       " 'GO_0006753',\n",
       " 'GO_0072141',\n",
       " 'GO_0033539',\n",
       " 'GO_1900148',\n",
       " 'GO_0061736',\n",
       " 'GO_1903934',\n",
       " 'GO_0035927',\n",
       " 'GO_0008228',\n",
       " 'GO_0090104',\n",
       " 'GO_0071651',\n",
       " 'GO_0008206',\n",
       " 'GO_0035791',\n",
       " 'GO_0010501',\n",
       " 'GO_0002221',\n",
       " 'GO_0033033',\n",
       " 'GO_0030214',\n",
       " 'GO_1903522',\n",
       " 'GO_1901379',\n",
       " 'GO_0006743',\n",
       " 'GO_0048757',\n",
       " 'GO_2000369',\n",
       " 'GO_0097168',\n",
       " 'GO_0048738',\n",
       " 'GO_0072583',\n",
       " 'GO_0006508',\n",
       " 'GO_0035709',\n",
       " 'GO_0045785',\n",
       " 'GO_0006679',\n",
       " 'GO_0061755',\n",
       " 'GO_0075509',\n",
       " 'GO_0060024',\n",
       " 'GO_0048468',\n",
       " 'GO_0061669',\n",
       " 'GO_0032388',\n",
       " 'GO_0031444',\n",
       " 'GO_0098915',\n",
       " 'GO_0006391',\n",
       " 'GO_0042428',\n",
       " 'GO_0071340',\n",
       " 'GO_0035989',\n",
       " 'GO_0043585',\n",
       " 'GO_0035802',\n",
       " 'GO_0046543',\n",
       " 'GO_1901492',\n",
       " 'GO_0070486',\n",
       " 'GO_1901346',\n",
       " 'GO_0090108',\n",
       " 'GO_0060061',\n",
       " 'GO_0043123',\n",
       " 'GO_0022602',\n",
       " 'GO_2001056',\n",
       " 'GO_0070781',\n",
       " 'GO_2000508',\n",
       " 'GO_1901911',\n",
       " 'GO_0043052',\n",
       " 'GO_0033274',\n",
       " 'GO_1902034',\n",
       " 'GO_1901899',\n",
       " 'GO_2000675',\n",
       " 'GO_0006883',\n",
       " 'GO_0042495',\n",
       " 'GO_0030502',\n",
       " 'GO_1903941',\n",
       " 'GO_1903215',\n",
       " 'GO_0006173',\n",
       " 'GO_0071451',\n",
       " 'GO_0035609',\n",
       " 'GO_0001666',\n",
       " 'GO_0055002',\n",
       " 'GO_0016077',\n",
       " 'GO_2000005',\n",
       " 'GO_0016562',\n",
       " 'GO_0045576',\n",
       " 'GO_0009404',\n",
       " 'GO_0036092',\n",
       " 'GO_0019254',\n",
       " 'GO_1904100',\n",
       " 'GO_0006482',\n",
       " 'GO_0048386',\n",
       " 'GO_2000357',\n",
       " 'GO_0034276',\n",
       " 'GO_2000607',\n",
       " 'GO_2001037',\n",
       " 'GO_0010762',\n",
       " 'GO_0007270',\n",
       " 'GO_0006043',\n",
       " 'GO_1904478',\n",
       " 'GO_2001244',\n",
       " 'GO_0046879',\n",
       " 'GO_0072489',\n",
       " 'GO_0009954',\n",
       " 'GO_0002353',\n",
       " 'GO_0006672',\n",
       " 'GO_0015931',\n",
       " 'GO_0018916',\n",
       " 'GO_0010811',\n",
       " 'GO_0046541',\n",
       " 'GO_0034638',\n",
       " 'GO_1904738',\n",
       " 'GO_0030148',\n",
       " 'GO_0019309',\n",
       " 'GO_0032354',\n",
       " 'GO_0033146',\n",
       " 'GO_0001975',\n",
       " 'GO_0042325',\n",
       " 'GO_0044788',\n",
       " 'GO_0009214',\n",
       " 'GO_0015709',\n",
       " 'GO_2000786',\n",
       " 'GO_0060312',\n",
       " 'GO_1902140',\n",
       " 'GO_0003340',\n",
       " 'GO_0099152',\n",
       " 'GO_0061737',\n",
       " 'GO_2000737',\n",
       " 'GO_0060621',\n",
       " 'GO_0046849',\n",
       " 'GO_0030326',\n",
       " 'GO_0071277',\n",
       " 'GO_2000288',\n",
       " 'GO_0035622',\n",
       " 'GO_0002018',\n",
       " 'GO_0072213',\n",
       " 'GO_1902553',\n",
       " 'GO_0031947',\n",
       " 'GO_1903235',\n",
       " 'GO_0051885',\n",
       " 'GO_0031086',\n",
       " 'GO_0000512',\n",
       " 'GO_0072133',\n",
       " 'GO_0042307',\n",
       " 'GO_0061364',\n",
       " 'GO_0061611',\n",
       " 'GO_0033594',\n",
       " 'GO_0006550',\n",
       " 'GO_0090134',\n",
       " 'GO_0002023',\n",
       " 'GO_0000820',\n",
       " 'GO_1900731',\n",
       " 'GO_0140718',\n",
       " 'GO_0048261',\n",
       " 'GO_0019218',\n",
       " 'GO_0097498',\n",
       " 'GO_0160176',\n",
       " 'GO_0048169',\n",
       " 'GO_0003208',\n",
       " 'GO_0043152',\n",
       " 'GO_0038116',\n",
       " 'GO_0045722',\n",
       " 'GO_0090384',\n",
       " 'GO_0061430',\n",
       " 'GO_0060458',\n",
       " 'GO_0061692',\n",
       " 'GO_0072061',\n",
       " 'GO_0032979',\n",
       " 'GO_0060055',\n",
       " 'GO_0030199',\n",
       " 'GO_0002491',\n",
       " 'GO_2000556',\n",
       " 'GO_0006493',\n",
       " 'GO_1903353',\n",
       " 'GO_0021942',\n",
       " 'GO_0031223',\n",
       " 'GO_0007174',\n",
       " 'GO_0009812',\n",
       " 'GO_0032387',\n",
       " 'GO_0060400',\n",
       " 'GO_0048697',\n",
       " 'GO_0034486',\n",
       " 'GO_0051141',\n",
       " 'GO_0015823',\n",
       " 'GO_0000098',\n",
       " 'GO_0032485',\n",
       " 'GO_0097475',\n",
       " 'GO_0000481',\n",
       " 'GO_0042371',\n",
       " 'GO_0007198',\n",
       " 'GO_0071353',\n",
       " 'GO_0090280',\n",
       " 'GO_0002757',\n",
       " 'GO_0110024',\n",
       " 'GO_0061724',\n",
       " 'GO_0015779',\n",
       " 'GO_1904035',\n",
       " 'GO_1903703',\n",
       " 'GO_0048008',\n",
       " 'GO_0070994',\n",
       " 'GO_0051967',\n",
       " 'GO_0010578',\n",
       " 'GO_0002901',\n",
       " 'GO_0032597',\n",
       " 'GO_1903445',\n",
       " 'GO_0044088',\n",
       " 'GO_0060517',\n",
       " 'GO_1903377',\n",
       " 'GO_0007603',\n",
       " 'GO_1903720',\n",
       " 'GO_0006104',\n",
       " 'GO_0000018',\n",
       " 'GO_1901642',\n",
       " 'GO_0006541',\n",
       " 'GO_0086100',\n",
       " 'GO_0061367',\n",
       " 'GO_0010070',\n",
       " 'GO_2001198',\n",
       " 'GO_2000194',\n",
       " 'GO_0090152',\n",
       " 'GO_0000731',\n",
       " 'GO_0021852',\n",
       " 'GO_0019941',\n",
       " 'GO_1902726',\n",
       " 'GO_0150175',\n",
       " 'GO_0061189',\n",
       " 'GO_0006702',\n",
       " 'GO_1903190',\n",
       " 'GO_0043320',\n",
       " 'GO_1903564',\n",
       " 'GO_1901980',\n",
       " 'GO_0072183',\n",
       " 'GO_0072593',\n",
       " 'GO_0038084',\n",
       " 'GO_0051946',\n",
       " 'GO_0030521',\n",
       " 'GO_0015734',\n",
       " 'GO_0006284',\n",
       " 'GO_0140639',\n",
       " 'GO_0022007',\n",
       " 'GO_0050673',\n",
       " 'GO_0021953',\n",
       " 'GO_0071908',\n",
       " 'GO_0046579',\n",
       " 'GO_0061795',\n",
       " 'GO_0032043',\n",
       " 'GO_0002014',\n",
       " 'GO_0099040',\n",
       " 'GO_1902527',\n",
       " 'GO_1904800',\n",
       " 'GO_1901017',\n",
       " 'GO_0036333',\n",
       " 'GO_0015862',\n",
       " 'GO_0032053',\n",
       " 'GO_1905795',\n",
       " 'GO_0002685',\n",
       " 'GO_0007338',\n",
       " 'GO_0042938',\n",
       " 'GO_0009266',\n",
       " 'GO_0035964',\n",
       " 'GO_0035088',\n",
       " 'GO_0008156',\n",
       " 'GO_1901143',\n",
       " 'GO_0060975',\n",
       " 'GO_0030490',\n",
       " 'GO_1905564',\n",
       " 'GO_0009650',\n",
       " 'GO_0010481',\n",
       " 'GO_0072169',\n",
       " 'GO_0031581',\n",
       " 'GO_0035397',\n",
       " 'GO_0000028',\n",
       " 'GO_1900006',\n",
       " 'GO_0090188',\n",
       " 'GO_0051250',\n",
       " 'GO_0043363',\n",
       " 'GO_0046321',\n",
       " 'GO_0036294',\n",
       " 'GO_2000464',\n",
       " 'GO_0060753',\n",
       " 'GO_0007167',\n",
       " 'GO_1901799',\n",
       " 'GO_0018158',\n",
       " 'GO_0007605',\n",
       " 'GO_0042415',\n",
       " 'GO_0150037',\n",
       " 'GO_0010711',\n",
       " 'GO_0032802',\n",
       " 'GO_0048014',\n",
       " 'GO_0043516',\n",
       " 'GO_0008543',\n",
       " 'GO_0006958',\n",
       " 'GO_0010800',\n",
       " 'GO_0120181',\n",
       " 'GO_0003272',\n",
       " 'GO_0090030',\n",
       " 'GO_1904231',\n",
       " 'GO_0048149',\n",
       " 'GO_0090085',\n",
       " 'GO_0090341',\n",
       " 'GO_0035793',\n",
       " 'GO_1904754',\n",
       " 'GO_0070207',\n",
       " 'GO_1900027',\n",
       " 'GO_0140627',\n",
       " 'GO_1905342',\n",
       " 'GO_0002366',\n",
       " 'GO_0070757',\n",
       " 'GO_1903463',\n",
       " 'GO_0018026',\n",
       " 'GO_1903919',\n",
       " 'GO_0106022',\n",
       " 'GO_1904694',\n",
       " 'GO_0033080',\n",
       " 'GO_0021615',\n",
       " 'GO_0032365',\n",
       " 'GO_0060595',\n",
       " 'GO_0045066',\n",
       " 'GO_0007059',\n",
       " 'GO_1902473',\n",
       " 'GO_0070278',\n",
       " 'GO_0046903',\n",
       " 'GO_1901894',\n",
       " 'GO_0060579',\n",
       " 'GO_0045977',\n",
       " 'GO_2001258',\n",
       " 'GO_0046296',\n",
       " 'GO_0034242',\n",
       " 'GO_0050770',\n",
       " 'GO_0009258',\n",
       " 'GO_1902074',\n",
       " 'GO_0006624',\n",
       " 'GO_0160054',\n",
       " 'GO_0030195',\n",
       " 'GO_1905450',\n",
       " 'GO_0042355',\n",
       " 'GO_1990432',\n",
       " 'GO_0098629',\n",
       " 'GO_0034139',\n",
       " 'GO_0032976',\n",
       " 'GO_0006677',\n",
       " 'GO_0099502',\n",
       " 'GO_0060255',\n",
       " 'GO_0051966',\n",
       " 'GO_0045617',\n",
       " 'GO_0018937',\n",
       " 'GO_0034439',\n",
       " 'GO_0160078',\n",
       " 'GO_1901329',\n",
       " 'GO_0034087',\n",
       " 'GO_0007199',\n",
       " 'GO_0010623',\n",
       " 'GO_0048668',\n",
       " 'GO_0019725',\n",
       " 'GO_0010693',\n",
       " 'GO_0071909',\n",
       " 'GO_0070444',\n",
       " 'GO_0072255',\n",
       " 'GO_0006271',\n",
       " 'GO_0032480',\n",
       " 'GO_0140744',\n",
       " 'GO_1901985',\n",
       " 'GO_0023041',\n",
       " 'GO_0048562',\n",
       " 'GO_0035022',\n",
       " 'GO_0035050',\n",
       " 'GO_0000961',\n",
       " 'GO_2000452',\n",
       " 'GO_0050674',\n",
       " 'GO_0036146',\n",
       " 'GO_0071217',\n",
       " 'GO_0035278',\n",
       " 'GO_0050666',\n",
       " 'GO_0050678',\n",
       " 'GO_0032495',\n",
       " 'GO_0006695',\n",
       " 'GO_0014910',\n",
       " 'GO_0015707',\n",
       " 'GO_2000969',\n",
       " 'GO_0051638',\n",
       " 'GO_0002270',\n",
       " 'GO_0006032',\n",
       " 'GO_0008361',\n",
       " 'GO_0097698',\n",
       " 'GO_0002021',\n",
       " 'GO_0001867',\n",
       " 'GO_2000121',\n",
       " 'GO_0042904',\n",
       " 'GO_0038060',\n",
       " 'GO_0010643',\n",
       " 'GO_1902775',\n",
       " 'GO_0090266',\n",
       " 'GO_0045793',\n",
       " 'GO_0000435',\n",
       " 'GO_0046092',\n",
       " 'GO_0045666',\n",
       " 'GO_0006072',\n",
       " 'GO_0019079',\n",
       " 'GO_0010499',\n",
       " 'GO_0021780',\n",
       " 'GO_0000019',\n",
       " 'GO_0090070',\n",
       " 'GO_0055088',\n",
       " 'GO_1905223',\n",
       " 'GO_1900235',\n",
       " 'GO_0006477',\n",
       " 'GO_0003221',\n",
       " 'GO_0033605',\n",
       " 'GO_0033574',\n",
       " 'GO_0051683',\n",
       " 'GO_0048745',\n",
       " 'GO_0045860',\n",
       " 'GO_0072148',\n",
       " 'GO_0140928',\n",
       " 'GO_0034983',\n",
       " 'GO_0003222',\n",
       " 'GO_0002737',\n",
       " 'GO_1990049',\n",
       " 'GO_0140745',\n",
       " 'GO_0016358',\n",
       " 'GO_0051490',\n",
       " 'GO_0046835',\n",
       " 'GO_0034332',\n",
       " 'GO_0006422',\n",
       " 'GO_0002130',\n",
       " 'GO_0071895',\n",
       " 'GO_0006287',\n",
       " 'GO_0032836',\n",
       " 'GO_0045751',\n",
       " 'GO_0006579',\n",
       " 'GO_0097084',\n",
       " 'GO_0042745',\n",
       " 'GO_0035710',\n",
       " 'GO_0071765',\n",
       " 'GO_1903909',\n",
       " 'GO_0034651',\n",
       " 'GO_0007057',\n",
       " 'GO_0045947',\n",
       " 'GO_0060407',\n",
       " 'GO_1902425',\n",
       " 'GO_0002194',\n",
       " 'GO_1905070',\n",
       " 'GO_0061323',\n",
       " 'GO_0021557',\n",
       " 'GO_1904491',\n",
       " ...}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_terms = {i.replace(\":\", \"_\") for i in go_terms}\n",
    "go_terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nodes before pruning: 24868\n"
     ]
    }
   ],
   "source": [
    "print(\"Nodes before pruning:\", len(G.nodes)) # 24k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes left 23731\n"
     ]
    }
   ],
   "source": [
    "# prune network\n",
    "def prune_tree(G, nodeset):\n",
    "    nodes_to_check = list(G.nodes)\n",
    "    while nodes_to_check:\n",
    "        node = nodes_to_check.pop()\n",
    "        if G.out_degree(node) == 0 and G.in_degree(node) > 0 and not node in nodeset: # if leaf is not in nodeset\n",
    "            parents = list(G.predecessors(node))\n",
    "            G.remove_node(node)\n",
    "            nodes_to_check.extend(parents)\n",
    "\n",
    "prune_tree(G, go_terms)\n",
    "\n",
    "print(\"Number of nodes left\", len(G.nodes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build full ontology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/tmp.20794777.dgankin/ipykernel_4021084/3296272266.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  snp_df['type'] = 'snp'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>parent</th>\n",
       "      <th>child</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO_0000001</td>\n",
       "      <td>GO_0048311</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO_0000002</td>\n",
       "      <td>GO_0007005</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO_0000012</td>\n",
       "      <td>GO_0006281</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO_0000017</td>\n",
       "      <td>GO_0042946</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO_0000018</td>\n",
       "      <td>GO_0051052</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644889</th>\n",
       "      <td>PASD1</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644890</th>\n",
       "      <td>GABRE</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644891</th>\n",
       "      <td>TREX2</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644892</th>\n",
       "      <td>SRPK3</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644893</th>\n",
       "      <td>IKBKG</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>792962 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            parent        child     type\n",
       "0       GO_0000001   GO_0048311  default\n",
       "1       GO_0000002   GO_0007005  default\n",
       "2       GO_0000012   GO_0006281  default\n",
       "3       GO_0000017   GO_0042946  default\n",
       "4       GO_0000018   GO_0051052  default\n",
       "...            ...          ...      ...\n",
       "644889       PASD1   rs58448897      snp\n",
       "644890       GABRE  rs145754054      snp\n",
       "644891       TREX2    rs5945389      snp\n",
       "644892       SRPK3  rs199915647      snp\n",
       "644893       IKBKG    rs5030868      snp\n",
       "\n",
       "[792962 rows x 3 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert the pruned graph to a dataframe for output\n",
    "pruned_edges = [(u, v) for u, v in G.edges]\n",
    "ontology_df = pd.DataFrame(pruned_edges, columns=['parent', 'child'])\n",
    "ontology_df[\"type\"] = \"default\"\n",
    "ontology_df\n",
    "\n",
    "# append gene to go term df\n",
    "tmp_df = merged_go[[\"index\",\"gene_name\"]].drop_duplicates()\n",
    "tmp_df['index'] = tmp_df['index'].str.replace(\":\",\"_\")\n",
    "tmp_df.columns = ['parent', 'child']\n",
    "tmp_df['type'] = 'gene'\n",
    "\n",
    "# append snp to gene df\n",
    "snp_df = go_bim[[\"gene_name\",\"snp\"]]\n",
    "snp_df.columns = ['parent', 'child']\n",
    "snp_df['type'] = 'snp'\n",
    "\n",
    "# concatenate\n",
    "ontology_df = pd.concat([ontology_df, tmp_df, snp_df])\n",
    "ontology_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nodes in Graph: 683236\n",
      "Edges in Graph: 792962\n",
      "Number of nodes left 683236\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>parent</th>\n",
       "      <th>child</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO_0000001</td>\n",
       "      <td>GO_0048311</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO_0048311</td>\n",
       "      <td>GO_0007005</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO_0048311</td>\n",
       "      <td>GO_0051646</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO_0048311</td>\n",
       "      <td>MEF2A</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO_0048311</td>\n",
       "      <td>HAP1</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       parent       child     type\n",
       "0  GO_0000001  GO_0048311  default\n",
       "1  GO_0048311  GO_0007005  default\n",
       "2  GO_0048311  GO_0051646  default\n",
       "3  GO_0048311       MEF2A     gene\n",
       "4  GO_0048311        HAP1     gene"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a directed graph from the ontology data\n",
    "OG = nx.DiGraph()\n",
    "\n",
    "# Add edges to the graph\n",
    "for index, row in ontology_df.iterrows():\n",
    "    OG.add_edge(row['parent'], row['child'], type=row['type']) # only add gene if in known genes\n",
    "\n",
    "print(\"Nodes in Graph:\", len(OG.nodes))\n",
    "print(\"Edges in Graph:\", len(OG.edges))\n",
    "\n",
    "#print(G.nodes)\n",
    "\n",
    "# Step 2: Iteratively remove nodes that do not lead to a known gene\n",
    "def prune_tree_(G):\n",
    "    nodes_to_check = list(G.nodes)\n",
    "    while nodes_to_check:\n",
    "        node = nodes_to_check.pop()\n",
    "        if G.out_degree(node) == 0 and G.in_degree(node) > 0 and \"GO\" in node:  # Check if it is a hanging node\n",
    "            print(\"yas\")\n",
    "            parents = list(G.predecessors(node))\n",
    "            G.remove_node(node)\n",
    "            nodes_to_check.extend(parents)\n",
    "\n",
    "prune_tree_(OG)\n",
    "\n",
    "print(\"Number of nodes left\", len(OG.nodes))\n",
    "\n",
    "# Convert the pruned graph back to a dataframe for output\n",
    "pruned_edges = [(u, v, OG[u][v]['type']) for u, v in OG.edges]\n",
    "pruned_df = pd.DataFrame(pruned_edges, columns=['parent', 'child', 'type'])\n",
    "\n",
    "# Save the pruned ontology\n",
    "pruned_ontology_file = '../data/go_biological_processes.txt'\n",
    "pruned_df.to_csv(pruned_ontology_file, sep='\\t', index=False)\n",
    "\n",
    "pruned_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
