{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cluster/project/beltrao/gankin/software/miniconda3/envs/vnn/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[EnsemblRelease(release=75, species='homo_sapiens'),\n",
       " EnsemblRelease(release=111, species='homo_sapiens')]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "import swifter\n",
    "\n",
    "from pyensembl import EnsemblRelease\n",
    "from pyensembl.genome import Genome\n",
    "from pyensembl.shell import collect_all_installed_ensembl_releases\n",
    "collect_all_installed_ensembl_releases()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load UKB BIM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/tmp.20794777.dgankin/ipykernel_4025407/1028760957.py:5: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  bim_df = pd.read_csv(bim_file, sep='\\t', header=None)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs28659788</td>\n",
       "      <td>0</td>\n",
       "      <td>723307</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>rs116587930</td>\n",
       "      <td>0</td>\n",
       "      <td>727841</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>rs116720794</td>\n",
       "      <td>0</td>\n",
       "      <td>729632</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>rs3131972</td>\n",
       "      <td>0</td>\n",
       "      <td>752721</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>rs12184325</td>\n",
       "      <td>0</td>\n",
       "      <td>754105</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805421</th>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-92047842</td>\n",
       "      <td>0</td>\n",
       "      <td>16337</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805422</th>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-79443531</td>\n",
       "      <td>0</td>\n",
       "      <td>16356</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805423</th>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-79443532</td>\n",
       "      <td>0</td>\n",
       "      <td>16362</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805424</th>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-89025709</td>\n",
       "      <td>0</td>\n",
       "      <td>16390</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805425</th>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-79381726</td>\n",
       "      <td>0</td>\n",
       "      <td>16391</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>805426 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom            snp  cm     pos a1 a2\n",
       "0          1     rs28659788   0  723307  C  G\n",
       "1          1    rs116587930   0  727841  G  A\n",
       "2          1    rs116720794   0  729632  C  T\n",
       "3          1      rs3131972   0  752721  A  G\n",
       "4          1     rs12184325   0  754105  C  T\n",
       "...      ...            ...  ..     ... .. ..\n",
       "805421    MT  Affx-92047842   0   16337  C  T\n",
       "805422    MT  Affx-79443531   0   16356  T  C\n",
       "805423    MT  Affx-79443532   0   16362  T  C\n",
       "805424    MT  Affx-89025709   0   16390  G  A\n",
       "805425    MT  Affx-79381726   0   16391  G  A\n",
       "\n",
       "[805426 rows x 6 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_dir = \"/cluster/project/beltrao/gankin/vnn/data\"\n",
    "\n",
    "# Read the .bim file to get SNP information\n",
    "bim_file = data_dir + \"/ukb.bim\"\n",
    "bim_df = pd.read_csv(bim_file, sep='\\t', header=None)\n",
    "bim_df.columns = ['chrom', 'snp', 'cm', 'pos', 'a1', 'a2']\n",
    "bim_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Open Targets data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>gene_id_any_distance</th>\n",
       "      <th>gene_id_any</th>\n",
       "      <th>gene_id_prot_coding_distance</th>\n",
       "      <th>gene_id_prot_coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>30794</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>30794</td>\n",
       "      <td>rs1311411458</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 0.177, 'raw': -0.402967}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>34625.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>34625.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>55330</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>55330</td>\n",
       "      <td>rs185215913</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>{'phred': 0.431, 'raw': -0.275244}</td>\n",
       "      <td>{'gnomad_afr': 0.016711590296495958, 'gnomad_a...</td>\n",
       "      <td>10089.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>10089.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>57248</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>57248</td>\n",
       "      <td>rs1249676487</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>{'phred': 7.268, 'raw': 0.304146}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>8171.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>8171.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>66381</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>66381</td>\n",
       "      <td>rs538833530</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 1.389, 'raw': -0.101295}</td>\n",
       "      <td>{'gnomad_afr': 0.15321011673151752, 'gnomad_am...</td>\n",
       "      <td>962.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>962.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>67631</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>67631</td>\n",
       "      <td>rs533896527</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.033, 'raw': -0.632235}</td>\n",
       "      <td>{'gnomad_afr': 0.041812315138478086, 'gnomad_a...</td>\n",
       "      <td>2212.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>2212.0</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72878704</th>\n",
       "      <td>X</td>\n",
       "      <td>156006466</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>155236131</td>\n",
       "      <td>rs1218662062</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 1.118, 'raw': -0.133403}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>8770.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "      <td>8770.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72878705</th>\n",
       "      <td>X</td>\n",
       "      <td>156006493</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>155236158</td>\n",
       "      <td>rs1417497789</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 1.776, 'raw': -0.063666}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0011848341...</td>\n",
       "      <td>8797.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "      <td>8797.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72878706</th>\n",
       "      <td>X</td>\n",
       "      <td>156008488</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>155238153</td>\n",
       "      <td>rs186430584</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 3.876, 'raw': 0.076619}</td>\n",
       "      <td>{'gnomad_afr': 0.030184331797235023, 'gnomad_a...</td>\n",
       "      <td>10792.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "      <td>10792.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72878707</th>\n",
       "      <td>X</td>\n",
       "      <td>156009035</td>\n",
       "      <td>CTG</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>155238700</td>\n",
       "      <td>rs1381721936</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': None, 'raw': None}</td>\n",
       "      <td>{'gnomad_afr': 0.0008291873963515755, 'gnomad_...</td>\n",
       "      <td>11339.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "      <td>11339.0</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72878708</th>\n",
       "      <td>X</td>\n",
       "      <td>156027432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>155257097</td>\n",
       "      <td>rs1041001787</td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>{'phred': 9.819, 'raw': 0.548259}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>6606.0</td>\n",
       "      <td>ENSG00000182484</td>\n",
       "      <td>6606.0</td>\n",
       "      <td>ENSG00000182484</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72878709 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0             1      30794          G          T          1         30794   \n",
       "1             1      55330          G          A          1         55330   \n",
       "2             1      57248          C          T          1         57248   \n",
       "3             1      66381          T          A          1         66381   \n",
       "4             1      67631          G          C          1         67631   \n",
       "...         ...        ...        ...        ...        ...           ...   \n",
       "72878704      X  156006466          G          C          X     155236131   \n",
       "72878705      X  156006493          C          T          X     155236158   \n",
       "72878706      X  156008488          C          T          X     155238153   \n",
       "72878707      X  156009035        CTG          C          X     155238700   \n",
       "72878708      X  156027432          C          T          X     155257097   \n",
       "\n",
       "                 rs_id             most_severe_consequence  \\\n",
       "0         rs1311411458                      intron_variant   \n",
       "1          rs185215913             downstream_gene_variant   \n",
       "2         rs1249676487             downstream_gene_variant   \n",
       "3          rs538833530               upstream_gene_variant   \n",
       "4          rs533896527               upstream_gene_variant   \n",
       "...                ...                                 ...   \n",
       "72878704  rs1218662062                      intron_variant   \n",
       "72878705  rs1417497789                      intron_variant   \n",
       "72878706   rs186430584                      intron_variant   \n",
       "72878707  rs1381721936                      intron_variant   \n",
       "72878708  rs1041001787  non_coding_transcript_exon_variant   \n",
       "\n",
       "                                        cadd  \\\n",
       "0         {'phred': 0.177, 'raw': -0.402967}   \n",
       "1         {'phred': 0.431, 'raw': -0.275244}   \n",
       "2          {'phred': 7.268, 'raw': 0.304146}   \n",
       "3         {'phred': 1.389, 'raw': -0.101295}   \n",
       "4         {'phred': 0.033, 'raw': -0.632235}   \n",
       "...                                      ...   \n",
       "72878704  {'phred': 1.118, 'raw': -0.133403}   \n",
       "72878705  {'phred': 1.776, 'raw': -0.063666}   \n",
       "72878706   {'phred': 3.876, 'raw': 0.076619}   \n",
       "72878707        {'phred': None, 'raw': None}   \n",
       "72878708   {'phred': 9.819, 'raw': 0.548259}   \n",
       "\n",
       "                                                         af  \\\n",
       "0         {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...   \n",
       "1         {'gnomad_afr': 0.016711590296495958, 'gnomad_a...   \n",
       "2         {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...   \n",
       "3         {'gnomad_afr': 0.15321011673151752, 'gnomad_am...   \n",
       "4         {'gnomad_afr': 0.041812315138478086, 'gnomad_a...   \n",
       "...                                                     ...   \n",
       "72878704  {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...   \n",
       "72878705  {'gnomad_afr': 0.0, 'gnomad_amr': 0.0011848341...   \n",
       "72878706  {'gnomad_afr': 0.030184331797235023, 'gnomad_a...   \n",
       "72878707  {'gnomad_afr': 0.0008291873963515755, 'gnomad_...   \n",
       "72878708  {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...   \n",
       "\n",
       "          gene_id_any_distance      gene_id_any  gene_id_prot_coding_distance  \\\n",
       "0                      34625.0  ENSG00000186092                       34625.0   \n",
       "1                      10089.0  ENSG00000186092                       10089.0   \n",
       "2                       8171.0  ENSG00000186092                        8171.0   \n",
       "3                        962.0  ENSG00000186092                         962.0   \n",
       "4                       2212.0  ENSG00000186092                        2212.0   \n",
       "...                        ...              ...                           ...   \n",
       "72878704                8770.0  ENSG00000124334                        8770.0   \n",
       "72878705                8797.0  ENSG00000124334                        8797.0   \n",
       "72878706               10792.0  ENSG00000124334                       10792.0   \n",
       "72878707               11339.0  ENSG00000124334                       11339.0   \n",
       "72878708                6606.0  ENSG00000182484                        6606.0   \n",
       "\n",
       "         gene_id_prot_coding  \n",
       "0            ENSG00000186092  \n",
       "1            ENSG00000186092  \n",
       "2            ENSG00000186092  \n",
       "3            ENSG00000186092  \n",
       "4            ENSG00000186092  \n",
       "...                      ...  \n",
       "72878704     ENSG00000124334  \n",
       "72878705     ENSG00000124334  \n",
       "72878706     ENSG00000124334  \n",
       "72878707     ENSG00000124334  \n",
       "72878708     ENSG00000182484  \n",
       "\n",
       "[72878709 rows x 14 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_index_parquet = data_dir + \"/complete_variant_index.parquet\"\n",
    "variant_index = pd.read_parquet(variant_index_parquet) # takes a few minutes\n",
    "variant_index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge with UKB SNP data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>gene_id_any_distance</th>\n",
       "      <th>gene_id_any</th>\n",
       "      <th>gene_id_prot_coding_distance</th>\n",
       "      <th>gene_id_prot_coding</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1169204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1104584</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.61, 'raw': -0.223786}</td>\n",
       "      <td>{'gnomad_afr': 0.1370131366674349, 'gnomad_amr...</td>\n",
       "      <td>4676.0</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>4676.0</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>1</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>0</td>\n",
       "      <td>1104584</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>12098.0</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>12098.0</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>1</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>0</td>\n",
       "      <td>1969752</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2194104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>2125543</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 4.241, 'raw': 0.098415}</td>\n",
       "      <td>{'gnomad_afr': 0.00034498620055197794, 'gnomad...</td>\n",
       "      <td>18616.0</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>18616.0</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>0</td>\n",
       "      <td>2125543</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2592668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>2524107</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 23.9, 'raw': 3.118097}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>6177.0</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>6177.0</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>0</td>\n",
       "      <td>2524107</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2828773</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2745338</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.058, 'raw': -0.555543}</td>\n",
       "      <td>{'gnomad_afr': 0.0060821666284140465, 'gnomad_...</td>\n",
       "      <td>27080.0</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>27080.0</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>0</td>\n",
       "      <td>2745338</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747754</th>\n",
       "      <td>X</td>\n",
       "      <td>152014563</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>151183035</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 2.456, 'raw': -0.011208}</td>\n",
       "      <td>{'gnomad_afr': 0.009298393913778529, 'gnomad_a...</td>\n",
       "      <td>39883.0</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>39883.0</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>X</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>0</td>\n",
       "      <td>151183035</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747755</th>\n",
       "      <td>X</td>\n",
       "      <td>153459591</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>152725049</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 2.968, 'raw': 0.022059}</td>\n",
       "      <td>{'gnomad_afr': 0.0032312925170068026, 'gnomad_...</td>\n",
       "      <td>10996.0</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>10996.0</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>0</td>\n",
       "      <td>152725049</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747756</th>\n",
       "      <td>X</td>\n",
       "      <td>153781540</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>153046995</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 22.5, 'raw': 2.480113}</td>\n",
       "      <td>{'gnomad_afr': 0.00017018379850238256, 'gnomad...</td>\n",
       "      <td>5128.0</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>5128.0</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>X</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>0</td>\n",
       "      <td>153046995</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747757</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>6780.0</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>6780.0</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>0</td>\n",
       "      <td>153762634</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747758</th>\n",
       "      <td>X</td>\n",
       "      <td>155780194</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>X</td>\n",
       "      <td>155009856</td>\n",
       "      <td>rs700449</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>{'phred': 5.474, 'raw': 0.175592}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0011876484...</td>\n",
       "      <td>61096.0</td>\n",
       "      <td>ENSG00000185973</td>\n",
       "      <td>61096.0</td>\n",
       "      <td>ENSG00000185973</td>\n",
       "      <td>XY</td>\n",
       "      <td>rs700449</td>\n",
       "      <td>0</td>\n",
       "      <td>155009856</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>747759 rows Ã— 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0           1    1169204          T          C          1       1104584   \n",
       "1           1    2038313          C          T          1       1969752   \n",
       "2           1    2194104          A          G          1       2125543   \n",
       "3           1    2592668          C          T          1       2524107   \n",
       "4           1    2828773          C          A          1       2745338   \n",
       "...       ...        ...        ...        ...        ...           ...   \n",
       "747754      X  152014563          G          A          X     151183035   \n",
       "747755      X  153459591          C          T          X     152725049   \n",
       "747756      X  153781540          A          C          X     153046995   \n",
       "747757      X  154534419          G          A          X     153762634   \n",
       "747758      X  155780194          T          G          X     155009856   \n",
       "\n",
       "              rs_id most_severe_consequence  \\\n",
       "0       rs112695918   upstream_gene_variant   \n",
       "1        rs74372554      intergenic_variant   \n",
       "2       rs146607591        missense_variant   \n",
       "3       rs146100312        missense_variant   \n",
       "4       rs116076619      intergenic_variant   \n",
       "...             ...                     ...   \n",
       "747754  rs145754054      intergenic_variant   \n",
       "747755    rs5945389          intron_variant   \n",
       "747756  rs199915647        missense_variant   \n",
       "747757    rs5030868        missense_variant   \n",
       "747758     rs700449     3_prime_UTR_variant   \n",
       "\n",
       "                                      cadd  \\\n",
       "0        {'phred': 0.61, 'raw': -0.223786}   \n",
       "1       {'phred': 0.309, 'raw': -0.324059}   \n",
       "2        {'phred': 4.241, 'raw': 0.098415}   \n",
       "3         {'phred': 23.9, 'raw': 3.118097}   \n",
       "4       {'phred': 0.058, 'raw': -0.555543}   \n",
       "...                                    ...   \n",
       "747754  {'phred': 2.456, 'raw': -0.011208}   \n",
       "747755   {'phred': 2.968, 'raw': 0.022059}   \n",
       "747756    {'phred': 22.5, 'raw': 2.480113}   \n",
       "747757    {'phred': 24.7, 'raw': 3.374046}   \n",
       "747758   {'phred': 5.474, 'raw': 0.175592}   \n",
       "\n",
       "                                                       af  \\\n",
       "0       {'gnomad_afr': 0.1370131366674349, 'gnomad_amr...   \n",
       "1       {'gnomad_afr': 0.01078971533516988, 'gnomad_am...   \n",
       "2       {'gnomad_afr': 0.00034498620055197794, 'gnomad...   \n",
       "3       {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...   \n",
       "4       {'gnomad_afr': 0.0060821666284140465, 'gnomad_...   \n",
       "...                                                   ...   \n",
       "747754  {'gnomad_afr': 0.009298393913778529, 'gnomad_a...   \n",
       "747755  {'gnomad_afr': 0.0032312925170068026, 'gnomad_...   \n",
       "747756  {'gnomad_afr': 0.00017018379850238256, 'gnomad...   \n",
       "747757  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...   \n",
       "747758  {'gnomad_afr': 0.0, 'gnomad_amr': 0.0011876484...   \n",
       "\n",
       "        gene_id_any_distance      gene_id_any  gene_id_prot_coding_distance  \\\n",
       "0                     4676.0  ENSG00000162571                        4676.0   \n",
       "1                    12098.0  ENSG00000067606                       12098.0   \n",
       "2                    18616.0  ENSG00000162585                       18616.0   \n",
       "3                     6177.0  ENSG00000157870                        6177.0   \n",
       "4                    27080.0  ENSG00000215912                       27080.0   \n",
       "...                      ...              ...                           ...   \n",
       "747754               39883.0  ENSG00000102287                       39883.0   \n",
       "747755               10996.0  ENSG00000183479                       10996.0   \n",
       "747756                5128.0  ENSG00000184343                        5128.0   \n",
       "747757                6780.0  ENSG00000269335                        6780.0   \n",
       "747758               61096.0  ENSG00000185973                       61096.0   \n",
       "\n",
       "       gene_id_prot_coding chrom          snp  cm        pos a1 a2  \n",
       "0          ENSG00000162571     1  rs112695918   0    1104584  T  C  \n",
       "1          ENSG00000067606     1   rs74372554   0    1969752  C  T  \n",
       "2          ENSG00000162585     1  rs146607591   0    2125543  A  G  \n",
       "3          ENSG00000157870     1  rs146100312   0    2524107  C  T  \n",
       "4          ENSG00000215912     1  rs116076619   0    2745338  C  A  \n",
       "...                    ...   ...          ...  ..        ... .. ..  \n",
       "747754     ENSG00000102287     X  rs145754054   0  151183035  G  A  \n",
       "747755     ENSG00000183479     X    rs5945389   0  152725049  C  T  \n",
       "747756     ENSG00000184343     X  rs199915647   0  153046995  A  C  \n",
       "747757     ENSG00000269335     X    rs5030868   0  153762634  G  A  \n",
       "747758     ENSG00000185973    XY     rs700449   0  155009856  T  C  \n",
       "\n",
       "[747759 rows x 20 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge bim with variant index\n",
    "ot_bim = variant_index.merge(bim_df, left_on='rs_id', right_on='snp', how='inner')\n",
    "ot_bim"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene ids to gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_data = EnsemblRelease(species='human', release=75)\n",
    "def get_gene_name(ensembl_id):\n",
    "    try:\n",
    "        return human_data.gene_name_of_gene_id(ensembl_id)\n",
    "    except:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Pandas Apply: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 18817/18817 [10:32<00:00, 29.76it/s]\n"
     ]
    }
   ],
   "source": [
    "name_map = pd.DataFrame()\n",
    "name_map[\"gene_id\"] = ot_bim.gene_id_prot_coding.unique()\n",
    "name_map[\"gene_name\"]  = name_map[\"gene_id\"].swifter.apply(lambda x: get_gene_name(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene Id not mappable:  298\n",
      "Unique gene names:  18510\n"
     ]
    }
   ],
   "source": [
    "# count None in gene names\n",
    "print(\"Gene Id not mappable: \", name_map.gene_name.isnull().sum())\n",
    "print(\"Unique gene names: \", name_map.gene_name.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "ot_bim = pd.merge(ot_bim, name_map, left_on='gene_id_prot_coding', right_on='gene_id', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_id_prot_coding_distance</th>\n",
       "      <th>gene_id_prot_coding</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1169204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1104584</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.61, 'raw': -0.223786}</td>\n",
       "      <td>{'gnomad_afr': 0.1370131366674349, 'gnomad_amr...</td>\n",
       "      <td>...</td>\n",
       "      <td>4676.0</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>1</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>0</td>\n",
       "      <td>1104584</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>TTLL10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>12098.0</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>1</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>0</td>\n",
       "      <td>1969752</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2194104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>2125543</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 4.241, 'raw': 0.098415}</td>\n",
       "      <td>{'gnomad_afr': 0.00034498620055197794, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>18616.0</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>0</td>\n",
       "      <td>2125543</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>C1orf86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2592668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>2524107</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 23.9, 'raw': 3.118097}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>6177.0</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>0</td>\n",
       "      <td>2524107</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>FAM213B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2828773</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2745338</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.058, 'raw': -0.555543}</td>\n",
       "      <td>{'gnomad_afr': 0.0060821666284140465, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>27080.0</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>0</td>\n",
       "      <td>2745338</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>TTC34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747754</th>\n",
       "      <td>X</td>\n",
       "      <td>152014563</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>151183035</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 2.456, 'raw': -0.011208}</td>\n",
       "      <td>{'gnomad_afr': 0.009298393913778529, 'gnomad_a...</td>\n",
       "      <td>...</td>\n",
       "      <td>39883.0</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>X</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>0</td>\n",
       "      <td>151183035</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>GABRE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747755</th>\n",
       "      <td>X</td>\n",
       "      <td>153459591</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>152725049</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 2.968, 'raw': 0.022059}</td>\n",
       "      <td>{'gnomad_afr': 0.0032312925170068026, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>10996.0</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>0</td>\n",
       "      <td>152725049</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>TREX2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747756</th>\n",
       "      <td>X</td>\n",
       "      <td>153781540</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>153046995</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 22.5, 'raw': 2.480113}</td>\n",
       "      <td>{'gnomad_afr': 0.00017018379850238256, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>5128.0</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>X</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>0</td>\n",
       "      <td>153046995</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>SRPK3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747757</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>6780.0</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>0</td>\n",
       "      <td>153762634</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747758</th>\n",
       "      <td>X</td>\n",
       "      <td>155780194</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>X</td>\n",
       "      <td>155009856</td>\n",
       "      <td>rs700449</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>{'phred': 5.474, 'raw': 0.175592}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0011876484...</td>\n",
       "      <td>...</td>\n",
       "      <td>61096.0</td>\n",
       "      <td>ENSG00000185973</td>\n",
       "      <td>XY</td>\n",
       "      <td>rs700449</td>\n",
       "      <td>0</td>\n",
       "      <td>155009856</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000185973</td>\n",
       "      <td>TMLHE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>741340 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0           1    1169204          T          C          1       1104584   \n",
       "1           1    2038313          C          T          1       1969752   \n",
       "2           1    2194104          A          G          1       2125543   \n",
       "3           1    2592668          C          T          1       2524107   \n",
       "4           1    2828773          C          A          1       2745338   \n",
       "...       ...        ...        ...        ...        ...           ...   \n",
       "747754      X  152014563          G          A          X     151183035   \n",
       "747755      X  153459591          C          T          X     152725049   \n",
       "747756      X  153781540          A          C          X     153046995   \n",
       "747757      X  154534419          G          A          X     153762634   \n",
       "747758      X  155780194          T          G          X     155009856   \n",
       "\n",
       "              rs_id most_severe_consequence  \\\n",
       "0       rs112695918   upstream_gene_variant   \n",
       "1        rs74372554      intergenic_variant   \n",
       "2       rs146607591        missense_variant   \n",
       "3       rs146100312        missense_variant   \n",
       "4       rs116076619      intergenic_variant   \n",
       "...             ...                     ...   \n",
       "747754  rs145754054      intergenic_variant   \n",
       "747755    rs5945389          intron_variant   \n",
       "747756  rs199915647        missense_variant   \n",
       "747757    rs5030868        missense_variant   \n",
       "747758     rs700449     3_prime_UTR_variant   \n",
       "\n",
       "                                      cadd  \\\n",
       "0        {'phred': 0.61, 'raw': -0.223786}   \n",
       "1       {'phred': 0.309, 'raw': -0.324059}   \n",
       "2        {'phred': 4.241, 'raw': 0.098415}   \n",
       "3         {'phred': 23.9, 'raw': 3.118097}   \n",
       "4       {'phred': 0.058, 'raw': -0.555543}   \n",
       "...                                    ...   \n",
       "747754  {'phred': 2.456, 'raw': -0.011208}   \n",
       "747755   {'phred': 2.968, 'raw': 0.022059}   \n",
       "747756    {'phred': 22.5, 'raw': 2.480113}   \n",
       "747757    {'phred': 24.7, 'raw': 3.374046}   \n",
       "747758   {'phred': 5.474, 'raw': 0.175592}   \n",
       "\n",
       "                                                       af  ...  \\\n",
       "0       {'gnomad_afr': 0.1370131366674349, 'gnomad_amr...  ...   \n",
       "1       {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   \n",
       "2       {'gnomad_afr': 0.00034498620055197794, 'gnomad...  ...   \n",
       "3       {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...  ...   \n",
       "4       {'gnomad_afr': 0.0060821666284140465, 'gnomad_...  ...   \n",
       "...                                                   ...  ...   \n",
       "747754  {'gnomad_afr': 0.009298393913778529, 'gnomad_a...  ...   \n",
       "747755  {'gnomad_afr': 0.0032312925170068026, 'gnomad_...  ...   \n",
       "747756  {'gnomad_afr': 0.00017018379850238256, 'gnomad...  ...   \n",
       "747757  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   \n",
       "747758  {'gnomad_afr': 0.0, 'gnomad_amr': 0.0011876484...  ...   \n",
       "\n",
       "        gene_id_prot_coding_distance gene_id_prot_coding  chrom          snp  \\\n",
       "0                             4676.0     ENSG00000162571      1  rs112695918   \n",
       "1                            12098.0     ENSG00000067606      1   rs74372554   \n",
       "2                            18616.0     ENSG00000162585      1  rs146607591   \n",
       "3                             6177.0     ENSG00000157870      1  rs146100312   \n",
       "4                            27080.0     ENSG00000215912      1  rs116076619   \n",
       "...                              ...                 ...    ...          ...   \n",
       "747754                       39883.0     ENSG00000102287      X  rs145754054   \n",
       "747755                       10996.0     ENSG00000183479      X    rs5945389   \n",
       "747756                        5128.0     ENSG00000184343      X  rs199915647   \n",
       "747757                        6780.0     ENSG00000269335      X    rs5030868   \n",
       "747758                       61096.0     ENSG00000185973     XY     rs700449   \n",
       "\n",
       "       cm        pos  a1  a2          gene_id gene_name  \n",
       "0       0    1104584   T   C  ENSG00000162571    TTLL10  \n",
       "1       0    1969752   C   T  ENSG00000067606     PRKCZ  \n",
       "2       0    2125543   A   G  ENSG00000162585   C1orf86  \n",
       "3       0    2524107   C   T  ENSG00000157870   FAM213B  \n",
       "4       0    2745338   C   A  ENSG00000215912     TTC34  \n",
       "...    ..        ...  ..  ..              ...       ...  \n",
       "747754  0  151183035   G   A  ENSG00000102287     GABRE  \n",
       "747755  0  152725049   C   T  ENSG00000183479     TREX2  \n",
       "747756  0  153046995   A   C  ENSG00000184343     SRPK3  \n",
       "747757  0  153762634   G   A  ENSG00000269335     IKBKG  \n",
       "747758  0  155009856   T   C  ENSG00000185973     TMLHE  \n",
       "\n",
       "[741340 rows x 22 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove all None gene names\n",
    "ot_bim = ot_bim[ot_bim.gene_name.notnull()]\n",
    "ot_bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "731084"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ot_bim.rs_id.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_id_prot_coding_distance</th>\n",
       "      <th>gene_id_prot_coding</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1169204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1104584</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.61, 'raw': -0.223786}</td>\n",
       "      <td>{'gnomad_afr': 0.1370131366674349, 'gnomad_amr...</td>\n",
       "      <td>...</td>\n",
       "      <td>4676.0</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>1</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>0</td>\n",
       "      <td>1104584</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>TTLL10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>12098.0</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>1</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>0</td>\n",
       "      <td>1969752</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2194104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>2125543</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 4.241, 'raw': 0.098415}</td>\n",
       "      <td>{'gnomad_afr': 0.00034498620055197794, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>18616.0</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>0</td>\n",
       "      <td>2125543</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>C1orf86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2592668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>2524107</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 23.9, 'raw': 3.118097}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>6177.0</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>0</td>\n",
       "      <td>2524107</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>FAM213B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2828773</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2745338</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.058, 'raw': -0.555543}</td>\n",
       "      <td>{'gnomad_afr': 0.0060821666284140465, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>27080.0</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>0</td>\n",
       "      <td>2745338</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>TTC34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747753</th>\n",
       "      <td>X</td>\n",
       "      <td>151609900</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>X</td>\n",
       "      <td>150778372</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 3.523, 'raw': 0.055668}</td>\n",
       "      <td>{'gnomad_afr': 0.11482740118046861, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>46225.0</td>\n",
       "      <td>ENSG00000166049</td>\n",
       "      <td>X</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>0</td>\n",
       "      <td>150778372</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000166049</td>\n",
       "      <td>PASD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747754</th>\n",
       "      <td>X</td>\n",
       "      <td>152014563</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>151183035</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 2.456, 'raw': -0.011208}</td>\n",
       "      <td>{'gnomad_afr': 0.009298393913778529, 'gnomad_a...</td>\n",
       "      <td>...</td>\n",
       "      <td>39883.0</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>X</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>0</td>\n",
       "      <td>151183035</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>GABRE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747755</th>\n",
       "      <td>X</td>\n",
       "      <td>153459591</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>152725049</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 2.968, 'raw': 0.022059}</td>\n",
       "      <td>{'gnomad_afr': 0.0032312925170068026, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>10996.0</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>0</td>\n",
       "      <td>152725049</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>TREX2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747756</th>\n",
       "      <td>X</td>\n",
       "      <td>153781540</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>153046995</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 22.5, 'raw': 2.480113}</td>\n",
       "      <td>{'gnomad_afr': 0.00017018379850238256, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>5128.0</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>X</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>0</td>\n",
       "      <td>153046995</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>SRPK3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747757</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>6780.0</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>0</td>\n",
       "      <td>153762634</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>731084 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0           1    1169204          T          C          1       1104584   \n",
       "1           1    2038313          C          T          1       1969752   \n",
       "2           1    2194104          A          G          1       2125543   \n",
       "3           1    2592668          C          T          1       2524107   \n",
       "4           1    2828773          C          A          1       2745338   \n",
       "...       ...        ...        ...        ...        ...           ...   \n",
       "747753      X  151609900          A          G          X     150778372   \n",
       "747754      X  152014563          G          A          X     151183035   \n",
       "747755      X  153459591          C          T          X     152725049   \n",
       "747756      X  153781540          A          C          X     153046995   \n",
       "747757      X  154534419          G          A          X     153762634   \n",
       "\n",
       "              rs_id most_severe_consequence  \\\n",
       "0       rs112695918   upstream_gene_variant   \n",
       "1        rs74372554      intergenic_variant   \n",
       "2       rs146607591        missense_variant   \n",
       "3       rs146100312        missense_variant   \n",
       "4       rs116076619      intergenic_variant   \n",
       "...             ...                     ...   \n",
       "747753   rs58448897          intron_variant   \n",
       "747754  rs145754054      intergenic_variant   \n",
       "747755    rs5945389          intron_variant   \n",
       "747756  rs199915647        missense_variant   \n",
       "747757    rs5030868        missense_variant   \n",
       "\n",
       "                                      cadd  \\\n",
       "0        {'phred': 0.61, 'raw': -0.223786}   \n",
       "1       {'phred': 0.309, 'raw': -0.324059}   \n",
       "2        {'phred': 4.241, 'raw': 0.098415}   \n",
       "3         {'phred': 23.9, 'raw': 3.118097}   \n",
       "4       {'phred': 0.058, 'raw': -0.555543}   \n",
       "...                                    ...   \n",
       "747753   {'phred': 3.523, 'raw': 0.055668}   \n",
       "747754  {'phred': 2.456, 'raw': -0.011208}   \n",
       "747755   {'phred': 2.968, 'raw': 0.022059}   \n",
       "747756    {'phred': 22.5, 'raw': 2.480113}   \n",
       "747757    {'phred': 24.7, 'raw': 3.374046}   \n",
       "\n",
       "                                                       af  ...  \\\n",
       "0       {'gnomad_afr': 0.1370131366674349, 'gnomad_amr...  ...   \n",
       "1       {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   \n",
       "2       {'gnomad_afr': 0.00034498620055197794, 'gnomad...  ...   \n",
       "3       {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...  ...   \n",
       "4       {'gnomad_afr': 0.0060821666284140465, 'gnomad_...  ...   \n",
       "...                                                   ...  ...   \n",
       "747753  {'gnomad_afr': 0.11482740118046861, 'gnomad_am...  ...   \n",
       "747754  {'gnomad_afr': 0.009298393913778529, 'gnomad_a...  ...   \n",
       "747755  {'gnomad_afr': 0.0032312925170068026, 'gnomad_...  ...   \n",
       "747756  {'gnomad_afr': 0.00017018379850238256, 'gnomad...  ...   \n",
       "747757  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   \n",
       "\n",
       "        gene_id_prot_coding_distance gene_id_prot_coding  chrom          snp  \\\n",
       "0                             4676.0     ENSG00000162571      1  rs112695918   \n",
       "1                            12098.0     ENSG00000067606      1   rs74372554   \n",
       "2                            18616.0     ENSG00000162585      1  rs146607591   \n",
       "3                             6177.0     ENSG00000157870      1  rs146100312   \n",
       "4                            27080.0     ENSG00000215912      1  rs116076619   \n",
       "...                              ...                 ...    ...          ...   \n",
       "747753                       46225.0     ENSG00000166049      X   rs58448897   \n",
       "747754                       39883.0     ENSG00000102287      X  rs145754054   \n",
       "747755                       10996.0     ENSG00000183479      X    rs5945389   \n",
       "747756                        5128.0     ENSG00000184343      X  rs199915647   \n",
       "747757                        6780.0     ENSG00000269335      X    rs5030868   \n",
       "\n",
       "       cm        pos  a1  a2          gene_id gene_name  \n",
       "0       0    1104584   T   C  ENSG00000162571    TTLL10  \n",
       "1       0    1969752   C   T  ENSG00000067606     PRKCZ  \n",
       "2       0    2125543   A   G  ENSG00000162585   C1orf86  \n",
       "3       0    2524107   C   T  ENSG00000157870   FAM213B  \n",
       "4       0    2745338   C   A  ENSG00000215912     TTC34  \n",
       "...    ..        ...  ..  ..              ...       ...  \n",
       "747753  0  150778372   A   G  ENSG00000166049     PASD1  \n",
       "747754  0  151183035   G   A  ENSG00000102287     GABRE  \n",
       "747755  0  152725049   C   T  ENSG00000183479     TREX2  \n",
       "747756  0  153046995   A   C  ENSG00000184343     SRPK3  \n",
       "747757  0  153762634   G   A  ENSG00000269335     IKBKG  \n",
       "\n",
       "[731084 rows x 22 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ot_bim = ot_bim.drop_duplicates(subset=[\"rs_id\", \"gene_name\"])\n",
    "# rename rs_id col to snp\n",
    "ot_bim = ot_bim.rename({\"rs_id\":\"snp\"})\n",
    "ot_bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_bed_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>rs28659788</td>\n",
       "      <td>0</td>\n",
       "      <td>723307</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116587930</td>\n",
       "      <td>0</td>\n",
       "      <td>727841</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116720794</td>\n",
       "      <td>0</td>\n",
       "      <td>729632</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>rs3131972</td>\n",
       "      <td>0</td>\n",
       "      <td>752721</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>rs12184325</td>\n",
       "      <td>0</td>\n",
       "      <td>754105</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805421</th>\n",
       "      <td>805421</td>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-92047842</td>\n",
       "      <td>0</td>\n",
       "      <td>16337</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805422</th>\n",
       "      <td>805422</td>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-79443531</td>\n",
       "      <td>0</td>\n",
       "      <td>16356</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805423</th>\n",
       "      <td>805423</td>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-79443532</td>\n",
       "      <td>0</td>\n",
       "      <td>16362</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805424</th>\n",
       "      <td>805424</td>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-89025709</td>\n",
       "      <td>0</td>\n",
       "      <td>16390</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805425</th>\n",
       "      <td>805425</td>\n",
       "      <td>MT</td>\n",
       "      <td>Affx-79381726</td>\n",
       "      <td>0</td>\n",
       "      <td>16391</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>805426 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        snp_bed_id chrom            snp  cm     pos a1 a2\n",
       "0                0     1     rs28659788   0  723307  C  G\n",
       "1                1     1    rs116587930   0  727841  G  A\n",
       "2                2     1    rs116720794   0  729632  C  T\n",
       "3                3     1      rs3131972   0  752721  A  G\n",
       "4                4     1     rs12184325   0  754105  C  T\n",
       "...            ...   ...            ...  ..     ... .. ..\n",
       "805421      805421    MT  Affx-92047842   0   16337  C  T\n",
       "805422      805422    MT  Affx-79443531   0   16356  T  C\n",
       "805423      805423    MT  Affx-79443532   0   16362  T  C\n",
       "805424      805424    MT  Affx-89025709   0   16390  G  A\n",
       "805425      805425    MT  Affx-79381726   0   16391  G  A\n",
       "\n",
       "[805426 rows x 7 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get snp_bed_id from bim_df \n",
    "bim_df = bim_df.reset_index()\n",
    "#bim_df.columns = ['snp_bed_id', 'chrom', 'snp', 'cm', 'pos', 'a1', 'a2']\n",
    "bim_df.columns = ['snp_bed_id', 'chrom', 'snp', 'cm', 'pos', 'a1', 'a2']\n",
    "bim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "ot_bim = ot_bim.merge(bim_df[[\"snp\", \"snp_bed_id\"]], left_on=\"snp\", right_on=\"snp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_id_prot_coding</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>snp_bed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1169204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1104584</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.61, 'raw': -0.223786}</td>\n",
       "      <td>{'gnomad_afr': 0.1370131366674349, 'gnomad_amr...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>1</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>0</td>\n",
       "      <td>1104584</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>TTLL10</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>1</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>0</td>\n",
       "      <td>1969752</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2194104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>2125543</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 4.241, 'raw': 0.098415}</td>\n",
       "      <td>{'gnomad_afr': 0.00034498620055197794, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>0</td>\n",
       "      <td>2125543</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>C1orf86</td>\n",
       "      <td>648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2592668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>2524107</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 23.9, 'raw': 3.118097}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>0</td>\n",
       "      <td>2524107</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>FAM213B</td>\n",
       "      <td>838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2828773</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>2745338</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.058, 'raw': -0.555543}</td>\n",
       "      <td>{'gnomad_afr': 0.0060821666284140465, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116076619</td>\n",
       "      <td>0</td>\n",
       "      <td>2745338</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000215912</td>\n",
       "      <td>TTC34</td>\n",
       "      <td>889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731079</th>\n",
       "      <td>X</td>\n",
       "      <td>151609900</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>X</td>\n",
       "      <td>150778372</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 3.523, 'raw': 0.055668}</td>\n",
       "      <td>{'gnomad_afr': 0.11482740118046861, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000166049</td>\n",
       "      <td>X</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>0</td>\n",
       "      <td>150778372</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000166049</td>\n",
       "      <td>PASD1</td>\n",
       "      <td>802181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731080</th>\n",
       "      <td>X</td>\n",
       "      <td>152014563</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>151183035</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 2.456, 'raw': -0.011208}</td>\n",
       "      <td>{'gnomad_afr': 0.009298393913778529, 'gnomad_a...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>X</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>0</td>\n",
       "      <td>151183035</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>GABRE</td>\n",
       "      <td>802336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731081</th>\n",
       "      <td>X</td>\n",
       "      <td>153459591</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>152725049</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 2.968, 'raw': 0.022059}</td>\n",
       "      <td>{'gnomad_afr': 0.0032312925170068026, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>0</td>\n",
       "      <td>152725049</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>TREX2</td>\n",
       "      <td>802547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731082</th>\n",
       "      <td>X</td>\n",
       "      <td>153781540</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>153046995</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 22.5, 'raw': 2.480113}</td>\n",
       "      <td>{'gnomad_afr': 0.00017018379850238256, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>X</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>0</td>\n",
       "      <td>153046995</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>SRPK3</td>\n",
       "      <td>802648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731083</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>0</td>\n",
       "      <td>153762634</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>731084 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0           1    1169204          T          C          1       1104584   \n",
       "1           1    2038313          C          T          1       1969752   \n",
       "2           1    2194104          A          G          1       2125543   \n",
       "3           1    2592668          C          T          1       2524107   \n",
       "4           1    2828773          C          A          1       2745338   \n",
       "...       ...        ...        ...        ...        ...           ...   \n",
       "731079      X  151609900          A          G          X     150778372   \n",
       "731080      X  152014563          G          A          X     151183035   \n",
       "731081      X  153459591          C          T          X     152725049   \n",
       "731082      X  153781540          A          C          X     153046995   \n",
       "731083      X  154534419          G          A          X     153762634   \n",
       "\n",
       "              rs_id most_severe_consequence  \\\n",
       "0       rs112695918   upstream_gene_variant   \n",
       "1        rs74372554      intergenic_variant   \n",
       "2       rs146607591        missense_variant   \n",
       "3       rs146100312        missense_variant   \n",
       "4       rs116076619      intergenic_variant   \n",
       "...             ...                     ...   \n",
       "731079   rs58448897          intron_variant   \n",
       "731080  rs145754054      intergenic_variant   \n",
       "731081    rs5945389          intron_variant   \n",
       "731082  rs199915647        missense_variant   \n",
       "731083    rs5030868        missense_variant   \n",
       "\n",
       "                                      cadd  \\\n",
       "0        {'phred': 0.61, 'raw': -0.223786}   \n",
       "1       {'phred': 0.309, 'raw': -0.324059}   \n",
       "2        {'phred': 4.241, 'raw': 0.098415}   \n",
       "3         {'phred': 23.9, 'raw': 3.118097}   \n",
       "4       {'phred': 0.058, 'raw': -0.555543}   \n",
       "...                                    ...   \n",
       "731079   {'phred': 3.523, 'raw': 0.055668}   \n",
       "731080  {'phred': 2.456, 'raw': -0.011208}   \n",
       "731081   {'phred': 2.968, 'raw': 0.022059}   \n",
       "731082    {'phred': 22.5, 'raw': 2.480113}   \n",
       "731083    {'phred': 24.7, 'raw': 3.374046}   \n",
       "\n",
       "                                                       af  ...  \\\n",
       "0       {'gnomad_afr': 0.1370131366674349, 'gnomad_amr...  ...   \n",
       "1       {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   \n",
       "2       {'gnomad_afr': 0.00034498620055197794, 'gnomad...  ...   \n",
       "3       {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...  ...   \n",
       "4       {'gnomad_afr': 0.0060821666284140465, 'gnomad_...  ...   \n",
       "...                                                   ...  ...   \n",
       "731079  {'gnomad_afr': 0.11482740118046861, 'gnomad_am...  ...   \n",
       "731080  {'gnomad_afr': 0.009298393913778529, 'gnomad_a...  ...   \n",
       "731081  {'gnomad_afr': 0.0032312925170068026, 'gnomad_...  ...   \n",
       "731082  {'gnomad_afr': 0.00017018379850238256, 'gnomad...  ...   \n",
       "731083  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   \n",
       "\n",
       "        gene_id_prot_coding chrom          snp cm        pos a1  a2  \\\n",
       "0           ENSG00000162571     1  rs112695918  0    1104584  T   C   \n",
       "1           ENSG00000067606     1   rs74372554  0    1969752  C   T   \n",
       "2           ENSG00000162585     1  rs146607591  0    2125543  A   G   \n",
       "3           ENSG00000157870     1  rs146100312  0    2524107  C   T   \n",
       "4           ENSG00000215912     1  rs116076619  0    2745338  C   A   \n",
       "...                     ...   ...          ... ..        ... ..  ..   \n",
       "731079      ENSG00000166049     X   rs58448897  0  150778372  A   G   \n",
       "731080      ENSG00000102287     X  rs145754054  0  151183035  G   A   \n",
       "731081      ENSG00000183479     X    rs5945389  0  152725049  C   T   \n",
       "731082      ENSG00000184343     X  rs199915647  0  153046995  A   C   \n",
       "731083      ENSG00000269335     X    rs5030868  0  153762634  G   A   \n",
       "\n",
       "                gene_id gene_name snp_bed_id  \n",
       "0       ENSG00000162571    TTLL10        220  \n",
       "1       ENSG00000067606     PRKCZ        599  \n",
       "2       ENSG00000162585   C1orf86        648  \n",
       "3       ENSG00000157870   FAM213B        838  \n",
       "4       ENSG00000215912     TTC34        889  \n",
       "...                 ...       ...        ...  \n",
       "731079  ENSG00000166049     PASD1     802181  \n",
       "731080  ENSG00000102287     GABRE     802336  \n",
       "731081  ENSG00000183479     TREX2     802547  \n",
       "731082  ENSG00000184343     SRPK3     802648  \n",
       "731083  ENSG00000269335     IKBKG     802968  \n",
       "\n",
       "[731084 rows x 23 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ot_bim"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load GO term to gene mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "import pickle "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#go_path = \"../../data/GO_biological_process.pickle\"\n",
    "go_path = \"../../ontology_operations/data/GO_cellular_component.pickle\"\n",
    "G = pickle.load(open(go_path, 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load go terms and genes\n",
    "#goterms_df = pd.read_csv(\"../../ontology_operations/data/GO_terms_biological_processes_merged_EBI_uniprot.csv\", index_col=0)\n",
    "goterms_df = pd.read_csv(\"../../ontology_operations/data/GO_terms_cellular_components_merged_EBI_uniprot.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>merged</th>\n",
       "      <th>len_merged</th>\n",
       "      <th>name</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GO:0000137</th>\n",
       "      <td>GLP;GOLGA8S;GOLGA6A;HUGL;GOT1B;GOLGA6B;GOLGA;N...</td>\n",
       "      <td>49</td>\n",
       "      <td>Golgi cis cisterna</td>\n",
       "      <td>GLP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0000137</th>\n",
       "      <td>GLP;GOLGA8S;GOLGA6A;HUGL;GOT1B;GOLGA6B;GOLGA;N...</td>\n",
       "      <td>49</td>\n",
       "      <td>Golgi cis cisterna</td>\n",
       "      <td>GOLGA8S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0000137</th>\n",
       "      <td>GLP;GOLGA8S;GOLGA6A;HUGL;GOT1B;GOLGA6B;GOLGA;N...</td>\n",
       "      <td>49</td>\n",
       "      <td>Golgi cis cisterna</td>\n",
       "      <td>GOLGA6A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0000137</th>\n",
       "      <td>GLP;GOLGA8S;GOLGA6A;HUGL;GOT1B;GOLGA6B;GOLGA;N...</td>\n",
       "      <td>49</td>\n",
       "      <td>Golgi cis cisterna</td>\n",
       "      <td>HUGL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0000137</th>\n",
       "      <td>GLP;GOLGA8S;GOLGA6A;HUGL;GOT1B;GOLGA6B;GOLGA;N...</td>\n",
       "      <td>49</td>\n",
       "      <td>Golgi cis cisterna</td>\n",
       "      <td>GOT1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:1903754</th>\n",
       "      <td>KIAA0627;CLASP2</td>\n",
       "      <td>2</td>\n",
       "      <td>cortical microtubule plus-end</td>\n",
       "      <td>CLASP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:1990836</th>\n",
       "      <td>HSPA10;HSPA8;HSC70;HSP73</td>\n",
       "      <td>4</td>\n",
       "      <td>lysosomal matrix</td>\n",
       "      <td>HSPA10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:1990836</th>\n",
       "      <td>HSPA10;HSPA8;HSC70;HSP73</td>\n",
       "      <td>4</td>\n",
       "      <td>lysosomal matrix</td>\n",
       "      <td>HSPA8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:1990836</th>\n",
       "      <td>HSPA10;HSPA8;HSC70;HSP73</td>\n",
       "      <td>4</td>\n",
       "      <td>lysosomal matrix</td>\n",
       "      <td>HSC70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:1990836</th>\n",
       "      <td>HSPA10;HSPA8;HSC70;HSP73</td>\n",
       "      <td>4</td>\n",
       "      <td>lysosomal matrix</td>\n",
       "      <td>HSP73</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190968 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       merged  len_merged  \\\n",
       "GO:0000137  GLP;GOLGA8S;GOLGA6A;HUGL;GOT1B;GOLGA6B;GOLGA;N...          49   \n",
       "GO:0000137  GLP;GOLGA8S;GOLGA6A;HUGL;GOT1B;GOLGA6B;GOLGA;N...          49   \n",
       "GO:0000137  GLP;GOLGA8S;GOLGA6A;HUGL;GOT1B;GOLGA6B;GOLGA;N...          49   \n",
       "GO:0000137  GLP;GOLGA8S;GOLGA6A;HUGL;GOT1B;GOLGA6B;GOLGA;N...          49   \n",
       "GO:0000137  GLP;GOLGA8S;GOLGA6A;HUGL;GOT1B;GOLGA6B;GOLGA;N...          49   \n",
       "...                                                       ...         ...   \n",
       "GO:1903754                                    KIAA0627;CLASP2           2   \n",
       "GO:1990836                           HSPA10;HSPA8;HSC70;HSP73           4   \n",
       "GO:1990836                           HSPA10;HSPA8;HSC70;HSP73           4   \n",
       "GO:1990836                           HSPA10;HSPA8;HSC70;HSP73           4   \n",
       "GO:1990836                           HSPA10;HSPA8;HSC70;HSP73           4   \n",
       "\n",
       "                                     name gene_name  \n",
       "GO:0000137             Golgi cis cisterna       GLP  \n",
       "GO:0000137             Golgi cis cisterna   GOLGA8S  \n",
       "GO:0000137             Golgi cis cisterna   GOLGA6A  \n",
       "GO:0000137             Golgi cis cisterna      HUGL  \n",
       "GO:0000137             Golgi cis cisterna     GOT1B  \n",
       "...                                   ...       ...  \n",
       "GO:1903754  cortical microtubule plus-end    CLASP2  \n",
       "GO:1990836               lysosomal matrix    HSPA10  \n",
       "GO:1990836               lysosomal matrix     HSPA8  \n",
       "GO:1990836               lysosomal matrix     HSC70  \n",
       "GO:1990836               lysosomal matrix     HSP73  \n",
       "\n",
       "[190968 rows x 4 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "goterms_df[\"gene_name\"] = goterms_df[\"merged\"].apply(lambda x: x.split(\";\"))\n",
    "goterms_df = goterms_df.explode(\"gene_name\")\n",
    "goterms_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df = pd.DataFrame()\n",
    "gene_df[\"gene_name\"] = goterms_df.gene_name.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_bim = ot_bim.merge(gene_df, on=\"gene_name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16042"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_bim.gene_name.nunique() # some genes not mapped to go terms .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16042"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_bim.gene_name.nunique() # some genes not mapped to go terms .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_id_prot_coding</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>snp_bed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1169204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1104584</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.61, 'raw': -0.223786}</td>\n",
       "      <td>{'gnomad_afr': 0.1370131366674349, 'gnomad_amr...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>1</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>0</td>\n",
       "      <td>1104584</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>TTLL10</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>1</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>0</td>\n",
       "      <td>1969752</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2194104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>2125543</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 4.241, 'raw': 0.098415}</td>\n",
       "      <td>{'gnomad_afr': 0.00034498620055197794, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>0</td>\n",
       "      <td>2125543</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>C1orf86</td>\n",
       "      <td>648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2592668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>2524107</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 23.9, 'raw': 3.118097}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>0</td>\n",
       "      <td>2524107</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>FAM213B</td>\n",
       "      <td>838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>3041036</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>2957600</td>\n",
       "      <td>rs12409277</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 1.823, 'raw': -0.059547}</td>\n",
       "      <td>{'gnomad_afr': 0.13893357848770396, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000169717</td>\n",
       "      <td>1</td>\n",
       "      <td>rs12409277</td>\n",
       "      <td>0</td>\n",
       "      <td>2957600</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000169717</td>\n",
       "      <td>ACTRT2</td>\n",
       "      <td>958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644889</th>\n",
       "      <td>X</td>\n",
       "      <td>151609900</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>X</td>\n",
       "      <td>150778372</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 3.523, 'raw': 0.055668}</td>\n",
       "      <td>{'gnomad_afr': 0.11482740118046861, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000166049</td>\n",
       "      <td>X</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>0</td>\n",
       "      <td>150778372</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000166049</td>\n",
       "      <td>PASD1</td>\n",
       "      <td>802181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644890</th>\n",
       "      <td>X</td>\n",
       "      <td>152014563</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>151183035</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 2.456, 'raw': -0.011208}</td>\n",
       "      <td>{'gnomad_afr': 0.009298393913778529, 'gnomad_a...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>X</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>0</td>\n",
       "      <td>151183035</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>GABRE</td>\n",
       "      <td>802336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644891</th>\n",
       "      <td>X</td>\n",
       "      <td>153459591</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>152725049</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 2.968, 'raw': 0.022059}</td>\n",
       "      <td>{'gnomad_afr': 0.0032312925170068026, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>0</td>\n",
       "      <td>152725049</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>TREX2</td>\n",
       "      <td>802547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644892</th>\n",
       "      <td>X</td>\n",
       "      <td>153781540</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>153046995</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 22.5, 'raw': 2.480113}</td>\n",
       "      <td>{'gnomad_afr': 0.00017018379850238256, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>X</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>0</td>\n",
       "      <td>153046995</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>SRPK3</td>\n",
       "      <td>802648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644893</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>0</td>\n",
       "      <td>153762634</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>644894 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0           1    1169204          T          C          1       1104584   \n",
       "1           1    2038313          C          T          1       1969752   \n",
       "2           1    2194104          A          G          1       2125543   \n",
       "3           1    2592668          C          T          1       2524107   \n",
       "4           1    3041036          T          C          1       2957600   \n",
       "...       ...        ...        ...        ...        ...           ...   \n",
       "644889      X  151609900          A          G          X     150778372   \n",
       "644890      X  152014563          G          A          X     151183035   \n",
       "644891      X  153459591          C          T          X     152725049   \n",
       "644892      X  153781540          A          C          X     153046995   \n",
       "644893      X  154534419          G          A          X     153762634   \n",
       "\n",
       "              rs_id most_severe_consequence  \\\n",
       "0       rs112695918   upstream_gene_variant   \n",
       "1        rs74372554      intergenic_variant   \n",
       "2       rs146607591        missense_variant   \n",
       "3       rs146100312        missense_variant   \n",
       "4        rs12409277      intergenic_variant   \n",
       "...             ...                     ...   \n",
       "644889   rs58448897          intron_variant   \n",
       "644890  rs145754054      intergenic_variant   \n",
       "644891    rs5945389          intron_variant   \n",
       "644892  rs199915647        missense_variant   \n",
       "644893    rs5030868        missense_variant   \n",
       "\n",
       "                                      cadd  \\\n",
       "0        {'phred': 0.61, 'raw': -0.223786}   \n",
       "1       {'phred': 0.309, 'raw': -0.324059}   \n",
       "2        {'phred': 4.241, 'raw': 0.098415}   \n",
       "3         {'phred': 23.9, 'raw': 3.118097}   \n",
       "4       {'phred': 1.823, 'raw': -0.059547}   \n",
       "...                                    ...   \n",
       "644889   {'phred': 3.523, 'raw': 0.055668}   \n",
       "644890  {'phred': 2.456, 'raw': -0.011208}   \n",
       "644891   {'phred': 2.968, 'raw': 0.022059}   \n",
       "644892    {'phred': 22.5, 'raw': 2.480113}   \n",
       "644893    {'phred': 24.7, 'raw': 3.374046}   \n",
       "\n",
       "                                                       af  ...  \\\n",
       "0       {'gnomad_afr': 0.1370131366674349, 'gnomad_amr...  ...   \n",
       "1       {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   \n",
       "2       {'gnomad_afr': 0.00034498620055197794, 'gnomad...  ...   \n",
       "3       {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...  ...   \n",
       "4       {'gnomad_afr': 0.13893357848770396, 'gnomad_am...  ...   \n",
       "...                                                   ...  ...   \n",
       "644889  {'gnomad_afr': 0.11482740118046861, 'gnomad_am...  ...   \n",
       "644890  {'gnomad_afr': 0.009298393913778529, 'gnomad_a...  ...   \n",
       "644891  {'gnomad_afr': 0.0032312925170068026, 'gnomad_...  ...   \n",
       "644892  {'gnomad_afr': 0.00017018379850238256, 'gnomad...  ...   \n",
       "644893  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   \n",
       "\n",
       "        gene_id_prot_coding chrom          snp cm        pos a1  a2  \\\n",
       "0           ENSG00000162571     1  rs112695918  0    1104584  T   C   \n",
       "1           ENSG00000067606     1   rs74372554  0    1969752  C   T   \n",
       "2           ENSG00000162585     1  rs146607591  0    2125543  A   G   \n",
       "3           ENSG00000157870     1  rs146100312  0    2524107  C   T   \n",
       "4           ENSG00000169717     1   rs12409277  0    2957600  T   C   \n",
       "...                     ...   ...          ... ..        ... ..  ..   \n",
       "644889      ENSG00000166049     X   rs58448897  0  150778372  A   G   \n",
       "644890      ENSG00000102287     X  rs145754054  0  151183035  G   A   \n",
       "644891      ENSG00000183479     X    rs5945389  0  152725049  C   T   \n",
       "644892      ENSG00000184343     X  rs199915647  0  153046995  A   C   \n",
       "644893      ENSG00000269335     X    rs5030868  0  153762634  G   A   \n",
       "\n",
       "                gene_id gene_name snp_bed_id  \n",
       "0       ENSG00000162571    TTLL10        220  \n",
       "1       ENSG00000067606     PRKCZ        599  \n",
       "2       ENSG00000162585   C1orf86        648  \n",
       "3       ENSG00000157870   FAM213B        838  \n",
       "4       ENSG00000169717    ACTRT2        958  \n",
       "...                 ...       ...        ...  \n",
       "644889  ENSG00000166049     PASD1     802181  \n",
       "644890  ENSG00000102287     GABRE     802336  \n",
       "644891  ENSG00000183479     TREX2     802547  \n",
       "644892  ENSG00000184343     SRPK3     802648  \n",
       "644893  ENSG00000269335     IKBKG     802968  \n",
       "\n",
       "[644894 rows x 23 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_bed_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TTLL10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>PRKCZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>C1orf86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>FAM213B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ACTRT2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16037</th>\n",
       "      <td>16037</td>\n",
       "      <td>RPL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16038</th>\n",
       "      <td>16038</td>\n",
       "      <td>HOXA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16039</th>\n",
       "      <td>16039</td>\n",
       "      <td>TAF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16040</th>\n",
       "      <td>16040</td>\n",
       "      <td>TPRN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16041</th>\n",
       "      <td>16041</td>\n",
       "      <td>NOMO1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16042 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_bed_id gene_name\n",
       "0                0    TTLL10\n",
       "1                1     PRKCZ\n",
       "2                2   C1orf86\n",
       "3                3   FAM213B\n",
       "4                4    ACTRT2\n",
       "...            ...       ...\n",
       "16037        16037     RPL18\n",
       "16038        16038     HOXA3\n",
       "16039        16039      TAF6\n",
       "16040        16040      TPRN\n",
       "16041        16041     NOMO1\n",
       "\n",
       "[16042 rows x 2 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df = pd.DataFrame()\n",
    "gene_df[\"gene_name\"] = go_bim.gene_name.unique()\n",
    "gene_df.reset_index(inplace=True)\n",
    "gene_df.columns = [\"gene_bed_id\", \"gene_name\"]\n",
    "gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>...</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>snp_bed_id</th>\n",
       "      <th>gene_bed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1169204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1104584</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.61, 'raw': -0.223786}</td>\n",
       "      <td>{'gnomad_afr': 0.1370131366674349, 'gnomad_amr...</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>0</td>\n",
       "      <td>1104584</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>TTLL10</td>\n",
       "      <td>220</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>0</td>\n",
       "      <td>1969752</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2194104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>2125543</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 4.241, 'raw': 0.098415}</td>\n",
       "      <td>{'gnomad_afr': 0.00034498620055197794, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146607591</td>\n",
       "      <td>0</td>\n",
       "      <td>2125543</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000162585</td>\n",
       "      <td>C1orf86</td>\n",
       "      <td>648</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2592668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>2524107</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 23.9, 'raw': 3.118097}</td>\n",
       "      <td>{'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>rs146100312</td>\n",
       "      <td>0</td>\n",
       "      <td>2524107</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000157870</td>\n",
       "      <td>FAM213B</td>\n",
       "      <td>838</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>3041036</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>2957600</td>\n",
       "      <td>rs12409277</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 1.823, 'raw': -0.059547}</td>\n",
       "      <td>{'gnomad_afr': 0.13893357848770396, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>rs12409277</td>\n",
       "      <td>0</td>\n",
       "      <td>2957600</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000169717</td>\n",
       "      <td>ACTRT2</td>\n",
       "      <td>958</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644889</th>\n",
       "      <td>X</td>\n",
       "      <td>151609900</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>X</td>\n",
       "      <td>150778372</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 3.523, 'raw': 0.055668}</td>\n",
       "      <td>{'gnomad_afr': 0.11482740118046861, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>0</td>\n",
       "      <td>150778372</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000166049</td>\n",
       "      <td>PASD1</td>\n",
       "      <td>802181</td>\n",
       "      <td>9628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644890</th>\n",
       "      <td>X</td>\n",
       "      <td>152014563</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>151183035</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 2.456, 'raw': -0.011208}</td>\n",
       "      <td>{'gnomad_afr': 0.009298393913778529, 'gnomad_a...</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>0</td>\n",
       "      <td>151183035</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>GABRE</td>\n",
       "      <td>802336</td>\n",
       "      <td>7660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644891</th>\n",
       "      <td>X</td>\n",
       "      <td>153459591</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X</td>\n",
       "      <td>152725049</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>{'phred': 2.968, 'raw': 0.022059}</td>\n",
       "      <td>{'gnomad_afr': 0.0032312925170068026, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>0</td>\n",
       "      <td>152725049</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000183479</td>\n",
       "      <td>TREX2</td>\n",
       "      <td>802547</td>\n",
       "      <td>2413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644892</th>\n",
       "      <td>X</td>\n",
       "      <td>153781540</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>X</td>\n",
       "      <td>153046995</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 22.5, 'raw': 2.480113}</td>\n",
       "      <td>{'gnomad_afr': 0.00017018379850238256, 'gnomad...</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>0</td>\n",
       "      <td>153046995</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000184343</td>\n",
       "      <td>SRPK3</td>\n",
       "      <td>802648</td>\n",
       "      <td>12949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644893</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>0</td>\n",
       "      <td>153762634</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "      <td>13478</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>644894 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0           1    1169204          T          C          1       1104584   \n",
       "1           1    2038313          C          T          1       1969752   \n",
       "2           1    2194104          A          G          1       2125543   \n",
       "3           1    2592668          C          T          1       2524107   \n",
       "4           1    3041036          T          C          1       2957600   \n",
       "...       ...        ...        ...        ...        ...           ...   \n",
       "644889      X  151609900          A          G          X     150778372   \n",
       "644890      X  152014563          G          A          X     151183035   \n",
       "644891      X  153459591          C          T          X     152725049   \n",
       "644892      X  153781540          A          C          X     153046995   \n",
       "644893      X  154534419          G          A          X     153762634   \n",
       "\n",
       "              rs_id most_severe_consequence  \\\n",
       "0       rs112695918   upstream_gene_variant   \n",
       "1        rs74372554      intergenic_variant   \n",
       "2       rs146607591        missense_variant   \n",
       "3       rs146100312        missense_variant   \n",
       "4        rs12409277      intergenic_variant   \n",
       "...             ...                     ...   \n",
       "644889   rs58448897          intron_variant   \n",
       "644890  rs145754054      intergenic_variant   \n",
       "644891    rs5945389          intron_variant   \n",
       "644892  rs199915647        missense_variant   \n",
       "644893    rs5030868        missense_variant   \n",
       "\n",
       "                                      cadd  \\\n",
       "0        {'phred': 0.61, 'raw': -0.223786}   \n",
       "1       {'phred': 0.309, 'raw': -0.324059}   \n",
       "2        {'phred': 4.241, 'raw': 0.098415}   \n",
       "3         {'phred': 23.9, 'raw': 3.118097}   \n",
       "4       {'phred': 1.823, 'raw': -0.059547}   \n",
       "...                                    ...   \n",
       "644889   {'phred': 3.523, 'raw': 0.055668}   \n",
       "644890  {'phred': 2.456, 'raw': -0.011208}   \n",
       "644891   {'phred': 2.968, 'raw': 0.022059}   \n",
       "644892    {'phred': 22.5, 'raw': 2.480113}   \n",
       "644893    {'phred': 24.7, 'raw': 3.374046}   \n",
       "\n",
       "                                                       af  ...  chrom  \\\n",
       "0       {'gnomad_afr': 0.1370131366674349, 'gnomad_amr...  ...      1   \n",
       "1       {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...      1   \n",
       "2       {'gnomad_afr': 0.00034498620055197794, 'gnomad...  ...      1   \n",
       "3       {'gnomad_afr': 0.0, 'gnomad_amr': 0.0, 'gnomad...  ...      1   \n",
       "4       {'gnomad_afr': 0.13893357848770396, 'gnomad_am...  ...      1   \n",
       "...                                                   ...  ...    ...   \n",
       "644889  {'gnomad_afr': 0.11482740118046861, 'gnomad_am...  ...      X   \n",
       "644890  {'gnomad_afr': 0.009298393913778529, 'gnomad_a...  ...      X   \n",
       "644891  {'gnomad_afr': 0.0032312925170068026, 'gnomad_...  ...      X   \n",
       "644892  {'gnomad_afr': 0.00017018379850238256, 'gnomad...  ...      X   \n",
       "644893  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...      X   \n",
       "\n",
       "                snp  cm        pos a1 a2          gene_id  gene_name  \\\n",
       "0       rs112695918   0    1104584  T  C  ENSG00000162571     TTLL10   \n",
       "1        rs74372554   0    1969752  C  T  ENSG00000067606      PRKCZ   \n",
       "2       rs146607591   0    2125543  A  G  ENSG00000162585    C1orf86   \n",
       "3       rs146100312   0    2524107  C  T  ENSG00000157870    FAM213B   \n",
       "4        rs12409277   0    2957600  T  C  ENSG00000169717     ACTRT2   \n",
       "...             ...  ..        ... .. ..              ...        ...   \n",
       "644889   rs58448897   0  150778372  A  G  ENSG00000166049      PASD1   \n",
       "644890  rs145754054   0  151183035  G  A  ENSG00000102287      GABRE   \n",
       "644891    rs5945389   0  152725049  C  T  ENSG00000183479      TREX2   \n",
       "644892  rs199915647   0  153046995  A  C  ENSG00000184343      SRPK3   \n",
       "644893    rs5030868   0  153762634  G  A  ENSG00000269335      IKBKG   \n",
       "\n",
       "       snp_bed_id gene_bed_id  \n",
       "0             220           0  \n",
       "1             599           1  \n",
       "2             648           2  \n",
       "3             838           3  \n",
       "4             958           4  \n",
       "...           ...         ...  \n",
       "644889     802181        9628  \n",
       "644890     802336        7660  \n",
       "644891     802547        2413  \n",
       "644892     802648       12949  \n",
       "644893     802968       13478  \n",
       "\n",
       "[644894 rows x 24 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_bim = go_bim.merge(gene_df, on=\"gene_name\")\n",
    "go_bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_snp_ids = go_bim.snp_bed_id.unique() # list of all snp ids, enumerate and map to ids\n",
    "snp_id_dict = dict(zip(all_snp_ids, range(len(all_snp_ids))))\n",
    "# snp position in gene_level bed file\n",
    "go_bim['snp_gbed_id'] = go_bim['snp_bed_id'].map(snp_id_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>snp_bed_id</th>\n",
       "      <th>snp_gbed_id</th>\n",
       "      <th>gene_bed_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs112695918</td>\n",
       "      <td>TTLL10</td>\n",
       "      <td>220</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs74372554</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs146607591</td>\n",
       "      <td>C1orf86</td>\n",
       "      <td>648</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs146100312</td>\n",
       "      <td>FAM213B</td>\n",
       "      <td>838</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs12409277</td>\n",
       "      <td>ACTRT2</td>\n",
       "      <td>958</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644889</th>\n",
       "      <td>rs58448897</td>\n",
       "      <td>PASD1</td>\n",
       "      <td>802181</td>\n",
       "      <td>644889</td>\n",
       "      <td>9628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644890</th>\n",
       "      <td>rs145754054</td>\n",
       "      <td>GABRE</td>\n",
       "      <td>802336</td>\n",
       "      <td>644890</td>\n",
       "      <td>7660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644891</th>\n",
       "      <td>rs5945389</td>\n",
       "      <td>TREX2</td>\n",
       "      <td>802547</td>\n",
       "      <td>644891</td>\n",
       "      <td>2413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644892</th>\n",
       "      <td>rs199915647</td>\n",
       "      <td>SRPK3</td>\n",
       "      <td>802648</td>\n",
       "      <td>644892</td>\n",
       "      <td>12949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644893</th>\n",
       "      <td>rs5030868</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "      <td>644893</td>\n",
       "      <td>13478</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>644894 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                snp gene_name  snp_bed_id  snp_gbed_id  gene_bed_id\n",
       "0       rs112695918    TTLL10         220            0            0\n",
       "1        rs74372554     PRKCZ         599            1            1\n",
       "2       rs146607591   C1orf86         648            2            2\n",
       "3       rs146100312   FAM213B         838            3            3\n",
       "4        rs12409277    ACTRT2         958            4            4\n",
       "...             ...       ...         ...          ...          ...\n",
       "644889   rs58448897     PASD1      802181       644889         9628\n",
       "644890  rs145754054     GABRE      802336       644890         7660\n",
       "644891    rs5945389     TREX2      802547       644891         2413\n",
       "644892  rs199915647     SRPK3      802648       644892        12949\n",
       "644893    rs5030868     IKBKG      802968       644893        13478\n",
       "\n",
       "[644894 rows x 5 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_bim[[\"snp\", \"gene_name\", \"snp_bed_id\", \"snp_gbed_id\", \"gene_bed_id\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_bim[[\"snp\", \"gene_name\", \"snp_bed_id\", \"snp_gbed_id\", \"gene_bed_id\"]].to_csv(\"../data/go_cellular_components_bim.csv\", index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#go_bim.to_csv(\"../data/go_biological_processes_bim_full.csv\", index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prune network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "import pickle\n",
    "\n",
    "#go_path = \"../../ontology_operations/data/GO_biological_process.pickle\"\n",
    "go_path = \"../../ontology_operations/data/GO_cellular_component.pickle\"\n",
    "G = pickle.load(open(go_path, 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_bim = pd.read_csv(\"../data/go_biological_processes_bim_full.csv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_go = go_bim.merge(goterms_df.reset_index(), on=\"gene_name\") # .... lot's of multiple connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_id</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>chr_id_b37</th>\n",
       "      <th>position_b37</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>cadd</th>\n",
       "      <th>af</th>\n",
       "      <th>...</th>\n",
       "      <th>a2</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>snp_bed_id</th>\n",
       "      <th>gene_bed_id</th>\n",
       "      <th>snp_gbed_id</th>\n",
       "      <th>index</th>\n",
       "      <th>merged</th>\n",
       "      <th>len_merged</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1169204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1104584</td>\n",
       "      <td>rs112695918</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>{'phred': 0.61, 'raw': -0.223786}</td>\n",
       "      <td>{'gnomad_afr': 0.1370131366674349, 'gnomad_amr...</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000162571</td>\n",
       "      <td>TTLL10</td>\n",
       "      <td>220</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GO:0005829</td>\n",
       "      <td>M130;NFATC1;GGCT;PKDL;TORC3;NIP2;CPGL;KNSL5;DN...</td>\n",
       "      <td>13464</td>\n",
       "      <td>cytosol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>GO:0001725</td>\n",
       "      <td>C20orf28;CTTNBP2NL;LPP;KIAA1296;KIAA0991;DRR1;...</td>\n",
       "      <td>211</td>\n",
       "      <td>stress fiber</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>DHX38;B4E246;NFATC1;CRAMP1L;ZNF526;CCNP;TORC3;...</td>\n",
       "      <td>17387</td>\n",
       "      <td>nucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>LSLCL;NFATC1;CCNP;TORC3;NIP2;B2R7M0;ATG16L2;GO...</td>\n",
       "      <td>16823</td>\n",
       "      <td>cytoplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2038313</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1969752</td>\n",
       "      <td>rs74372554</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>{'phred': 0.309, 'raw': -0.324059}</td>\n",
       "      <td>{'gnomad_afr': 0.01078971533516988, 'gnomad_am...</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000067606</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>599</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>GO:0005768</td>\n",
       "      <td>KIAA0319;A0A384ME16;PKN;ARL8B;B3KQD2;B4E2S7;TM...</td>\n",
       "      <td>1498</td>\n",
       "      <td>endosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2946977</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "      <td>13478</td>\n",
       "      <td>644893</td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>DHX38;B4E246;NFATC1;CRAMP1L;ZNF526;CCNP;TORC3;...</td>\n",
       "      <td>17387</td>\n",
       "      <td>nucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2946978</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "      <td>13478</td>\n",
       "      <td>644893</td>\n",
       "      <td>GO:0005654</td>\n",
       "      <td>DHX38;NFATC1;TTF1;TORC3;CPGL;C20orf77;PRO1146;...</td>\n",
       "      <td>10042</td>\n",
       "      <td>nucleoplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2946979</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "      <td>13478</td>\n",
       "      <td>644893</td>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>LSLCL;NFATC1;CCNP;TORC3;NIP2;B2R7M0;ATG16L2;GO...</td>\n",
       "      <td>16823</td>\n",
       "      <td>cytoplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2946980</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "      <td>13478</td>\n",
       "      <td>644893</td>\n",
       "      <td>GO:0005829</td>\n",
       "      <td>M130;NFATC1;GGCT;PKDL;TORC3;NIP2;CPGL;KNSL5;DN...</td>\n",
       "      <td>13464</td>\n",
       "      <td>cytosol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2946981</th>\n",
       "      <td>X</td>\n",
       "      <td>154534419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>X</td>\n",
       "      <td>153762634</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>{'phred': 24.7, 'raw': 3.374046}</td>\n",
       "      <td>{'gnomad_afr': 0.0001721763085399449, 'gnomad_...</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000269335</td>\n",
       "      <td>IKBKG</td>\n",
       "      <td>802968</td>\n",
       "      <td>13478</td>\n",
       "      <td>644893</td>\n",
       "      <td>GO:0072686</td>\n",
       "      <td>NGRN;IL8RB;IRA1;DCDC1;B4DY90;USP44;STK15;CDC14...</td>\n",
       "      <td>359</td>\n",
       "      <td>mitotic spindle</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2946982 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr_id   position ref_allele alt_allele chr_id_b37  position_b37  \\\n",
       "0            1    1169204          T          C          1       1104584   \n",
       "1            1    2038313          C          T          1       1969752   \n",
       "2            1    2038313          C          T          1       1969752   \n",
       "3            1    2038313          C          T          1       1969752   \n",
       "4            1    2038313          C          T          1       1969752   \n",
       "...        ...        ...        ...        ...        ...           ...   \n",
       "2946977      X  154534419          G          A          X     153762634   \n",
       "2946978      X  154534419          G          A          X     153762634   \n",
       "2946979      X  154534419          G          A          X     153762634   \n",
       "2946980      X  154534419          G          A          X     153762634   \n",
       "2946981      X  154534419          G          A          X     153762634   \n",
       "\n",
       "               rs_id most_severe_consequence  \\\n",
       "0        rs112695918   upstream_gene_variant   \n",
       "1         rs74372554      intergenic_variant   \n",
       "2         rs74372554      intergenic_variant   \n",
       "3         rs74372554      intergenic_variant   \n",
       "4         rs74372554      intergenic_variant   \n",
       "...              ...                     ...   \n",
       "2946977    rs5030868        missense_variant   \n",
       "2946978    rs5030868        missense_variant   \n",
       "2946979    rs5030868        missense_variant   \n",
       "2946980    rs5030868        missense_variant   \n",
       "2946981    rs5030868        missense_variant   \n",
       "\n",
       "                                       cadd  \\\n",
       "0         {'phred': 0.61, 'raw': -0.223786}   \n",
       "1        {'phred': 0.309, 'raw': -0.324059}   \n",
       "2        {'phred': 0.309, 'raw': -0.324059}   \n",
       "3        {'phred': 0.309, 'raw': -0.324059}   \n",
       "4        {'phred': 0.309, 'raw': -0.324059}   \n",
       "...                                     ...   \n",
       "2946977    {'phred': 24.7, 'raw': 3.374046}   \n",
       "2946978    {'phred': 24.7, 'raw': 3.374046}   \n",
       "2946979    {'phred': 24.7, 'raw': 3.374046}   \n",
       "2946980    {'phred': 24.7, 'raw': 3.374046}   \n",
       "2946981    {'phred': 24.7, 'raw': 3.374046}   \n",
       "\n",
       "                                                        af  ...  a2  \\\n",
       "0        {'gnomad_afr': 0.1370131366674349, 'gnomad_amr...  ...   C   \n",
       "1        {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   T   \n",
       "2        {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   T   \n",
       "3        {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   T   \n",
       "4        {'gnomad_afr': 0.01078971533516988, 'gnomad_am...  ...   T   \n",
       "...                                                    ...  ...  ..   \n",
       "2946977  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   A   \n",
       "2946978  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   A   \n",
       "2946979  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   A   \n",
       "2946980  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   A   \n",
       "2946981  {'gnomad_afr': 0.0001721763085399449, 'gnomad_...  ...   A   \n",
       "\n",
       "                 gene_id  gene_name snp_bed_id gene_bed_id snp_gbed_id  \\\n",
       "0        ENSG00000162571     TTLL10        220           0           0   \n",
       "1        ENSG00000067606      PRKCZ        599           1           1   \n",
       "2        ENSG00000067606      PRKCZ        599           1           1   \n",
       "3        ENSG00000067606      PRKCZ        599           1           1   \n",
       "4        ENSG00000067606      PRKCZ        599           1           1   \n",
       "...                  ...        ...        ...         ...         ...   \n",
       "2946977  ENSG00000269335      IKBKG     802968       13478      644893   \n",
       "2946978  ENSG00000269335      IKBKG     802968       13478      644893   \n",
       "2946979  ENSG00000269335      IKBKG     802968       13478      644893   \n",
       "2946980  ENSG00000269335      IKBKG     802968       13478      644893   \n",
       "2946981  ENSG00000269335      IKBKG     802968       13478      644893   \n",
       "\n",
       "              index                                             merged  \\\n",
       "0        GO:0005829  M130;NFATC1;GGCT;PKDL;TORC3;NIP2;CPGL;KNSL5;DN...   \n",
       "1        GO:0001725  C20orf28;CTTNBP2NL;LPP;KIAA1296;KIAA0991;DRR1;...   \n",
       "2        GO:0005634  DHX38;B4E246;NFATC1;CRAMP1L;ZNF526;CCNP;TORC3;...   \n",
       "3        GO:0005737  LSLCL;NFATC1;CCNP;TORC3;NIP2;B2R7M0;ATG16L2;GO...   \n",
       "4        GO:0005768  KIAA0319;A0A384ME16;PKN;ARL8B;B3KQD2;B4E2S7;TM...   \n",
       "...             ...                                                ...   \n",
       "2946977  GO:0005634  DHX38;B4E246;NFATC1;CRAMP1L;ZNF526;CCNP;TORC3;...   \n",
       "2946978  GO:0005654  DHX38;NFATC1;TTF1;TORC3;CPGL;C20orf77;PRO1146;...   \n",
       "2946979  GO:0005737  LSLCL;NFATC1;CCNP;TORC3;NIP2;B2R7M0;ATG16L2;GO...   \n",
       "2946980  GO:0005829  M130;NFATC1;GGCT;PKDL;TORC3;NIP2;CPGL;KNSL5;DN...   \n",
       "2946981  GO:0072686  NGRN;IL8RB;IRA1;DCDC1;B4DY90;USP44;STK15;CDC14...   \n",
       "\n",
       "        len_merged             name  \n",
       "0            13464          cytosol  \n",
       "1              211     stress fiber  \n",
       "2            17387          nucleus  \n",
       "3            16823        cytoplasm  \n",
       "4             1498         endosome  \n",
       "...            ...              ...  \n",
       "2946977      17387          nucleus  \n",
       "2946978      10042      nucleoplasm  \n",
       "2946979      16823        cytoplasm  \n",
       "2946980      13464          cytosol  \n",
       "2946981        359  mitotic spindle  \n",
       "\n",
       "[2946982 rows x 29 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "859"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_go[\"index\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11364"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_go[\"index\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_terms = merged_go[\"index\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GO_0000137',\n",
       " 'GO_0000138',\n",
       " 'GO_0000139',\n",
       " 'GO_0000145',\n",
       " 'GO_0000220',\n",
       " 'GO_0000221',\n",
       " 'GO_0000228',\n",
       " 'GO_0000235',\n",
       " 'GO_0000242',\n",
       " 'GO_0000262',\n",
       " 'GO_0000322',\n",
       " 'GO_0000323',\n",
       " 'GO_0000407',\n",
       " 'GO_0000421',\n",
       " 'GO_0000793',\n",
       " 'GO_0000794',\n",
       " 'GO_0000800',\n",
       " 'GO_0000801',\n",
       " 'GO_0000802',\n",
       " 'GO_0000803',\n",
       " 'GO_0000805',\n",
       " 'GO_0000806',\n",
       " 'GO_0000923',\n",
       " 'GO_0000932',\n",
       " 'GO_0001520',\n",
       " 'GO_0001527',\n",
       " 'GO_0001533',\n",
       " 'GO_0001534',\n",
       " 'GO_0001535',\n",
       " 'GO_0001650',\n",
       " 'GO_0001651',\n",
       " 'GO_0001652',\n",
       " 'GO_0001669',\n",
       " 'GO_0001725',\n",
       " 'GO_0001726',\n",
       " 'GO_0001739',\n",
       " 'GO_0001740',\n",
       " 'GO_0001741',\n",
       " 'GO_0001750',\n",
       " 'GO_0001772',\n",
       " 'GO_0001891',\n",
       " 'GO_0001917',\n",
       " 'GO_0001931',\n",
       " 'GO_0001939',\n",
       " 'GO_0001940',\n",
       " 'GO_0002079',\n",
       " 'GO_0002080',\n",
       " 'GO_0002081',\n",
       " 'GO_0002096',\n",
       " 'GO_0002102',\n",
       " 'GO_0002141',\n",
       " 'GO_0002177',\n",
       " 'GO_0005576',\n",
       " 'GO_0005581',\n",
       " 'GO_0005582',\n",
       " 'GO_0005583',\n",
       " 'GO_0005584',\n",
       " 'GO_0005585',\n",
       " 'GO_0005586',\n",
       " 'GO_0005587',\n",
       " 'GO_0005588',\n",
       " 'GO_0005589',\n",
       " 'GO_0005590',\n",
       " 'GO_0005591',\n",
       " 'GO_0005592',\n",
       " 'GO_0005594',\n",
       " 'GO_0005595',\n",
       " 'GO_0005596',\n",
       " 'GO_0005597',\n",
       " 'GO_0005599',\n",
       " 'GO_0005600',\n",
       " 'GO_0005614',\n",
       " 'GO_0005615',\n",
       " 'GO_0005634',\n",
       " 'GO_0005637',\n",
       " 'GO_0005638',\n",
       " 'GO_0005640',\n",
       " 'GO_0005642',\n",
       " 'GO_0005652',\n",
       " 'GO_0005654',\n",
       " 'GO_0005657',\n",
       " 'GO_0005682',\n",
       " 'GO_0005683',\n",
       " 'GO_0005685',\n",
       " 'GO_0005686',\n",
       " 'GO_0005687',\n",
       " 'GO_0005688',\n",
       " 'GO_0005690',\n",
       " 'GO_0005691',\n",
       " 'GO_0005694',\n",
       " 'GO_0005700',\n",
       " 'GO_0005712',\n",
       " 'GO_0005713',\n",
       " 'GO_0005715',\n",
       " 'GO_0005726',\n",
       " 'GO_0005730',\n",
       " 'GO_0005737',\n",
       " 'GO_0005739',\n",
       " 'GO_0005740',\n",
       " 'GO_0005741',\n",
       " 'GO_0005743',\n",
       " 'GO_0005761',\n",
       " 'GO_0005762',\n",
       " 'GO_0005763',\n",
       " 'GO_0005764',\n",
       " 'GO_0005765',\n",
       " 'GO_0005767',\n",
       " 'GO_0005768',\n",
       " 'GO_0005769',\n",
       " 'GO_0005770',\n",
       " 'GO_0005771',\n",
       " 'GO_0005773',\n",
       " 'GO_0005774',\n",
       " 'GO_0005775',\n",
       " 'GO_0005776',\n",
       " 'GO_0005778',\n",
       " 'GO_0005783',\n",
       " 'GO_0005789',\n",
       " 'GO_0005790',\n",
       " 'GO_0005791',\n",
       " 'GO_0005793',\n",
       " 'GO_0005794',\n",
       " 'GO_0005796',\n",
       " 'GO_0005797',\n",
       " 'GO_0005798',\n",
       " 'GO_0005801',\n",
       " 'GO_0005802',\n",
       " 'GO_0005811',\n",
       " 'GO_0005813',\n",
       " 'GO_0005814',\n",
       " 'GO_0005815',\n",
       " 'GO_0005816',\n",
       " 'GO_0005818',\n",
       " 'GO_0005819',\n",
       " 'GO_0005827',\n",
       " 'GO_0005828',\n",
       " 'GO_0005829',\n",
       " 'GO_0005838',\n",
       " 'GO_0005856',\n",
       " 'GO_0005862',\n",
       " 'GO_0005863',\n",
       " 'GO_0005865',\n",
       " 'GO_0005874',\n",
       " 'GO_0005876',\n",
       " 'GO_0005879',\n",
       " 'GO_0005880',\n",
       " 'GO_0005881',\n",
       " 'GO_0005882',\n",
       " 'GO_0005883',\n",
       " 'GO_0005884',\n",
       " 'GO_0005902',\n",
       " 'GO_0005903',\n",
       " 'GO_0005905',\n",
       " 'GO_0005911',\n",
       " 'GO_0005914',\n",
       " 'GO_0005915',\n",
       " 'GO_0005916',\n",
       " 'GO_0005921',\n",
       " 'GO_0005923',\n",
       " 'GO_0005927',\n",
       " 'GO_0005938',\n",
       " 'GO_0008021',\n",
       " 'GO_0008091',\n",
       " 'GO_0008180',\n",
       " 'GO_0008282',\n",
       " 'GO_0009897',\n",
       " 'GO_0009898',\n",
       " 'GO_0009925',\n",
       " 'GO_0010008',\n",
       " 'GO_0010009',\n",
       " 'GO_0010369',\n",
       " 'GO_0010370',\n",
       " 'GO_0010494',\n",
       " 'GO_0012505',\n",
       " 'GO_0012506',\n",
       " 'GO_0012510',\n",
       " 'GO_0014701',\n",
       " 'GO_0014704',\n",
       " 'GO_0014705',\n",
       " 'GO_0014731',\n",
       " 'GO_0014801',\n",
       " 'GO_0014802',\n",
       " 'GO_0014804',\n",
       " 'GO_0015030',\n",
       " 'GO_0015629',\n",
       " 'GO_0015630',\n",
       " 'GO_0015934',\n",
       " 'GO_0015935',\n",
       " 'GO_0016234',\n",
       " 'GO_0016235',\n",
       " 'GO_0016323',\n",
       " 'GO_0016324',\n",
       " 'GO_0016327',\n",
       " 'GO_0016328',\n",
       " 'GO_0016363',\n",
       " 'GO_0016528',\n",
       " 'GO_0016591',\n",
       " 'GO_0016604',\n",
       " 'GO_0016605',\n",
       " 'GO_0016607',\n",
       " 'GO_0019866',\n",
       " 'GO_0019867',\n",
       " 'GO_0020018',\n",
       " 'GO_0030016',\n",
       " 'GO_0030017',\n",
       " 'GO_0030018',\n",
       " 'GO_0030027',\n",
       " 'GO_0030054',\n",
       " 'GO_0030055',\n",
       " 'GO_0030056',\n",
       " 'GO_0030057',\n",
       " 'GO_0030061',\n",
       " 'GO_0030117',\n",
       " 'GO_0030120',\n",
       " 'GO_0030125',\n",
       " 'GO_0030127',\n",
       " 'GO_0030130',\n",
       " 'GO_0030132',\n",
       " 'GO_0030133',\n",
       " 'GO_0030136',\n",
       " 'GO_0030137',\n",
       " 'GO_0030139',\n",
       " 'GO_0030140',\n",
       " 'GO_0030141',\n",
       " 'GO_0030142',\n",
       " 'GO_0030315',\n",
       " 'GO_0030424',\n",
       " 'GO_0030425',\n",
       " 'GO_0030426',\n",
       " 'GO_0030485',\n",
       " 'GO_0030496',\n",
       " 'GO_0030658',\n",
       " 'GO_0030659',\n",
       " 'GO_0030660',\n",
       " 'GO_0030662',\n",
       " 'GO_0030663',\n",
       " 'GO_0030665',\n",
       " 'GO_0030666',\n",
       " 'GO_0030667',\n",
       " 'GO_0030669',\n",
       " 'GO_0030670',\n",
       " 'GO_0030672',\n",
       " 'GO_0030673',\n",
       " 'GO_0030684',\n",
       " 'GO_0030686',\n",
       " 'GO_0030687',\n",
       " 'GO_0030688',\n",
       " 'GO_0030849',\n",
       " 'GO_0030863',\n",
       " 'GO_0030864',\n",
       " 'GO_0030867',\n",
       " 'GO_0030868',\n",
       " 'GO_0030894',\n",
       " 'GO_0030981',\n",
       " 'GO_0030990',\n",
       " 'GO_0030991',\n",
       " 'GO_0030992',\n",
       " 'GO_0031021',\n",
       " 'GO_0031045',\n",
       " 'GO_0031080',\n",
       " 'GO_0031088',\n",
       " 'GO_0031089',\n",
       " 'GO_0031090',\n",
       " 'GO_0031091',\n",
       " 'GO_0031092',\n",
       " 'GO_0031093',\n",
       " 'GO_0031094',\n",
       " 'GO_0031095',\n",
       " 'GO_0031143',\n",
       " 'GO_0031232',\n",
       " 'GO_0031234',\n",
       " 'GO_0031252',\n",
       " 'GO_0031253',\n",
       " 'GO_0031254',\n",
       " 'GO_0031256',\n",
       " 'GO_0031258',\n",
       " 'GO_0031259',\n",
       " 'GO_0031260',\n",
       " 'GO_0031312',\n",
       " 'GO_0031313',\n",
       " 'GO_0031314',\n",
       " 'GO_0031430',\n",
       " 'GO_0031526',\n",
       " 'GO_0031527',\n",
       " 'GO_0031528',\n",
       " 'GO_0031592',\n",
       " 'GO_0031594',\n",
       " 'GO_0031616',\n",
       " 'GO_0031672',\n",
       " 'GO_0031673',\n",
       " 'GO_0031674',\n",
       " 'GO_0031901',\n",
       " 'GO_0031902',\n",
       " 'GO_0031904',\n",
       " 'GO_0031905',\n",
       " 'GO_0031906',\n",
       " 'GO_0031941',\n",
       " 'GO_0031965',\n",
       " 'GO_0031966',\n",
       " 'GO_0031981',\n",
       " 'GO_0031985',\n",
       " 'GO_0032009',\n",
       " 'GO_0032010',\n",
       " 'GO_0032040',\n",
       " 'GO_0032059',\n",
       " 'GO_0032127',\n",
       " 'GO_0032153',\n",
       " 'GO_0032154',\n",
       " 'GO_0032156',\n",
       " 'GO_0032173',\n",
       " 'GO_0032279',\n",
       " 'GO_0032280',\n",
       " 'GO_0032391',\n",
       " 'GO_0032420',\n",
       " 'GO_0032421',\n",
       " 'GO_0032426',\n",
       " 'GO_0032433',\n",
       " 'GO_0032437',\n",
       " 'GO_0032473',\n",
       " 'GO_0032541',\n",
       " 'GO_0032580',\n",
       " 'GO_0032584',\n",
       " 'GO_0032585',\n",
       " 'GO_0032587',\n",
       " 'GO_0032588',\n",
       " 'GO_0032589',\n",
       " 'GO_0032590',\n",
       " 'GO_0032591',\n",
       " 'GO_0032593',\n",
       " 'GO_0032809',\n",
       " 'GO_0032839',\n",
       " 'GO_0032982',\n",
       " 'GO_0033010',\n",
       " 'GO_0033011',\n",
       " 'GO_0033017',\n",
       " 'GO_0033018',\n",
       " 'GO_0033093',\n",
       " 'GO_0033106',\n",
       " 'GO_0033116',\n",
       " 'GO_0033150',\n",
       " 'GO_0033162',\n",
       " 'GO_0033163',\n",
       " 'GO_0033165',\n",
       " 'GO_0033179',\n",
       " 'GO_0033180',\n",
       " 'GO_0033268',\n",
       " 'GO_0033269',\n",
       " 'GO_0033270',\n",
       " 'GO_0033391',\n",
       " 'GO_0034045',\n",
       " 'GO_0034358',\n",
       " 'GO_0034359',\n",
       " 'GO_0034360',\n",
       " 'GO_0034361',\n",
       " 'GO_0034362',\n",
       " 'GO_0034363',\n",
       " 'GO_0034364',\n",
       " 'GO_0034365',\n",
       " 'GO_0034366',\n",
       " 'GO_0034399',\n",
       " 'GO_0034423',\n",
       " 'GO_0034451',\n",
       " 'GO_0034464',\n",
       " 'GO_0034466',\n",
       " 'GO_0034506',\n",
       " 'GO_0034515',\n",
       " 'GO_0034592',\n",
       " 'GO_0034693',\n",
       " 'GO_0034709',\n",
       " 'GO_0034774',\n",
       " 'GO_0034777',\n",
       " 'GO_0035061',\n",
       " 'GO_0035253',\n",
       " 'GO_0035363',\n",
       " 'GO_0035371',\n",
       " 'GO_0035577',\n",
       " 'GO_0035578',\n",
       " 'GO_0035579',\n",
       " 'GO_0035580',\n",
       " 'GO_0035686',\n",
       " 'GO_0035748',\n",
       " 'GO_0035749',\n",
       " 'GO_0035770',\n",
       " 'GO_0035805',\n",
       " 'GO_0035838',\n",
       " 'GO_0035841',\n",
       " 'GO_0035861',\n",
       " 'GO_0035869',\n",
       " 'GO_0035985',\n",
       " 'GO_0036019',\n",
       " 'GO_0036020',\n",
       " 'GO_0036021',\n",
       " 'GO_0036053',\n",
       " 'GO_0036057',\n",
       " 'GO_0036117',\n",
       " 'GO_0036126',\n",
       " 'GO_0036156',\n",
       " 'GO_0036157',\n",
       " 'GO_0036186',\n",
       " 'GO_0036195',\n",
       " 'GO_0036379',\n",
       " 'GO_0036398',\n",
       " 'GO_0036449',\n",
       " 'GO_0036457',\n",
       " 'GO_0036464',\n",
       " 'GO_0036477',\n",
       " 'GO_0042175',\n",
       " 'GO_0042382',\n",
       " 'GO_0042383',\n",
       " 'GO_0042405',\n",
       " 'GO_0042406',\n",
       " 'GO_0042470',\n",
       " 'GO_0042581',\n",
       " 'GO_0042582',\n",
       " 'GO_0042583',\n",
       " 'GO_0042584',\n",
       " 'GO_0042585',\n",
       " 'GO_0042587',\n",
       " 'GO_0042588',\n",
       " 'GO_0042589',\n",
       " 'GO_0042599',\n",
       " 'GO_0042622',\n",
       " 'GO_0042627',\n",
       " 'GO_0042629',\n",
       " 'GO_0042641',\n",
       " 'GO_0042645',\n",
       " 'GO_0042734',\n",
       " 'GO_0042827',\n",
       " 'GO_0042995',\n",
       " 'GO_0043005',\n",
       " 'GO_0043025',\n",
       " 'GO_0043034',\n",
       " 'GO_0043073',\n",
       " 'GO_0043076',\n",
       " 'GO_0043083',\n",
       " 'GO_0043159',\n",
       " 'GO_0043160',\n",
       " 'GO_0043194',\n",
       " 'GO_0043195',\n",
       " 'GO_0043196',\n",
       " 'GO_0043197',\n",
       " 'GO_0043198',\n",
       " 'GO_0043202',\n",
       " 'GO_0043203',\n",
       " 'GO_0043204',\n",
       " 'GO_0043209',\n",
       " 'GO_0043218',\n",
       " 'GO_0043219',\n",
       " 'GO_0043226',\n",
       " 'GO_0043227',\n",
       " 'GO_0043228',\n",
       " 'GO_0043229',\n",
       " 'GO_0043231',\n",
       " 'GO_0043232',\n",
       " 'GO_0043292',\n",
       " 'GO_0043293',\n",
       " 'GO_0043505',\n",
       " 'GO_0043596',\n",
       " 'GO_0043657',\n",
       " 'GO_0043679',\n",
       " 'GO_0044194',\n",
       " 'GO_0044195',\n",
       " 'GO_0044224',\n",
       " 'GO_0044225',\n",
       " 'GO_0044232',\n",
       " 'GO_0044233',\n",
       " 'GO_0044284',\n",
       " 'GO_0044291',\n",
       " 'GO_0044292',\n",
       " 'GO_0044294',\n",
       " 'GO_0044295',\n",
       " 'GO_0044297',\n",
       " 'GO_0044298',\n",
       " 'GO_0044299',\n",
       " 'GO_0044300',\n",
       " 'GO_0044301',\n",
       " 'GO_0044302',\n",
       " 'GO_0044304',\n",
       " 'GO_0044305',\n",
       " 'GO_0044306',\n",
       " 'GO_0044307',\n",
       " 'GO_0044308',\n",
       " 'GO_0044309',\n",
       " 'GO_0044316',\n",
       " 'GO_0044322',\n",
       " 'GO_0044326',\n",
       " 'GO_0044327',\n",
       " 'GO_0044354',\n",
       " 'GO_0044391',\n",
       " 'GO_0044393',\n",
       " 'GO_0044611',\n",
       " 'GO_0044613',\n",
       " 'GO_0044614',\n",
       " 'GO_0044615',\n",
       " 'GO_0044753',\n",
       " 'GO_0044754',\n",
       " 'GO_0044853',\n",
       " 'GO_0045025',\n",
       " 'GO_0045095',\n",
       " 'GO_0045098',\n",
       " 'GO_0045111',\n",
       " 'GO_0045120',\n",
       " 'GO_0045121',\n",
       " 'GO_0045171',\n",
       " 'GO_0045177',\n",
       " 'GO_0045178',\n",
       " 'GO_0045179',\n",
       " 'GO_0045180',\n",
       " 'GO_0045202',\n",
       " 'GO_0045211',\n",
       " 'GO_0045334',\n",
       " 'GO_0045335',\n",
       " 'GO_0046581',\n",
       " 'GO_0046691',\n",
       " 'GO_0048237',\n",
       " 'GO_0048471',\n",
       " 'GO_0048500',\n",
       " 'GO_0048786',\n",
       " 'GO_0048787',\n",
       " 'GO_0048788',\n",
       " 'GO_0051233',\n",
       " 'GO_0051286',\n",
       " 'GO_0055028',\n",
       " 'GO_0055037',\n",
       " 'GO_0055038',\n",
       " 'GO_0055120',\n",
       " 'GO_0060053',\n",
       " 'GO_0060076',\n",
       " 'GO_0060077',\n",
       " 'GO_0060091',\n",
       " 'GO_0060171',\n",
       " 'GO_0060187',\n",
       " 'GO_0060201',\n",
       " 'GO_0060203',\n",
       " 'GO_0060205',\n",
       " 'GO_0060342',\n",
       " 'GO_0060473',\n",
       " 'GO_0060987',\n",
       " 'GO_0061202',\n",
       " 'GO_0061474',\n",
       " 'GO_0061673',\n",
       " 'GO_0061689',\n",
       " 'GO_0061826',\n",
       " 'GO_0061827',\n",
       " 'GO_0061832',\n",
       " 'GO_0061845',\n",
       " 'GO_0061846',\n",
       " 'GO_0061850',\n",
       " 'GO_0061851',\n",
       " 'GO_0061908',\n",
       " 'GO_0062023',\n",
       " 'GO_0065010',\n",
       " 'GO_0070062',\n",
       " 'GO_0070083',\n",
       " 'GO_0070160',\n",
       " 'GO_0070161',\n",
       " 'GO_0070319',\n",
       " 'GO_0070381',\n",
       " 'GO_0070382',\n",
       " 'GO_0070685',\n",
       " 'GO_0070701',\n",
       " 'GO_0070702',\n",
       " 'GO_0070703',\n",
       " 'GO_0070762',\n",
       " 'GO_0070820',\n",
       " 'GO_0070821',\n",
       " 'GO_0070852',\n",
       " 'GO_0070931',\n",
       " 'GO_0070938',\n",
       " 'GO_0070971',\n",
       " 'GO_0071001',\n",
       " 'GO_0071002',\n",
       " 'GO_0071004',\n",
       " 'GO_0071005',\n",
       " 'GO_0071006',\n",
       " 'GO_0071007',\n",
       " 'GO_0071011',\n",
       " 'GO_0071013',\n",
       " 'GO_0071254',\n",
       " 'GO_0071546',\n",
       " 'GO_0071547',\n",
       " 'GO_0071561',\n",
       " 'GO_0071598',\n",
       " 'GO_0071682',\n",
       " 'GO_0071782',\n",
       " 'GO_0071914',\n",
       " 'GO_0071920',\n",
       " 'GO_0071944',\n",
       " 'GO_0071953',\n",
       " 'GO_0072534',\n",
       " 'GO_0072562',\n",
       " 'GO_0072563',\n",
       " 'GO_0072686',\n",
       " 'GO_0072687',\n",
       " 'GO_0090543',\n",
       " 'GO_0090619',\n",
       " 'GO_0090636',\n",
       " 'GO_0090637',\n",
       " 'GO_0090651',\n",
       " 'GO_0090658',\n",
       " 'GO_0090724',\n",
       " 'GO_0090734',\n",
       " 'GO_0097013',\n",
       " 'GO_0097014',\n",
       " 'GO_0097038',\n",
       " 'GO_0097060',\n",
       " 'GO_0097165',\n",
       " 'GO_0097197',\n",
       " 'GO_0097208',\n",
       " 'GO_0097209',\n",
       " 'GO_0097225',\n",
       " 'GO_0097226',\n",
       " 'GO_0097227',\n",
       " 'GO_0097228',\n",
       " 'GO_0097229',\n",
       " 'GO_0097232',\n",
       " 'GO_0097233',\n",
       " 'GO_0097234',\n",
       " 'GO_0097268',\n",
       " 'GO_0097342',\n",
       " 'GO_0097356',\n",
       " 'GO_0097381',\n",
       " 'GO_0097386',\n",
       " 'GO_0097409',\n",
       " 'GO_0097413',\n",
       " 'GO_0097414',\n",
       " 'GO_0097418',\n",
       " 'GO_0097422',\n",
       " 'GO_0097427',\n",
       " 'GO_0097431',\n",
       " 'GO_0097433',\n",
       " 'GO_0097440',\n",
       " 'GO_0097441',\n",
       " 'GO_0097442',\n",
       " 'GO_0097443',\n",
       " 'GO_0097444',\n",
       " 'GO_0097447',\n",
       " 'GO_0097449',\n",
       " 'GO_0097450',\n",
       " 'GO_0097451',\n",
       " 'GO_0097453',\n",
       " 'GO_0097457',\n",
       " 'GO_0097462',\n",
       " 'GO_0097470',\n",
       " 'GO_0097471',\n",
       " 'GO_0097486',\n",
       " 'GO_0097487',\n",
       " 'GO_0097489',\n",
       " 'GO_0097504',\n",
       " 'GO_0097512',\n",
       " 'GO_0097513',\n",
       " 'GO_0097524',\n",
       " 'GO_0097539',\n",
       " 'GO_0097541',\n",
       " 'GO_0097542',\n",
       " 'GO_0097543',\n",
       " 'GO_0097545',\n",
       " 'GO_0097546',\n",
       " 'GO_0097575',\n",
       " 'GO_0097598',\n",
       " 'GO_0097629',\n",
       " 'GO_0097632',\n",
       " 'GO_0097708',\n",
       " 'GO_0097728',\n",
       " 'GO_0097729',\n",
       " 'GO_0097731',\n",
       " 'GO_0097733',\n",
       " 'GO_0098536',\n",
       " 'GO_0098541',\n",
       " 'GO_0098548',\n",
       " 'GO_0098552',\n",
       " 'GO_0098553',\n",
       " 'GO_0098554',\n",
       " 'GO_0098556',\n",
       " 'GO_0098559',\n",
       " 'GO_0098560',\n",
       " 'GO_0098562',\n",
       " 'GO_0098574',\n",
       " 'GO_0098575',\n",
       " 'GO_0098576',\n",
       " 'GO_0098577',\n",
       " 'GO_0098588',\n",
       " 'GO_0098591',\n",
       " 'GO_0098592',\n",
       " 'GO_0098594',\n",
       " 'GO_0098595',\n",
       " 'GO_0098674',\n",
       " 'GO_0098683',\n",
       " 'GO_0098684',\n",
       " 'GO_0098685',\n",
       " 'GO_0098686',\n",
       " 'GO_0098687',\n",
       " 'GO_0098688',\n",
       " 'GO_0098690',\n",
       " 'GO_0098691',\n",
       " 'GO_0098723',\n",
       " 'GO_0098793',\n",
       " 'GO_0098794',\n",
       " 'GO_0098804',\n",
       " 'GO_0098826',\n",
       " 'GO_0098830',\n",
       " 'GO_0098831',\n",
       " 'GO_0098833',\n",
       " 'GO_0098835',\n",
       " 'GO_0098837',\n",
       " 'GO_0098839',\n",
       " 'GO_0098842',\n",
       " 'GO_0098843',\n",
       " 'GO_0098844',\n",
       " 'GO_0098845',\n",
       " 'GO_0098846',\n",
       " 'GO_0098850',\n",
       " 'GO_0098857',\n",
       " 'GO_0098858',\n",
       " 'GO_0098871',\n",
       " 'GO_0098875',\n",
       " 'GO_0098888',\n",
       " 'GO_0098890',\n",
       " 'GO_0098894',\n",
       " 'GO_0098895',\n",
       " 'GO_0098944',\n",
       " 'GO_0098954',\n",
       " 'GO_0098965',\n",
       " 'GO_0098966',\n",
       " 'GO_0098978',\n",
       " 'GO_0098981',\n",
       " 'GO_0098982',\n",
       " 'GO_0098983',\n",
       " 'GO_0098984',\n",
       " 'GO_0098985',\n",
       " 'GO_0098992',\n",
       " 'GO_0098998',\n",
       " 'GO_0099012',\n",
       " 'GO_0099013',\n",
       " 'GO_0099020',\n",
       " 'GO_0099073',\n",
       " 'GO_0099091',\n",
       " 'GO_0099092',\n",
       " 'GO_0099147',\n",
       " 'GO_0099154',\n",
       " 'GO_0099160',\n",
       " 'GO_0099164',\n",
       " 'GO_0099182',\n",
       " 'GO_0099189',\n",
       " 'GO_0099192',\n",
       " 'GO_0099243',\n",
       " 'GO_0099400',\n",
       " 'GO_0099503',\n",
       " 'GO_0099513',\n",
       " 'GO_0099523',\n",
       " 'GO_0099524',\n",
       " 'GO_0099544',\n",
       " 'GO_0099569',\n",
       " 'GO_0099572',\n",
       " 'GO_0099573',\n",
       " 'GO_0099617',\n",
       " 'GO_0099629',\n",
       " 'GO_0099631',\n",
       " 'GO_0099634',\n",
       " 'GO_0099738',\n",
       " 'GO_0101003',\n",
       " 'GO_0101004',\n",
       " 'GO_0106139',\n",
       " 'GO_0110070',\n",
       " 'GO_0120001',\n",
       " 'GO_0120003',\n",
       " 'GO_0120025',\n",
       " 'GO_0120043',\n",
       " 'GO_0120044',\n",
       " 'GO_0120095',\n",
       " 'GO_0120098',\n",
       " 'GO_0120103',\n",
       " 'GO_0120111',\n",
       " 'GO_0120134',\n",
       " 'GO_0120135',\n",
       " 'GO_0120199',\n",
       " 'GO_0120200',\n",
       " 'GO_0120202',\n",
       " 'GO_0120206',\n",
       " 'GO_0120212',\n",
       " 'GO_0120219',\n",
       " 'GO_0120220',\n",
       " 'GO_0120234',\n",
       " 'GO_0120249',\n",
       " 'GO_0120259',\n",
       " 'GO_0120281',\n",
       " 'GO_0120282',\n",
       " 'GO_0120293',\n",
       " 'GO_0120336',\n",
       " 'GO_0120337',\n",
       " 'GO_0120338',\n",
       " 'GO_0140240',\n",
       " 'GO_0140268',\n",
       " 'GO_0140284',\n",
       " 'GO_0140445',\n",
       " 'GO_0140494',\n",
       " 'GO_0150001',\n",
       " 'GO_0150002',\n",
       " 'GO_0150014',\n",
       " 'GO_0150034',\n",
       " 'GO_0150048',\n",
       " 'GO_0150051',\n",
       " 'GO_0150053',\n",
       " 'GO_0160045',\n",
       " 'GO_0160056',\n",
       " 'GO_0160110',\n",
       " 'GO_0160111',\n",
       " 'GO_0160112',\n",
       " 'GO_0170047',\n",
       " 'GO_1901588',\n",
       " 'GO_1902636',\n",
       " 'GO_1902737',\n",
       " 'GO_1903349',\n",
       " 'GO_1903561',\n",
       " 'GO_1903754',\n",
       " 'GO_1904115',\n",
       " 'GO_1904724',\n",
       " 'GO_1904813',\n",
       " 'GO_1904856',\n",
       " 'GO_1904930',\n",
       " 'GO_1905720',\n",
       " 'GO_1905721',\n",
       " 'GO_1990005',\n",
       " 'GO_1990023',\n",
       " 'GO_1990026',\n",
       " 'GO_1990030',\n",
       " 'GO_1990031',\n",
       " 'GO_1990032',\n",
       " 'GO_1990037',\n",
       " 'GO_1990038',\n",
       " 'GO_1990047',\n",
       " 'GO_1990075',\n",
       " 'GO_1990324',\n",
       " 'GO_1990357',\n",
       " 'GO_1990385',\n",
       " 'GO_1990427',\n",
       " 'GO_1990435',\n",
       " 'GO_1990462',\n",
       " 'GO_1990498',\n",
       " 'GO_1990578',\n",
       " 'GO_1990635',\n",
       " 'GO_1990674',\n",
       " 'GO_1990675',\n",
       " 'GO_1990676',\n",
       " 'GO_1990716',\n",
       " 'GO_1990718',\n",
       " 'GO_1990742',\n",
       " 'GO_1990752',\n",
       " 'GO_1990769',\n",
       " 'GO_1990777',\n",
       " 'GO_1990780',\n",
       " 'GO_1990794',\n",
       " 'GO_1990812',\n",
       " 'GO_1990836',\n",
       " 'GO_1990876',\n",
       " 'GO_1990877',\n",
       " 'GO_1990909',\n",
       " 'GO_1990913',\n",
       " 'GO_1990917'}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_terms = {i.replace(\":\", \"_\") for i in go_terms}\n",
    "go_terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nodes before pruning: 2064\n"
     ]
    }
   ],
   "source": [
    "print(\"Nodes before pruning:\", len(G.nodes)) # 24k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes left 1940\n"
     ]
    }
   ],
   "source": [
    "# prune network\n",
    "def prune_tree(G, nodeset):\n",
    "    nodes_to_check = list(G.nodes)\n",
    "    while nodes_to_check:\n",
    "        node = nodes_to_check.pop()\n",
    "        if G.out_degree(node) == 0 and G.in_degree(node) > 0 and not node in nodeset: # if leaf is not in nodeset\n",
    "            parents = list(G.predecessors(node))\n",
    "            G.remove_node(node)\n",
    "            nodes_to_check.extend(parents)\n",
    "\n",
    "prune_tree(G, go_terms)\n",
    "\n",
    "print(\"Number of nodes left\", len(G.nodes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build full ontology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/tmp.20794777.dgankin/ipykernel_4025407/1703353129.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  snp_df['type'] = 'snp'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>parent</th>\n",
       "      <th>child</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO_0000137</td>\n",
       "      <td>GO_0031985</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO_0000138</td>\n",
       "      <td>GO_0031985</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO_0000139</td>\n",
       "      <td>GO_0098588</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO_0000220</td>\n",
       "      <td>GO_0033179</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO_0000221</td>\n",
       "      <td>GO_0033180</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644889</th>\n",
       "      <td>PASD1</td>\n",
       "      <td>rs58448897</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644890</th>\n",
       "      <td>GABRE</td>\n",
       "      <td>rs145754054</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644891</th>\n",
       "      <td>TREX2</td>\n",
       "      <td>rs5945389</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644892</th>\n",
       "      <td>SRPK3</td>\n",
       "      <td>rs199915647</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644893</th>\n",
       "      <td>IKBKG</td>\n",
       "      <td>rs5030868</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>716635 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            parent        child     type\n",
       "0       GO_0000137   GO_0031985  default\n",
       "1       GO_0000138   GO_0031985  default\n",
       "2       GO_0000139   GO_0098588  default\n",
       "3       GO_0000220   GO_0033179  default\n",
       "4       GO_0000221   GO_0033180  default\n",
       "...            ...          ...      ...\n",
       "644889       PASD1   rs58448897      snp\n",
       "644890       GABRE  rs145754054      snp\n",
       "644891       TREX2    rs5945389      snp\n",
       "644892       SRPK3  rs199915647      snp\n",
       "644893       IKBKG    rs5030868      snp\n",
       "\n",
       "[716635 rows x 3 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert the pruned graph to a dataframe for output\n",
    "pruned_edges = [(u, v) for u, v in G.edges]\n",
    "ontology_df = pd.DataFrame(pruned_edges, columns=['parent', 'child'])\n",
    "ontology_df[\"type\"] = \"default\"\n",
    "ontology_df\n",
    "\n",
    "# append gene to go term df\n",
    "tmp_df = merged_go[[\"index\",\"gene_name\"]].drop_duplicates()\n",
    "tmp_df['index'] = tmp_df['index'].str.replace(\":\",\"_\")\n",
    "tmp_df.columns = ['parent', 'child']\n",
    "tmp_df['type'] = 'gene'\n",
    "\n",
    "# append snp to gene df\n",
    "snp_df = go_bim[[\"gene_name\",\"snp\"]]\n",
    "snp_df.columns = ['parent', 'child']\n",
    "snp_df['type'] = 'snp'\n",
    "\n",
    "# concatenate\n",
    "ontology_df = pd.concat([ontology_df, tmp_df, snp_df])\n",
    "ontology_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nodes in Graph: 662311\n",
      "Edges in Graph: 716635\n",
      "Number of nodes left 662311\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>parent</th>\n",
       "      <th>child</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO_0000137</td>\n",
       "      <td>GO_0031985</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO_0000137</td>\n",
       "      <td>GOLGA8J</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO_0000137</td>\n",
       "      <td>GOLGA6A</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO_0000137</td>\n",
       "      <td>XYLT1</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO_0000137</td>\n",
       "      <td>ATL1</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       parent       child     type\n",
       "0  GO_0000137  GO_0031985  default\n",
       "1  GO_0000137     GOLGA8J     gene\n",
       "2  GO_0000137     GOLGA6A     gene\n",
       "3  GO_0000137       XYLT1     gene\n",
       "4  GO_0000137        ATL1     gene"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a directed graph from the ontology data\n",
    "OG = nx.DiGraph()\n",
    "\n",
    "# Add edges to the graph\n",
    "for index, row in ontology_df.iterrows():\n",
    "    OG.add_edge(row['parent'], row['child'], type=row['type']) # only add gene if in known genes\n",
    "\n",
    "print(\"Nodes in Graph:\", len(OG.nodes))\n",
    "print(\"Edges in Graph:\", len(OG.edges))\n",
    "\n",
    "#print(G.nodes)\n",
    "\n",
    "# Step 2: Iteratively remove nodes that do not lead to a known gene\n",
    "def prune_tree_(G):\n",
    "    nodes_to_check = list(G.nodes)\n",
    "    while nodes_to_check:\n",
    "        node = nodes_to_check.pop()\n",
    "        if G.out_degree(node) == 0 and G.in_degree(node) > 0 and \"GO\" in node:  # Check if it is a hanging node\n",
    "            print(\"yas\")\n",
    "            parents = list(G.predecessors(node))\n",
    "            G.remove_node(node)\n",
    "            nodes_to_check.extend(parents)\n",
    "\n",
    "prune_tree_(OG)\n",
    "\n",
    "print(\"Number of nodes left\", len(OG.nodes))\n",
    "\n",
    "# Convert the pruned graph back to a dataframe for output\n",
    "pruned_edges = [(u, v, OG[u][v]['type']) for u, v in OG.edges]\n",
    "pruned_df = pd.DataFrame(pruned_edges, columns=['parent', 'child', 'type'])\n",
    "\n",
    "# Save the pruned ontology\n",
    "#pruned_ontology_file = '../data/go_biological_processes.txt'\n",
    "pruned_ontology_file = '../data/go_cellular_components.txt'\n",
    "pruned_df.to_csv(pruned_ontology_file, sep='\\t', index=False)\n",
    "\n",
    "pruned_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>parent</th>\n",
       "      <th>child</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO_0000137</td>\n",
       "      <td>GO_0031985</td>\n",
       "      <td>default</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO_0000137</td>\n",
       "      <td>GOLGA8J</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO_0000137</td>\n",
       "      <td>GOLGA6A</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO_0000137</td>\n",
       "      <td>XYLT1</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO_0000137</td>\n",
       "      <td>ATL1</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716630</th>\n",
       "      <td>RPL18</td>\n",
       "      <td>rs12608448</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716631</th>\n",
       "      <td>HOXA3</td>\n",
       "      <td>rs118016485</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716632</th>\n",
       "      <td>TAF6</td>\n",
       "      <td>rs143557149</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716633</th>\n",
       "      <td>TPRN</td>\n",
       "      <td>rs144769563</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716634</th>\n",
       "      <td>NOMO1</td>\n",
       "      <td>rs137963033</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>716635 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            parent        child     type\n",
       "0       GO_0000137   GO_0031985  default\n",
       "1       GO_0000137      GOLGA8J     gene\n",
       "2       GO_0000137      GOLGA6A     gene\n",
       "3       GO_0000137        XYLT1     gene\n",
       "4       GO_0000137         ATL1     gene\n",
       "...            ...          ...      ...\n",
       "716630       RPL18   rs12608448      snp\n",
       "716631       HOXA3  rs118016485      snp\n",
       "716632        TAF6  rs143557149      snp\n",
       "716633        TPRN  rs144769563      snp\n",
       "716634       NOMO1  rs137963033      snp\n",
       "\n",
       "[716635 rows x 3 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pruned_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
